{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# COVID-19 Open Research Dataset Challenge (CORD-19)\n",
    "\n",
    "To get all the papers go here:\n",
    "https://www.kaggle.com/allen-institute-for-ai/CORD-19-research-challenge"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Parsing docs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Python libraries have been imported.\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import json\n",
    "from pathlib import Path\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "# If you need to parse docs of differend kind than JSON, use the libraries below:\n",
    "# import cv2\n",
    "# import fitz\n",
    "# import docx2txt\n",
    "# import textract\n",
    "# import pytesseract\n",
    "\n",
    "print(\"Python libraries have been imported.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Parser:\n",
    "    def __init__(self, path):\n",
    "        self.path = path\n",
    "        \n",
    "#     def readExcel(self, path):\n",
    "#         xls = pd.read_excel(path)\n",
    "#         return xls.to_string()\n",
    "\n",
    "    # Jsons from COVID competition\n",
    "    def readJson(self, path):\n",
    "        json_text = \"\"\n",
    "        with open(path) as file:\n",
    "            paper_in_json = json.load(file)\n",
    "            text_list = paper_in_json['body_text']\n",
    "            if len(text_list) > 0:\n",
    "                for text_dict in text_list:\n",
    "                    json_text += text_dict['text']+\" \"  \n",
    "        return json_text   \n",
    "\n",
    "    def readExcel(self, path):\n",
    "        xls = pd.read_excel(path, None)\n",
    "        xls_text = ''\n",
    "        for df in xls.values():\n",
    "            xls_text += df.to_string()+\" \"\n",
    "        return xls_text    \n",
    "    \n",
    "    def readImg(self, path):\n",
    "        img = cv2.imread(path)\n",
    "        return pytesseract.image_to_string(img)\n",
    "    \n",
    "    def readTxt(self, path):\n",
    "        text = ''\n",
    "        with open(path,'r') as f:\n",
    "            text = f.read()\n",
    "        return text\n",
    "    \n",
    "    def readPdf(self, path):\n",
    "        doc = fitz.open(path)\n",
    "        text = ''\n",
    "        for page in doc:\n",
    "            text += page.getText(\"text\")+\" \"\n",
    "        return text\n",
    "    \n",
    "    def readDocx(self, path):\n",
    "        text = docx2txt.process(path,img_dir=\n",
    "                                'imgs_from_docs')\n",
    "        return text\n",
    "    \n",
    "    def readDoc(self, path):\n",
    "        text = textract.process(path).decode('utf-8')\n",
    "        return text\n",
    "\n",
    "    def readFiles(self):\n",
    "        text_list = []\n",
    "        for root, dirs, files in os.walk(self.path):\n",
    "            for filename in files:\n",
    "                if not filename.startswith('._'):         \n",
    "                    if Path(filename).suffix.lower() in ['.jpg', '.png','.jpeg']:\n",
    "                        print('Parsing doc: ',filename)\n",
    "                        text = self.readImg(os.path.join(root, filename))\n",
    "                        if text.strip() != '':\n",
    "                            text_list.append({'title':filename,'content':text})\n",
    "                            # print(text)\n",
    "                    if Path(filename).suffix.lower() in ['.xls','.xlsx']:\n",
    "                        print('Parsing doc: ',filename)\n",
    "                        text = self.readExcel(os.path.join(root, filename))\n",
    "                        if text.strip() != '':\n",
    "                            text_list.append({'title':filename,'content':text})\n",
    "                            # print(text)\n",
    "                    if Path(filename).suffix.lower() == '.txt':\n",
    "                        print('Parsing doc: ',filename)\n",
    "                        text = self.readTxt(os.path.join(root, filename))\n",
    "                        if text.strip() !=  '':\n",
    "                            text_list.append({'title':filename,'content':text})\n",
    "                            # print(text)   \n",
    "                    if Path(filename).suffix.lower() == '.pdf':\n",
    "                        print('Parsing doc: ',filename)\n",
    "                        text = self.readPdf(os.path.join(root, filename))\n",
    "                        if text.strip() !=  '':\n",
    "                            text_list.append({'title':filename,'content':text})\n",
    "                            # print(text)    \n",
    "                    if Path(filename).suffix.lower() == '.docx':\n",
    "                        print('Parsing doc: ',filename)\n",
    "                        text = self.readDocx(os.path.join(root, filename))                \n",
    "                        if text.strip() !=  '':\n",
    "                            text_list.append({'title':filename,'content':text})\n",
    "                            # print(text)           \n",
    "                    if Path(filename).suffix.lower() == '.doc':\n",
    "                        print('Parsing doc: ',filename)\n",
    "                        text = self.readDoc(os.path.join(root, filename))                    \n",
    "                        if text.strip() !=  '':\n",
    "                            text_list.append({'title':filename,'content':text})\n",
    "                            # print(text)\n",
    "                    if Path(filename).suffix.lower() == '.json':\n",
    "                        print('Parsing doc: ',filename)\n",
    "                        text = self.readJson(os.path.join(root, filename))\n",
    "                        if text.strip() !=  '':\n",
    "                            text_list.append({'title':filename,'content':text})                             \n",
    "        return text_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Parsing doc:  PMC1636417.xml.json\n",
      "Parsing doc:  PMC1636460.xml.json\n",
      "Parsing doc:  PMC1635287.xml.json\n",
      "Parsing doc:  PMC1616946.xml.json\n",
      "Parsing doc:  PMC2115309.xml.json\n",
      "Parsing doc:  PMC1994808.xml.json\n",
      "Parsing doc:  PMC2106561.xml.json\n",
      "Parsing doc:  PMC1994780.xml.json\n",
      "Parsing doc:  PMC1807961.xml.json\n",
      "Parsing doc:  PMC2116283.xml.json\n",
      "Parsing doc:  PMC1802563.xml.json\n",
      "Parsing doc:  PMC1934988.xml.json\n",
      "Parsing doc:  PMC2118498.xml.json\n",
      "Parsing doc:  PMC2115421.xml.json\n",
      "Parsing doc:  PMC2115043.xml.json\n",
      "Parsing doc:  PMC1781101.xml.json\n",
      "Parsing doc:  PMC2114808.xml.json\n",
      "Parsing doc:  PMC2114809.xml.json\n",
      "Parsing doc:  PMC2115162.xml.json\n",
      "Parsing doc:  PMC2018615.xml.json\n",
      "Parsing doc:  PMC2116277.xml.json\n",
      "Parsing doc:  PMC2116293.xml.json\n",
      "Parsing doc:  PMC2119224.xml.json\n",
      "Parsing doc:  PMC2115898.xml.json\n",
      "Parsing doc:  PMC1914094.xml.json\n",
      "Parsing doc:  PMC1874606.xml.json\n",
      "Parsing doc:  PMC2115535.xml.json\n",
      "Parsing doc:  PMC1616953.xml.json\n",
      "Parsing doc:  PMC2114261.xml.json\n",
      "Parsing doc:  PMC2115698.xml.json\n",
      "Parsing doc:  PMC1976451.xml.json\n",
      "Parsing doc:  PMC1933184.xml.json\n",
      "Parsing doc:  PMC2115513.xml.json\n",
      "Parsing doc:  PMC2118742.xml.json\n",
      "Parsing doc:  PMC2115831.xml.json\n",
      "Parsing doc:  PMC2113917.xml.json\n",
      "Parsing doc:  PMC2113985.xml.json\n",
      "Parsing doc:  PMC1751549.xml.json\n",
      "Parsing doc:  PMC2113797.xml.json\n",
      "Parsing doc:  PMC1616970.xml.json\n",
      "Parsing doc:  PMC2114239.xml.json\n",
      "Parsing doc:  PMC1636451.xml.json\n",
      "Parsing doc:  PMC2115992.xml.json\n",
      "Parsing doc:  PMC1919510.xml.json\n",
      "Parsing doc:  PMC2115569.xml.json\n",
      "Parsing doc:  PMC2113407.xml.json\n",
      "Parsing doc:  PMC2018635.xml.json\n",
      "Parsing doc:  PMC1669763.xml.json\n",
      "Parsing doc:  PMC1933182.xml.json\n",
      "Parsing doc:  PMC1636472.xml.json\n",
      "All docs have been parsed.\n"
     ]
    }
   ],
   "source": [
    "parser = Parser('covid-19-papers')\n",
    "docs_list = parser.readFiles()\n",
    "print('All docs have been parsed.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 PMC1636417.xml.json\n",
      "1 PMC1636460.xml.json\n",
      "2 PMC1635287.xml.json\n",
      "3 PMC1616946.xml.json\n",
      "4 PMC1994808.xml.json\n",
      "5 PMC1994780.xml.json\n",
      "6 PMC1807961.xml.json\n",
      "7 PMC1802563.xml.json\n",
      "8 PMC1934988.xml.json\n",
      "9 PMC2118498.xml.json\n",
      "10 PMC1781101.xml.json\n",
      "11 PMC2018615.xml.json\n",
      "12 PMC1914094.xml.json\n",
      "13 PMC1874606.xml.json\n",
      "14 PMC1616953.xml.json\n",
      "15 PMC1976451.xml.json\n",
      "16 PMC1933184.xml.json\n",
      "17 PMC1751549.xml.json\n",
      "18 PMC1616970.xml.json\n",
      "19 PMC1636451.xml.json\n",
      "20 PMC1919510.xml.json\n",
      "21 PMC2018635.xml.json\n",
      "22 PMC1669763.xml.json\n",
      "23 PMC1933182.xml.json\n",
      "24 PMC1636472.xml.json\n"
     ]
    }
   ],
   "source": [
    "for nr, doc in enumerate(docs_list):\n",
    "    print(nr,doc['title'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of documents: 25\n"
     ]
    }
   ],
   "source": [
    "print('Number of documents:',len(docs_list))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'title': 'PMC1636417.xml.json',\n",
       " 'content': \"For both surveillance and diagnostic applications, fine-scale pathogen identification and near-neighbor discrimination is important; therefore, an assay that monitors at this very specific level is desirable for many types of samples such as clinical and environmental (1–3). To successfully use any method based on DNA or RNA detection, these assays must be coupled with large databases of nucleic acid sequence information for assay design to ensure that the desired information is provided and for the interpretation of raw data. Several well-established techniques use PCR to amplify individual target pieces of sequenced genomes to provide detection of organisms (4). These methods can roughly be divided into approaches that target individual short sequence lengths or probes (<40 bp) and methods that examine longer probes. The advantage of using short probes is that when the uniqueness of the probe has been assured and unique primers are also selected, this method gives good specificity. This approach is capable of providing fine-scale identification of several genetically close organisms by selecting a sufficient number of probes. However, this can rapidly lead to a very large number of total probes being required to detect all organisms of interest. In addition these selected probes, which in the initial selection process were determined to be unique, are often later found to be less specific as more organisms are sequenced or are less specific under conditions that differ from the original conditions. This is particularly a problem for organisms belonging to a family with a high mutation rate and also for pathogens that have relatively few neighboring pathogens sequenced. In addition, PCR approaches focused on short unique probes are not capable of detecting the presence of new significant mutations nor can they easily resolve base sequence details. Approaches that use longer individual probes avoid many of these issues at the cost of being less specific. This issue means most of these approaches are not suitable for providing the information desired, providing impetus to this work. High-density resequencing microarrays produce variable length segments, 102–105 bp, of direct sequence information. This target sequence falls in the longer target regime of PCR approaches but rather than being hybridized to a longer less-specific probe on the microarray, many shorter specific probes are placed on the microarray to allow more detailed determinations from the entire PCR amplicon. This also means that the specificity of the primers used can be relaxed. They have been successfully used to detect single nucleotide polymorphism (SNP) and genetic variants from viral, bacterial and eukaryotic genomes (5–12). Their use for SNP detection has clearly established their ability to provide reliable quality sequence information. In most cases, the microarrays were designed to study a limited number of genetically similar target pathogens and for many cases, the detection methods relied only on recognizing hybridization patterns for identification (6,9,10,13,14). Taking advantage of the sequential base resolution capability of resequencing microarrays that is required for SNP detection, resequencing has recently been successfully adapted recently using a different approach for organism identification of multiple bacterial and viral pathogens while allowing for fine detailed discrimination of closely related organisms and tracking mutations within the targeted pathogen (15–16). The new methodology differed from earlier work by using the resolved bases as the query of a similarity search of DNA databases to identify the most likely species and variants that match the base calls from the hybridization observed. The system was capable of testing for 26 pathogens simultaneously and could detect the presence of multiple pathogens. A software program, resequencing pathogen identifier (REPI), was used to simplify data analysis by performing similarity searches of a genetic database using basic local alignment search tool (BLAST) (17). The REPI program used BLAST default settings and would only return sequences that might represent the hybridization if the expect value, a quantity calculated by the BLAST program that indicates the likelihood that the sequence match found would have occurred by random chance in the database, was <10−9. This screened out all cases that had insufficient signal; however, the final determination of what pathogen(s) was detected and to what degree discrimination was possible required manual examination of the returned results. This method successfully allowed fine discrimination of various adenoviruses and strain identifications of Flu A and B samples in agreement with conventional sampling results (15,16). Two important advantages of this approach were that the information was always recovered at the most detailed level possible and that it was capable of still recognizing organisms with recent mutations. This approach also maintained specificity well, as it was not dependent on the uniqueness of a few individual short probes. Although this analysis method has utility, there are several shortcomings: it is time consuming, not optimized to maximize sensitivity, has complicated results, is suitable only for an expert, and contains redundant or duplicate information. The process was time consuming because only the initial screening was handled automatically while the remaining steps required manual interpretation before the detection analysis was complete. Because a simple criterion (expect value cutoff of 10−9) and non-optimized BLAST parameters were used to consider a pathogen detected, the REPI algorithm provided a list of candidate organisms but did not make a final simple conclusion or relate the results of one prototype sequence to another. Instead a manual process was used to make the final determination, but because the REPI program provided all similar results and the use of public nucleic acid databases containing redundant entries, a large amount of data was presented to a user that was not useful. In addition, with a manual process it was not possible to establish that the algorithm developed was generally applicable for any organism where nucleic acid base resolved sequence information has been provided. In this paper, we describe a new software expert system, Computer-Implemented Biological Sequence Identifier system 2.0 (CIBSI 2.0), that successfully uses resolved base sequence information from custom designed Affymetrix resequencing microarrays to provide a simple list of organisms that are detected. This algorithm addresses the most important shortcoming of previous methods by incorporating new features to completely automate pathogen identification. We have demonstrated the effectiveness of this algorithm for identification via several examples. The single program is capable of making correct decisions for all 26 pathogens contained on the Respiratory Pathogen Microarray v.1 (RPM v.1), whether detected alone or in combinations, with improved sensitivity. Although the program is currently applied to resequencing microarrays, the methodologies developed remain generally applicable. Only the first portion of the algorithm handles issues specific to microarrays while the remainder deals with sequences that are suitable for use as a query by the BLAST algorithm. In developing the general identification algorithm, we have identified and resolved issues specific to resequencing microarrays that complicate their use. Because the entire decision process for what is detected has been automated, it is straightforward to test whether the rules used to make identifications are rigorous and applicable to any pathogen. With this efficient program, resequencing based assays can provide a competitive method to test simultaneously for many possible pathogens, providing output that can be interpreted by a non-expert. The details of the RPM v.1 design and the experimental methods have been discussed in previous work (15,16,18) (Lin et al., submitted for publication). Briefly, the RPM v.1 chip design includes 57 tiled regions allowing resequencing of 29.7 kb of sequences from 27 respiratory pathogens and biothreat agents. These were selected based upon clinical relevance for the population of immediate interest (United States military recruit in training) (19–21). Partial sequences from the genes containing diagnostic regions were tiled for the detection of these pathogens. The experimental microarray data used in the present analysis were obtained using a variety of purified nucleic acid templates and clinical samples culture (throat swabs and nasal washes) using random and multiplexed RT–PCR amplification schemes (for more detail description of amplification methods see Supplementary Data). Resequencing microarrays provide base call resolution by comparing the intensities between a set of four 25mer probes that differ from each other at the same position (13th base). An amplicon or target sequence is represented by numerous overlapping probe sets. GCOS™ software v1.3 (Affymetrix Inc., Santa Clara, CA) was used to align and scan hybridized microarrays to determine the intensity of each probe in every probe set. Base calls were made based on the intensity data of each probe set using GDAS v3.0.2.8 software (Affymetrix Inc.) which used an implementation of the ABACUS algorithm (5). The sequences were represented in FASTA format for later analysis steps. In this paper, target pathogens are the organisms the assay was specifically designed to detect. The sets of probes that represent reference sequence selected from target pathogen genomes are referred to as a Prototype Sequence or ‘ProSeq’ for brevity. The set of resolved bases that result from hybridization of genomic material to a ProSeq is referred to as the hybridized sequence or ‘HybSeq’. The HybSeq is split into possible subsequences or ‘SubSeqs’. The CIBSI 2.0 program implemented in Perl described in this study handled a hierarchy of three tasks (Figure 1): (I) ProSeq identification; (II) ProSeq grouping; and (III) pathogen determination. The most developed and important portion of the algorithm deals with ProSeq identification Task(I) and is handled in three important subtasks: initial filtering of individual HybSeqs into SubSeqs suitable for sequence similarity comparisons (Figure 2), database querying of individual SubSeqs (Figure 3) and taxonomic comparison of BLAST returns for each SubSeq (Figure 4). The NCBI BLAST and taxonomy databases were used for the queries and images were obtained on February 7, 2006. For the ProSeq grouping Task(II), ProSeqs were compared to determine if they supported the same identified organism. In the pathogen determination Task(III), detected organisms were compared to the list of target pathogens the assay was designed for in order to determine if any were positively detected or were possibly related close genetic near neighbors. The level of discrimination that a particular sample supported was automatically determined. An initial filtering algorithm, REPI, was developed previously (16) and the general concepts with revisions were incorporated into the current (automated detection) algorithm used in the CIBSI 2.0 program. Filtering and subsequence selection were used to remove potential biasing caused by reference sequence choice and by other sources (i.e. primers). When PCR amplification was used, microarrays were hybridized in the presence of only primers to determine locations where they resulted in hybridization. Any portions of the ProSeqs that hybridized with the primers were masked as N calls so that the HybSeq did not contain biased information. Normally the primers are designed to be outside the ProSeq region to minimize the interference caused by primers, and so minimize the bases to be masked. There is still the chance that some bases require masking because with the large number of primers used in the multiplex, short stretches of a ProSeq not corresponding to primer locations may still hybridize with the primers. Such regions could be removed from the reference sequences and so not appear on the microarray. However, determining such locations are a difficult and time-consuming task that for most cases is not worth the effort. The first subtask of ProSeq identification Task(I) is noted in Figure 1 and shown schematically in detail in Figure 2. This subtask uses a procedure to examine a HybSeq to find the longest possible subsequence of base calls (SubSeq) that can be submitted as a query to BLAST. It produces a group of SubSeq that contain all portions of a HybSeq that have a chance of producing a limited list of returns from a BLAST query. When a HybSeq has two regions separated by a long stretch of continuous N calls, the relational positioning of the two regions cannot be trusted and so must be sent as separate queries. In addition for shorter subsequences, the number of base calls that must be made is dependent on the length. It was also recognized that for very long sequences a longer WORD size in BLAST may be used. A detailed description of the criteria and process used for each step is contained in Supplementary Data. Upon completion, the algorithm returned to the Task(I) loop and performed the BLAST subtask. The database query subtask performed a batch similarity search of a database using SubSeq as the queries. The BLAST program used was the NCBI Blastall –p blastn version 2.12 with a defined set of parameters. The masking of low complex regions was performed for the seeding phase to speed up the query; however, low complexity repeats were included in the actual scoring. The entire nucleotide database from NCBI acquired on February 7, 2006 was used as the reference database. (Note that earlier images of the database were used during development but all experiments were rerun with the algorithm as described with the image of the database obtained on this date.) The default gap penalty and nucleotide match score were used. The nucleotide mismatch penalty, –q, parameter was set to −1 rather than the default. The results of any BLAST query with an expect value <0.0001 were returned in tabular format from the blastall program. The information about each return (bit score, expect value, mismatches, length of match) was placed in the Return{hash key}{info} hash using the SubSeq identity as the hash key for further analysis. The next subtask of ProSeq identification Task(I) carried out was the determination of SubSeq() states and is shown in Figure 3. The BLAST algorithm gives a ranking score which can be reported as accounting for the size of the database (expect value) or not (bit score). The full taxonomic classifications of every return for a SubSeq were retrieved from the NCBI taxonomy database obtained on February 7, 2006. Using the scores and taxonomy relationships it was possible to find a reduced number of returns that had the best match with the HybSeq. These results were summarized by identifying the taxonomic class to which all the returns belonged to, ‘identified organism’, and a parameter that indicated how they are related to each other, ‘organism uniqueness’. A detailed description of the steps is contained in Supplementary Data. After each SubSeq was examined, the algorithm moved to the next subtask, which was to determine the identified organism of the ProSeq from the SubSeq (Figure 4). The subsequences from the same ProSeq were only allowed to support a single ‘identified organism’ determination. The procedure shown in Figure 4 demonstrates the decision method used to arrive at this determination (detailed description in Supplementary Data). After the subtask covered in Figure 4 was completed, the ProSeq identification Task(I) loop continued until all ProSeqs were examined. A list of ProSeqs that had detected organisms was built up in the Result1 array. After the ProSeq identification Task(I) was completed, ProSeq grouping Task(II) (Figure 1) was used to examine the identified organism values listed in Result1 and grouped them together if they identified the same taxonomic class. Each entry in Result1 was examined and a new entry was created in Result2 if the identified organism did not appear in this list. The entries of Result2 represented the distinct individual organisms identified, but might still contain redundant information. When the ProSeqs were designed to detect the same organism and they all hybridized well, this grouping led to a reduction in redundant information being reported. But, when one ProSeq did not hybridize as well for a variety of possible reasons, multiple entries would appear in Result2 that actually represent hybridization from the same pathogen. This is because there is an alternative cause for the ProSeq hybridizing in this manner. This hybridization could be caused by two different but closely related organisms both being present in a sample and hybridizing to the microarray. Because we have not yet developed methods to distinguish these cases, no further reduction of the list of organisms is made for in ProSeq grouping Task(II) in cases where the level of identification varied on different ProSeq targeted for the same organism. Although it was difficult to relate results from separate ProSeqs to each other, it was desired to have a simple final detection decision be made in pathogen determination Task(III). The first task was specifically implemented so that information about what a ProSeq was intended to detect was not considered and the second task only minimal consideration of this was taken into account. This allowed these initial tasks to be capable of recognizing not only just positive and negative identifications of target pathogens but also cases that were indeterminate. In the final task, the algorithm considered whether the identified organisms belonged to the list of organisms the ProSeqs were designed to detect. The task would group organisms from ProSeq grouping Task(II) together that belonged to or were child classes of the taxonomic class of a target organism. The taxonomic class reported was the common taxonomic group of all the organisms. When all the ProSeqs for a pathogen hybridized well, a fine level discrimination was reported. But if one or more ProSeqs hybridized less well, the reported positive target pathogen was only identified at the level of the less detailed level. This is conservative because methods have not yet been developed to clearly discriminate mixtures of very closely related organisms causing different ProSeqs to hybridize from variable hybridization of a single organism on several ProSeqs. The results of all three tasks were reported and a more experienced user can view ProSeq grouping Task(II) results to clarify some cases. Note that organisms identified in ProSeq grouping Task(II) that only belonged to target pathogens were reported as positives. Clear negative ProSeqs were not mentioned in the output. ProSeqs that were indeterminate or that detected close genetic organisms were never reported as positives. These organisms were instead reported as being detected. Purified Chlamydia pneumoniae nucleic acid samples with 10–1000 genome copies (via method in Lin et al., submitted for publication) were chosen to illustrate how pathogen detection and identification were done when multiple ProSeqs were targeted for the same pathogen. RPM v.1 has three highly conserved ProSeqs selected from the genes encoding for the major outer membrane proteins VD2 and VD4, and the DNA-directed RNA polymerase (rpoB) gene. The HybSeqs from the different samples differed only in the number of unique base calls as shown in Table 1. The percentage of the ProSeq called varied from 80 to 100% except for one case at a concentration of 10 that had only 11% of the rpoB ProSeq producing unique base calls. Because the samples at this concentration are not reproducibly generating the same percentage of base calls, this is probably the detection limit of this ProSeq of the assay. Table 1 listed the determinations made for the SubSeq and at the end of each task for the various samples. The ProSeq from the different cases produced the same number of SubSeqs. These SubSeqs from different samples reported different bit scores for the same top ranked returns from BLAST. In fact VD2 and VD4 produced exactly the same results. The NCBI taxonomy database classified the returns into four distinct groups, which represented the C.pneumoniae taxonomic group and three child strain groups. AE001652, AE002167, AE017159 and BA000008 appeared in the returns of all the ProSeqs for each sample, since they represented database entries of completely sequenced genomes. One rpoB SubSeq produced for its organism uniqueness, SeqUniqu. All other SubSeqs were TaxAmbig as multiple returns from different taxonomic classes were returned. Since the VD2 and VD4 ProSeq each have a single SubSeq, Task(I) assigned the ProSeq the state of the SubSeq. For the rpoB ProSeq, the bit score of one SubSeq was large enough that the algorithm assigned that SubSeq's identification to the ProSeq. Task II of the algorithm grouped all three ProSeqs together since they all had the same identified organism and TaxAmbig was assigned. The result of Task(III) was positive for target pathogen C.pneumoniae and this decision was straightforward as all the ProSeqs agreed with each other and belonged to the same target pathogen taxonomic class. Although the rpoB ProSeq was SeqUniqu, this was not the final conclusion for Task(II) as the ProSeq that was SeqUniqu was not the child taxonomic group and other ProSeq were TaxAmbig. The three recognized strains scored the same, which indicated that the sequence selected for the ProSeqs was very conserved and would not allow discrimination between the strains. Influenza and Human Adenovirus (HAdV) were the only pathogens that had ProSeq selected that would permit detailed strain level discrimination as discussed in previous work (15,16). This previous work using manual analysis found that the microarray results were in excellent agreement with the conventional sequencing results for clinical samples. A few of the results of running the CIBSI 2.0 program using the updated NCBI database on the raw microarray results are presented in Table 2 (the results for all samples used in the previous work are presented in Supplementary Table A). The identified organisms were not identical to the original findings due to the difference in database used and because all ProSeqs were considered rather than only the Flu A and B hemagglutinin. In fact, the conventional sequencing results that were submitted to NCBI from that work were found for every sample to be among the returns with the best score for the hemagglutinin ProSeq (Supplementary Table B). It should be noted that the previous work based its analysis upon only the results of the hemagglutinin ProSeq. For 8 of 13 Influenza A and 3 of 12 Influenza B cases, the results of ProSeq identification Task(I) and ProSeq grouping Task(II) found that the conventional sequencing was the single best return for the hemagglutinin ProSeq. Owing to the large number of isolate sequences in the database for the hemagglutinin gene it was not surprising that in some cases a single unique entry was not found. In each of the remaining five Influenza A samples, the other sequences returned differed by <0.2% from the conventional sequence. The fewer samples with unique isolate identifications for Influenza B were due to an older reference sequence used for the ProSeq, which allowed less hybridization to occur (18). This also meant that when multiple sequences were returned for a sample they represented greater genetic variation, up to 2%. As a result of the current method of making pathogen determination Task(III) level identification, the final organism reported was less specific (H3N2 or Flu B) for every sample than what was reported as possible in ProSeq grouping Task(II). For HAdV samples, the algorithm also reproduced the finer scale discriminations that had been made previously by manual methods (data not shown). The next example of detection for the Mycoplasma pneumoniae pathogen demonstrated a case where there was only a single ProSeq for the target pathogen. A total of 48 test samples were performed using multiplex PCR (via method in Lin et al., submitted for publication) where for 46 of the samples M.pneumoniae organism was spiked into nasal wash with several other pathogens from 100 to 100 000 colony forming units per ml, the remaining 2 samples were purified with nucleic acid from culture stock at a concentration of 1000 genome copies per reaction volume. This ProSeq was also not optimal for fine discrimination because it was selected from a highly conserved region (345 bp) of the cytadhesin P1 gene. In every case taxonomic database entries for M.pneumoniae or its one recognized distinct strain tied for MaxScore (Supplementary Table 3). To better understand these returns, the database sequences were examined and subdivided into three groups of sequences, A, B and C, based on how well they matched the reference sequence used to make the ProSeq. The placement of the database entries into the three groups was determined from a CLUSTAL alignment of the sequences of this gene. This alignment confirmed that the database entries differed significantly more from each other in regions not represented by the ProSeq and contained sufficient variability that would have allowed finer discrimination. Members of Group A exactly matched the ProSeq and could not be distinguished between on the microarray. Similarly, members of group B matched the ProSeq except at the 199th position where the base called was C rather than T. Group C sequences contained a few database entries that were more variable and might be distinguished from other entries within the ProSeq. For the 48 experimental tests of M.pneumoniae, as much as 80% of the ProSeq hybridized for 19 samples, yet only 5 of these samples had an unambiguous base call at the 199th position. When it was unambiguous, it always matched group B sequences. In the cases where an N base call was made at the 199th location, both groups A and B sequences were returned with the same score. Regardless of this, the target pathogen positively identified was M.pneumoniae for every sample tested. These examples showed how decisions were made independent of whether single or multiple ProSeqs were dedicated to a target pathogen. They also illustrated that the level of discrimination possible was strongly determined by the quality of the selected ProSeq. It is possible that for some pathogens fine level discrimination is not required and the currently tested selections on RPM v.1 would provide satisfactory information. The CIBSI 2.0 algorithm demonstrated its capability to automatically report the maximum level of discrimination that could be supported by the HybSeq information. To demonstrate how the algorithm handled closely related genetic species, a sample of a non-targeted pathogen was considered using multiplex PCR (via method in Lin et al., submitted for publication). For Variola major virus, one of the biothreat pathogens on the RPMv.1, the validation runs demonstrated that Variola major virus purified DNA templates of plasmids were always positively identified when detected (Table 3). Table 4 shows the results when purified Vaccinia genomic DNA was spiked into nasal washes and processed at various concentrations using multiplex PCR. The array has two ProSeqs from hemagglutinin (VMVHA, ∼500 bp) and cytokine response modifier B (VMVcrmB, ∼300 bp) genes for Variola major virus detection. The percentage of the ProSeq that hybridizes is sufficient that if hybridization patterns were only considered one might assume that this tile is identifying the presence of its target. This would indicate that reference sequence selected was not the best choice. However, when our algorithm was applied none of the samples is in fact identified as Variola major or minor virus. Vaccinia was always one of the Orthopoxvirus species listed with the highest scores for VMVcrmB ProSeq, but in only seven cases was it uniquely identified as the probable species detected. In only one sample at the lowest concentration and fraction of VMVcrmB hybridizing, did this ProSeq even identify Variola major and minor virus as one among the Orthopoxvirus species that could be the cause of the hybridization. The lower limit of detection for the amplification method used was between this concentration and the one above it for Variola major itself. The VMVHA ProSeq exhibited much lower sensitivity and made identifications of Orthopoxvirus species in only two experiments and Variola major virus was listed as one of the tied best scoring returns. In both cases, VMVcrmB ProSeq specifically identified Vaccinia virus as the best match. The percentage of the hybridized ProSeq correlated with concentration of the sample. This example demonstrated the importance of the filtering portion of the algorithm by considering the HybSeqs of the ProSeqs for the H1N1 neuraminidase (NA1) and matrix genes from human Influenza A/Puerto Rico/8/34 (H1N1) strain. Filtering was necessary because sending the HybSeq of a ProSeq to BLAST in a single query can bias the scores against strains that have insertions or deletions relative to the ProSeq, especially when using BLAST parameters that maximized the use of base calls. The sliding window test was the portion of the algorithm that controlled filtering. If filtering were turned off, the entire HybSeq would be used in a single subsequence for two influenza ProSeqs that showed significant hybridization. A/Weiss/43 (H1N1) strain was identified as the most likely strain from the HybSeq of the NA1 ProSeq while the HybSeq of the matrix ProSeq correctly identified A/Puerto Rico/8/34. To better understand the source of biasing, CLUSTAL alignment of the NA1 gene of the two strains and the reference sequence used to make the ProSeq are shown in Figure 5. The two strains showed 95% identity (67 mismatches in 1362 aligned bases); however, there was a stretch of 45 bases inserted in both A/Weiss/43 and the NA1 ProSeq compared to A/Puerto Rico/8/34. With the default filtering on, the NA1 ProSeq was split into five SubSeqs as the algorithm encountered large stretches of no calls. In ProSeq identification Task(I), the algorithm determined that three of the SubSeqs, which were shorter, had an identified organism of H1N1 as several isolates including A/Puerto Rico/8/34 tied with the best score, while the two other SubSeqs had an identified organism of only the A/Puerto Rico/8/34 strain as being the closest match. The organism identified by the NA1 ProSeq was A/Puerto Rico/8/34 because one of the SubSeqs had a much higher score. This ProSeq supported the same strain identification made in the matrix ProSeq. The organism identified was A/Puerto Rico/8/34 because two ProSeqs detected only that organism. The correct target pathogen was detected with filtering while without filtering the target pathogen level of identification would have been Influenza A (H1N1 subtype) because two organisms were detected, A/Puerto Rico/8/34 and A/Weiss/43. Splitting the HybSeq into SubSeqs to remove biasing may reduce the level of identification as occurred for the three of the five SubSeqs in this case. The previous example for Vaccinia was another instance where the wrong species (Camel Pox or Callithrix jacchus) identification would occur if the filtering was not used. Clinical samples in Table 2 showed that HybSeqs split into multiple SubSeqs were capable of very specific identification. To illustrate the other filtering performed, when multiplex strategies rather than generic were used for amplification Figure 5 also contains the raw and mask filtered results of this region for the A/Puerto Rico/8/34 sample. It was necessary to perform additional filtering to remove potential biasing caused by the specific primers as described in the methods. In the case shown in Figure 5, a sequence of 18 bases present in the raw result are made N after filtering since they are in a location that interacts with the primers. If these base calls were included in the subsequences constructed, even though the HybSeq would still be split into the same number of SubSeq, the query for the ProSeq would favor an incorrect strain. The algorithm we have developed successfully provided pathogen identification to the maximum level of detail possible (species or strain) depending on the quality of each ProSeq. This identification capability requires minimal input on the identity of the pathogens, making non-expert use feasible. The crucial feature incorporated that allowed complete automation was the use of taxonomic databases, which classify organisms into ordered groups and provide relationships between organism entries, allowing removal of redundancies, comparison of different related prototype sequences and simplification of data presentation. This allows databases, i.e. NCBI, that are redundant and subject to minimal curation but which constantly receive updated and new sequence information to be used with great success. Although we have demonstrated this using only the NCBI databases, other databases or custom made ones could have easily been used, which might improve performance. The algorithm is capable of providing accurate identifications at all analysis levels for pathogens that are less variable or are represented by highly conserved ProSeqs. For more variable or rapidly mutating pathogens, e.g. Influenza A virus, ProSeq identification Task(I) and ProSeq grouping Task(II) still provided accurate detailed identifications, but the pathogen determination Task(III) was unable to report fine scale discrimination. The comparison of the conventionally sequenced Influenza virus gene sequences illustrated that the algorithm is capable of automatically adjusting for updates in databases. The algorithm demonstrated its capability to properly distinguish hybridizations on a ProSeq caused by the specified pathogen from those caused by genetically close (near neighbor) strains and did not make incorrect identifications, eliminating one potential cause of false positives. Filtering the raw hybridization results served to reduce the computation time, accounted for potential primer interference and more importantly reduced potential biasing. This simple integrated algorithm provided sufficient and accurate identification, so that immediate use of the RPM v.1 or similar resequencing arrays and assay is possible. Although not discussed in this paper, the algorithm has successfully detected the presence of simulated multi-infections (Lin et al., submitted for publication). The algorithm as currently developed will detect mixtures when the organisms have sufficient variation; however, detection of a mixture of an organism and its mutation strain in a sample is uncertain in its present phase. In principle it may be possible to detect such mixtures as the resequencing microarray can detect and sequence diploid organisms. Besides demonstrating the success of the CIBSI 2.0 program, the work involved in developing the algorithm allowed insight into the importance of proper ProSeq selection. The RPM v.1 was the first resequencing array designed specifically for multiple pathogen detection using database similarity searching and served as a prototype for this application. We have demonstrated that a single ProSeq with as few as 100 bp, when designed correctly, can be sufficient to unambiguously identify an organism. However, it is clearly indicated that several longer ProSeqs provide better confirmation and more detailed information of a pathogen. Although the algorithm provides accuracy equivalent to manual analysis for determinations of individual ProSeqs, the current algorithm is only partially successful in integrating information from multiple ProSeqs. The emphasis of the design to this point has been on capabilities that are generally applicable to any pathogen. We are incorporating these insights in our newer more comprehensive resequencing array designs. Improving on level of detail reported in pathogen determination Task(III) will require more information about an individual pathogen and may have to be developed for each specific pathogen or class of pathogens. This information is also required for the algorithm to identify which differences between a sample and database entries represent significant mutations. Future work will involve improving the use of the current taxonomic database or potentially developing a new relational database that is specific to our needs and then incorporating more specific information of target pathogens. The hierarchal design of the data analysis makes it easy to incorporate analysis that build upon the analysis already performed. We have met with some success in the current version but want to have increased automated discrimination. We have a well-defined path to completing this aim. The use of properly designed resequencing microarrays and this automated detection algorithm provides a way forward to developing assays that can test for multiple organisms simultaneously while providing fine strain level discrimination giving access to information about detailed strain recognition, antibiotic resistance markers and pathogenicity. This is a capability that other approaches cannot currently provide. In addition, since the design of the original 30 kb RPM microarray, the possible sequence content of the current array has increased 10-fold to 300 kb and further increases in array density are still attainable. This, coupled with our identification algorithms, will allow the analysis of partial sequence information from even more organisms for applications such as differential diagnostics for illnesses with multiple potential causes (i.e. febrile respiratory illness), tracking of emergent pathogens, distinction of biological threats from harmless near genetic neighbors in surveillance applications and for tracking the impact of co-infections or super infections. The concept of categorizing and reporting different degrees of identification depending on the quality of samples and set of target sequences is not limited to resequencing microarrays but is more generally applicable to any platform that is capable of returning sequence level calls that can be used to query a reference DNA database. As the trend for assays that test for multiple pathogens increases, automated analysis tools, such as this one, become more crucial for rapid identification in simple formats useful to the non-expert on a day to day basis. The remaining hurdle to using resequencing microarrays as a routine assay method is now clearly the sample processing methods. Further automating these steps is an important area of future research and development. The program can be obtained free of charge for research purposes by contacting the authors. Supplementary Data are available at NAR Online. \"}"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# title and content of doc no. 0\n",
    "docs_list[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Python libraries have been imported.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[nltk_data] Downloading package stopwords to\n",
      "[nltk_data]     /Users/robert.trypuz/nltk_data...\n",
      "[nltk_data]   Package stopwords is already up-to-date!\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "import nltk\n",
    "#nltk.download()\n",
    "from nltk.stem import WordNetLemmatizer\n",
    "nltk.download('stopwords')\n",
    "from nltk.corpus import stopwords\n",
    "from nltk.corpus import wordnet\n",
    "print(\"Python libraries have been imported.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "class TextPreprocessor:\n",
    "    def __init__(self, docs_list):\n",
    "        self.docs_list = docs_list\n",
    "    \n",
    "    # doc title plays a role of id; in the future doc id should be uuid\n",
    "    def addField(self, doc_title, field_key, field_value):\n",
    "        for doc in self.docs_list:\n",
    "            if doc['title'] == doc_title:\n",
    "                doc[field_key] = field_value\n",
    "                break\n",
    "        \n",
    "    def wordSentTokenize(self, text):\n",
    "        return nltk.sent_tokenize(text)\n",
    "            \n",
    "    def cleanSentences(self, sentences_list):\n",
    "        sentences = []\n",
    "        for sentence in sentences_list:\n",
    "#             sentence = re.sub(r\"that's\",\"that is\",sentence)\n",
    "#             sentence = re.sub(r\"there's\",\"there is\",sentence)\n",
    "#             sentence = re.sub(r\"what's\",\"what is\",sentence)\n",
    "#             sentence = re.sub(r\"where's\",\"where is\",sentence)\n",
    "#             sentence = re.sub(r\"it's\",\"it is\",sentence)\n",
    "#             sentence = re.sub(r\"who's\",\"who is\",sentence)\n",
    "#             sentence = re.sub(r\"i'm\",\"i am\",sentence)\n",
    "#             sentence = re.sub(r\"she's\",\"she is\",sentence)\n",
    "#             sentence = re.sub(r\"he's\",\"he is\",sentence)\n",
    "#             sentence = re.sub(r\"they're\",\"they are\",sentence)\n",
    "#             sentence = re.sub(r\"who're\",\"who are\",sentence)\n",
    "#             sentence = re.sub(r\"ain't\",\"am not\",sentence)\n",
    "#             sentence = re.sub(r\"wouldn't\",\"would not\",sentence)\n",
    "#             sentence = re.sub(r\"shouldn't\",\"should not\",sentence)\n",
    "#             sentence = re.sub(r\"can't\",\"can not\",sentence)\n",
    "#             sentence = re.sub(r\"couldn't\",\"could not\",sentence)\n",
    "#             sentence = re.sub(r\"won't\",\"will not\",sentence)\n",
    "            sentence = re.sub(r\"\\W\",\" \",sentence) # non-word character\n",
    "            sentence = re.sub(r\"\\d\",\" \",sentence) # digit character\n",
    "            sentence = re.sub(r\"nan\",\" \",sentence)\n",
    "            sentence = re.sub(r\"NaN\",\" \",sentence)  \n",
    "#             sentence = re.sub(r\"is\",\" \",sentence)\n",
    "#             sentence = re.sub(r\"IS\",\" \",sentence)\n",
    "#             sentence = re.sub(r\"for\",\" \",sentence)\n",
    "#             sentence = re.sub(r\"the\",\" \",sentence)\n",
    "#             sentence = re.sub(r\"unnamed\",\" \",sentence)\n",
    "#             sentence = re.sub(r\"yes\",\" \",sentence)\n",
    "#             sentence = re.sub(r\"Yes\",\" \",sentence)\n",
    "#             sentence = re.sub(r\"no\",\" \",sentence)\n",
    "#             sentence = re.sub(r\"No\",\" \",sentence)\n",
    "#             sentence = re.sub(r\"com\",\" \",sentence)\n",
    "#             sentence = re.sub(r\"Com\",\" \",sentence)\n",
    "#             sentence = re.sub(r\"COM\",\" \",sentence)\n",
    "            sentence = re.sub(r\"_\",\" \",sentence)\n",
    "            sentence = re.sub(r\"\\s+[a-z]\\s+\",\" \",sentence) # clean a whitespace\n",
    "            sentence = re.sub(r\"\\s+[a-z]$\",\" \",sentence)\n",
    "            sentence = re.sub(r\"^[a-z]\\s+\",\" \",sentence)\n",
    "            sentence = re.sub(r\"\\s+\",\" \",sentence)           \n",
    "            sentences.append(sentence.strip())\n",
    "        return sentences\n",
    "    \n",
    "    def wordTokenizeCorpus(self, sentences_list):\n",
    "        return  [nltk.word_tokenize(sentence) for sentence in sentences_list]\n",
    "\n",
    "    def getWordnetPos(self, word):\n",
    "        \"\"\"Map POS tag to first character lemmatize() accepts\"\"\"\n",
    "        tag = nltk.pos_tag([word])[0][1][0].upper()\n",
    "        tag_dict = {\"J\": wordnet.ADJ,\n",
    "                    \"N\": wordnet.NOUN,\n",
    "                    \"V\": wordnet.VERB,\n",
    "                    \"R\": wordnet.ADV}    \n",
    "        return tag_dict.get(tag, wordnet.NOUN)\n",
    "\n",
    "    def lemamatizePosSensitive(self, sentences_list):\n",
    "        lemmatizer = WordNetLemmatizer()\n",
    "        sentences = []\n",
    "        for sentence in sentences_list:\n",
    "            sentences.append([lemmatizer.lemmatize(word, self.getWordnetPos(word)) \n",
    "                              for word in sentence])\n",
    "        return sentences\n",
    "    \n",
    "    def lemmatizeWords(self, sentences_list):\n",
    "        lemmatizer = WordNetLemmatizer()\n",
    "        sentences = []\n",
    "        for sentence in sentences_list:\n",
    "            words = [lemmatizer.lemmatize(word) for word in sentence]\n",
    "            sentences.append(words)\n",
    "        return sentences\n",
    "            \n",
    "    \n",
    "    def removeStopwords(self, sentences_list):\n",
    "        sentences = []\n",
    "        for sentence in sentences_list:\n",
    "            words = [word for word in sentence if word.lower() not in stopwords.words('english')]\n",
    "            sentences.append(words)\n",
    "        return sentences\n",
    "\n",
    "    def posTagging(self, sentences_list):\n",
    "        sentences = []\n",
    "        for sentence in sentences_list:\n",
    "            pos = nltk.pos_tag(sentence)\n",
    "            sentences.append(pos)\n",
    "        return sentences\n",
    "    \n",
    "    def regognizeNamedEntities(self, sentences_list_with_pos_tags):\n",
    "        sentences = []\n",
    "        for sentence in sentences_list_with_pos_tags:\n",
    "            ne = nltk.ne_chunk(sentence)\n",
    "            sentences.append(ne)\n",
    "        return sentences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Preprocessing has been initiated.\n",
      "Preprocessing accomplished! Be patient, it may take a while.\n"
     ]
    }
   ],
   "source": [
    "text_preprocessor = TextPreprocessor(docs_list)\n",
    "print('Preprocessing has been initiated.')\n",
    "for doc in docs_list:\n",
    "    text_preprocessor.addField(doc['title'],\n",
    "                              'content_sent_tokenized',\n",
    "                              text_preprocessor.wordSentTokenize(doc['content']))  \n",
    "\n",
    "    text_preprocessor.addField(doc['title'],\n",
    "                              'clean_content_sent_tokenized',\n",
    "                              text_preprocessor.cleanSentences(doc['content_sent_tokenized']))  \n",
    "\n",
    "    text_preprocessor.addField(doc['title'],\n",
    "                              'content_word_tokenized',\n",
    "                              text_preprocessor.wordTokenizeCorpus(doc['clean_content_sent_tokenized']))  \n",
    "\n",
    "    text_preprocessor.addField(doc['title'],\n",
    "                              'lemmatized',\n",
    "                              text_preprocessor.lemamatizePosSensitive(doc['content_word_tokenized']))  \n",
    "\n",
    "    text_preprocessor.addField(doc['title'],\n",
    "                              'no_stopwords',\n",
    "                              text_preprocessor.removeStopwords(doc['lemmatized']))  \n",
    "\n",
    "    text_preprocessor.addField(doc['title'],\n",
    "                              'pos_tagged',\n",
    "                              text_preprocessor.posTagging(doc['no_stopwords']))  \n",
    "    \n",
    "    text_preprocessor.addField(doc['title'],\n",
    "                             'ne_tagged',\n",
    "                             text_preprocessor.regognizeNamedEntities(doc['pos_tagged']))\n",
    "\n",
    "print('Preprocessing accomplished! Be patient, it may take a while.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PMC1636417.xml.json\n"
     ]
    }
   ],
   "source": [
    "print(docs_list[0]['title'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "For both surveillance and diagnostic applications, fine-scale pathogen identification and near-neighbor discrimination is important; therefore, an assay that monitors at this very specific level is desirable for many types of samples such as clinical and environmental (1–3). To successfully use any method based on DNA or RNA detection, these assays must be coupled with large databases of nucleic acid sequence information for assay design to ensure that the desired information is provided and for the interpretation of raw data. Several well-established techniques use PCR to amplify individual target pieces of sequenced genomes to provide detection of organisms (4). These methods can roughly be divided into approaches that target individual short sequence lengths or probes (<40 bp) and methods that examine longer probes. The advantage of using short probes is that when the uniqueness of the probe has been assured and unique primers are also selected, this method gives good specificity. This approach is capable of providing fine-scale identification of several genetically close organisms by selecting a sufficient number of probes. However, this can rapidly lead to a very large number of total probes being required to detect all organisms of interest. In addition these selected probes, which in the initial selection process were determined to be unique, are often later found to be less specific as more organisms are sequenced or are less specific under conditions that differ from the original conditions. This is particularly a problem for organisms belonging to a family with a high mutation rate and also for pathogens that have relatively few neighboring pathogens sequenced. In addition, PCR approaches focused on short unique probes are not capable of detecting the presence of new significant mutations nor can they easily resolve base sequence details. Approaches that use longer individual probes avoid many of these issues at the cost of being less specific. This issue means most of these approaches are not suitable for providing the information desired, providing impetus to this work. High-density resequencing microarrays produce variable length segments, 102–105 bp, of direct sequence information. This target sequence falls in the longer target regime of PCR approaches but rather than being hybridized to a longer less-specific probe on the microarray, many shorter specific probes are placed on the microarray to allow more detailed determinations from the entire PCR amplicon. This also means that the specificity of the primers used can be relaxed. They have been successfully used to detect single nucleotide polymorphism (SNP) and genetic variants from viral, bacterial and eukaryotic genomes (5–12). Their use for SNP detection has clearly established their ability to provide reliable quality sequence information. In most cases, the microarrays were designed to study a limited number of genetically similar target pathogens and for many cases, the detection methods relied only on recognizing hybridization patterns for identification (6,9,10,13,14). Taking advantage of the sequential base resolution capability of resequencing microarrays that is required for SNP detection, resequencing has recently been successfully adapted recently using a different approach for organism identification of multiple bacterial and viral pathogens while allowing for fine detailed discrimination of closely related organisms and tracking mutations within the targeted pathogen (15–16). The new methodology differed from earlier work by using the resolved bases as the query of a similarity search of DNA databases to identify the most likely species and variants that match the base calls from the hybridization observed. The system was capable of testing for 26 pathogens simultaneously and could detect the presence of multiple pathogens. A software program, resequencing pathogen identifier (REPI), was used to simplify data analysis by performing similarity searches of a genetic database using basic local alignment search tool (BLAST) (17). The REPI program used BLAST default settings and would only return sequences that might represent the hybridization if the expect value, a quantity calculated by the BLAST program that indicates the likelihood that the sequence match found would have occurred by random chance in the database, was <10−9. This screened out all cases that had insufficient signal; however, the final determination of what pathogen(s) was detected and to what degree discrimination was possible required manual examination of the returned results. This method successfully allowed fine discrimination of various adenoviruses and strain identifications of Flu A and B samples in agreement with conventional sampling results (15,16). Two important advantages of this approach were that the information was always recovered at the most detailed level possible and that it was capable of still recognizing organisms with recent mutations. This approach also maintained specificity well, as it was not dependent on the uniqueness of a few individual short probes. Although this analysis method has utility, there are several shortcomings: it is time consuming, not optimized to maximize sensitivity, has complicated results, is suitable only for an expert, and contains redundant or duplicate information. The process was time consuming because only the initial screening was handled automatically while the remaining steps required manual interpretation before the detection analysis was complete. Because a simple criterion (expect value cutoff of 10−9) and non-optimized BLAST parameters were used to consider a pathogen detected, the REPI algorithm provided a list of candidate organisms but did not make a final simple conclusion or relate the results of one prototype sequence to another. Instead a manual process was used to make the final determination, but because the REPI program provided all similar results and the use of public nucleic acid databases containing redundant entries, a large amount of data was presented to a user that was not useful. In addition, with a manual process it was not possible to establish that the algorithm developed was generally applicable for any organism where nucleic acid base resolved sequence information has been provided. In this paper, we describe a new software expert system, Computer-Implemented Biological Sequence Identifier system 2.0 (CIBSI 2.0), that successfully uses resolved base sequence information from custom designed Affymetrix resequencing microarrays to provide a simple list of organisms that are detected. This algorithm addresses the most important shortcoming of previous methods by incorporating new features to completely automate pathogen identification. We have demonstrated the effectiveness of this algorithm for identification via several examples. The single program is capable of making correct decisions for all 26 pathogens contained on the Respiratory Pathogen Microarray v.1 (RPM v.1), whether detected alone or in combinations, with improved sensitivity. Although the program is currently applied to resequencing microarrays, the methodologies developed remain generally applicable. Only the first portion of the algorithm handles issues specific to microarrays while the remainder deals with sequences that are suitable for use as a query by the BLAST algorithm. In developing the general identification algorithm, we have identified and resolved issues specific to resequencing microarrays that complicate their use. Because the entire decision process for what is detected has been automated, it is straightforward to test whether the rules used to make identifications are rigorous and applicable to any pathogen. With this efficient program, resequencing based assays can provide a competitive method to test simultaneously for many possible pathogens, providing output that can be interpreted by a non-expert. The details of the RPM v.1 design and the experimental methods have been discussed in previous work (15,16,18) (Lin et al., submitted for publication). Briefly, the RPM v.1 chip design includes 57 tiled regions allowing resequencing of 29.7 kb of sequences from 27 respiratory pathogens and biothreat agents. These were selected based upon clinical relevance for the population of immediate interest (United States military recruit in training) (19–21). Partial sequences from the genes containing diagnostic regions were tiled for the detection of these pathogens. The experimental microarray data used in the present analysis were obtained using a variety of purified nucleic acid templates and clinical samples culture (throat swabs and nasal washes) using random and multiplexed RT–PCR amplification schemes (for more detail description of amplification methods see Supplementary Data). Resequencing microarrays provide base call resolution by comparing the intensities between a set of four 25mer probes that differ from each other at the same position (13th base). An amplicon or target sequence is represented by numerous overlapping probe sets. GCOS™ software v1.3 (Affymetrix Inc., Santa Clara, CA) was used to align and scan hybridized microarrays to determine the intensity of each probe in every probe set. Base calls were made based on the intensity data of each probe set using GDAS v3.0.2.8 software (Affymetrix Inc.) which used an implementation of the ABACUS algorithm (5). The sequences were represented in FASTA format for later analysis steps. In this paper, target pathogens are the organisms the assay was specifically designed to detect. The sets of probes that represent reference sequence selected from target pathogen genomes are referred to as a Prototype Sequence or ‘ProSeq’ for brevity. The set of resolved bases that result from hybridization of genomic material to a ProSeq is referred to as the hybridized sequence or ‘HybSeq’. The HybSeq is split into possible subsequences or ‘SubSeqs’. The CIBSI 2.0 program implemented in Perl described in this study handled a hierarchy of three tasks (Figure 1): (I) ProSeq identification; (II) ProSeq grouping; and (III) pathogen determination. The most developed and important portion of the algorithm deals with ProSeq identification Task(I) and is handled in three important subtasks: initial filtering of individual HybSeqs into SubSeqs suitable for sequence similarity comparisons (Figure 2), database querying of individual SubSeqs (Figure 3) and taxonomic comparison of BLAST returns for each SubSeq (Figure 4). The NCBI BLAST and taxonomy databases were used for the queries and images were obtained on February 7, 2006. For the ProSeq grouping Task(II), ProSeqs were compared to determine if they supported the same identified organism. In the pathogen determination Task(III), detected organisms were compared to the list of target pathogens the assay was designed for in order to determine if any were positively detected or were possibly related close genetic near neighbors. The level of discrimination that a particular sample supported was automatically determined. An initial filtering algorithm, REPI, was developed previously (16) and the general concepts with revisions were incorporated into the current (automated detection) algorithm used in the CIBSI 2.0 program. Filtering and subsequence selection were used to remove potential biasing caused by reference sequence choice and by other sources (i.e. primers). When PCR amplification was used, microarrays were hybridized in the presence of only primers to determine locations where they resulted in hybridization. Any portions of the ProSeqs that hybridized with the primers were masked as N calls so that the HybSeq did not contain biased information. Normally the primers are designed to be outside the ProSeq region to minimize the interference caused by primers, and so minimize the bases to be masked. There is still the chance that some bases require masking because with the large number of primers used in the multiplex, short stretches of a ProSeq not corresponding to primer locations may still hybridize with the primers. Such regions could be removed from the reference sequences and so not appear on the microarray. However, determining such locations are a difficult and time-consuming task that for most cases is not worth the effort. The first subtask of ProSeq identification Task(I) is noted in Figure 1 and shown schematically in detail in Figure 2. This subtask uses a procedure to examine a HybSeq to find the longest possible subsequence of base calls (SubSeq) that can be submitted as a query to BLAST. It produces a group of SubSeq that contain all portions of a HybSeq that have a chance of producing a limited list of returns from a BLAST query. When a HybSeq has two regions separated by a long stretch of continuous N calls, the relational positioning of the two regions cannot be trusted and so must be sent as separate queries. In addition for shorter subsequences, the number of base calls that must be made is dependent on the length. It was also recognized that for very long sequences a longer WORD size in BLAST may be used. A detailed description of the criteria and process used for each step is contained in Supplementary Data. Upon completion, the algorithm returned to the Task(I) loop and performed the BLAST subtask. The database query subtask performed a batch similarity search of a database using SubSeq as the queries. The BLAST program used was the NCBI Blastall –p blastn version 2.12 with a defined set of parameters. The masking of low complex regions was performed for the seeding phase to speed up the query; however, low complexity repeats were included in the actual scoring. The entire nucleotide database from NCBI acquired on February 7, 2006 was used as the reference database. (Note that earlier images of the database were used during development but all experiments were rerun with the algorithm as described with the image of the database obtained on this date.) The default gap penalty and nucleotide match score were used. The nucleotide mismatch penalty, –q, parameter was set to −1 rather than the default. The results of any BLAST query with an expect value <0.0001 were returned in tabular format from the blastall program. The information about each return (bit score, expect value, mismatches, length of match) was placed in the Return{hash key}{info} hash using the SubSeq identity as the hash key for further analysis. The next subtask of ProSeq identification Task(I) carried out was the determination of SubSeq() states and is shown in Figure 3. The BLAST algorithm gives a ranking score which can be reported as accounting for the size of the database (expect value) or not (bit score). The full taxonomic classifications of every return for a SubSeq were retrieved from the NCBI taxonomy database obtained on February 7, 2006. Using the scores and taxonomy relationships it was possible to find a reduced number of returns that had the best match with the HybSeq. These results were summarized by identifying the taxonomic class to which all the returns belonged to, ‘identified organism’, and a parameter that indicated how they are related to each other, ‘organism uniqueness’. A detailed description of the steps is contained in Supplementary Data. After each SubSeq was examined, the algorithm moved to the next subtask, which was to determine the identified organism of the ProSeq from the SubSeq (Figure 4). The subsequences from the same ProSeq were only allowed to support a single ‘identified organism’ determination. The procedure shown in Figure 4 demonstrates the decision method used to arrive at this determination (detailed description in Supplementary Data). After the subtask covered in Figure 4 was completed, the ProSeq identification Task(I) loop continued until all ProSeqs were examined. A list of ProSeqs that had detected organisms was built up in the Result1 array. After the ProSeq identification Task(I) was completed, ProSeq grouping Task(II) (Figure 1) was used to examine the identified organism values listed in Result1 and grouped them together if they identified the same taxonomic class. Each entry in Result1 was examined and a new entry was created in Result2 if the identified organism did not appear in this list. The entries of Result2 represented the distinct individual organisms identified, but might still contain redundant information. When the ProSeqs were designed to detect the same organism and they all hybridized well, this grouping led to a reduction in redundant information being reported. But, when one ProSeq did not hybridize as well for a variety of possible reasons, multiple entries would appear in Result2 that actually represent hybridization from the same pathogen. This is because there is an alternative cause for the ProSeq hybridizing in this manner. This hybridization could be caused by two different but closely related organisms both being present in a sample and hybridizing to the microarray. Because we have not yet developed methods to distinguish these cases, no further reduction of the list of organisms is made for in ProSeq grouping Task(II) in cases where the level of identification varied on different ProSeq targeted for the same organism. Although it was difficult to relate results from separate ProSeqs to each other, it was desired to have a simple final detection decision be made in pathogen determination Task(III). The first task was specifically implemented so that information about what a ProSeq was intended to detect was not considered and the second task only minimal consideration of this was taken into account. This allowed these initial tasks to be capable of recognizing not only just positive and negative identifications of target pathogens but also cases that were indeterminate. In the final task, the algorithm considered whether the identified organisms belonged to the list of organisms the ProSeqs were designed to detect. The task would group organisms from ProSeq grouping Task(II) together that belonged to or were child classes of the taxonomic class of a target organism. The taxonomic class reported was the common taxonomic group of all the organisms. When all the ProSeqs for a pathogen hybridized well, a fine level discrimination was reported. But if one or more ProSeqs hybridized less well, the reported positive target pathogen was only identified at the level of the less detailed level. This is conservative because methods have not yet been developed to clearly discriminate mixtures of very closely related organisms causing different ProSeqs to hybridize from variable hybridization of a single organism on several ProSeqs. The results of all three tasks were reported and a more experienced user can view ProSeq grouping Task(II) results to clarify some cases. Note that organisms identified in ProSeq grouping Task(II) that only belonged to target pathogens were reported as positives. Clear negative ProSeqs were not mentioned in the output. ProSeqs that were indeterminate or that detected close genetic organisms were never reported as positives. These organisms were instead reported as being detected. Purified Chlamydia pneumoniae nucleic acid samples with 10–1000 genome copies (via method in Lin et al., submitted for publication) were chosen to illustrate how pathogen detection and identification were done when multiple ProSeqs were targeted for the same pathogen. RPM v.1 has three highly conserved ProSeqs selected from the genes encoding for the major outer membrane proteins VD2 and VD4, and the DNA-directed RNA polymerase (rpoB) gene. The HybSeqs from the different samples differed only in the number of unique base calls as shown in Table 1. The percentage of the ProSeq called varied from 80 to 100% except for one case at a concentration of 10 that had only 11% of the rpoB ProSeq producing unique base calls. Because the samples at this concentration are not reproducibly generating the same percentage of base calls, this is probably the detection limit of this ProSeq of the assay. Table 1 listed the determinations made for the SubSeq and at the end of each task for the various samples. The ProSeq from the different cases produced the same number of SubSeqs. These SubSeqs from different samples reported different bit scores for the same top ranked returns from BLAST. In fact VD2 and VD4 produced exactly the same results. The NCBI taxonomy database classified the returns into four distinct groups, which represented the C.pneumoniae taxonomic group and three child strain groups. AE001652, AE002167, AE017159 and BA000008 appeared in the returns of all the ProSeqs for each sample, since they represented database entries of completely sequenced genomes. One rpoB SubSeq produced for its organism uniqueness, SeqUniqu. All other SubSeqs were TaxAmbig as multiple returns from different taxonomic classes were returned. Since the VD2 and VD4 ProSeq each have a single SubSeq, Task(I) assigned the ProSeq the state of the SubSeq. For the rpoB ProSeq, the bit score of one SubSeq was large enough that the algorithm assigned that SubSeq's identification to the ProSeq. Task II of the algorithm grouped all three ProSeqs together since they all had the same identified organism and TaxAmbig was assigned. The result of Task(III) was positive for target pathogen C.pneumoniae and this decision was straightforward as all the ProSeqs agreed with each other and belonged to the same target pathogen taxonomic class. Although the rpoB ProSeq was SeqUniqu, this was not the final conclusion for Task(II) as the ProSeq that was SeqUniqu was not the child taxonomic group and other ProSeq were TaxAmbig. The three recognized strains scored the same, which indicated that the sequence selected for the ProSeqs was very conserved and would not allow discrimination between the strains. Influenza and Human Adenovirus (HAdV) were the only pathogens that had ProSeq selected that would permit detailed strain level discrimination as discussed in previous work (15,16). This previous work using manual analysis found that the microarray results were in excellent agreement with the conventional sequencing results for clinical samples. A few of the results of running the CIBSI 2.0 program using the updated NCBI database on the raw microarray results are presented in Table 2 (the results for all samples used in the previous work are presented in Supplementary Table A). The identified organisms were not identical to the original findings due to the difference in database used and because all ProSeqs were considered rather than only the Flu A and B hemagglutinin. In fact, the conventional sequencing results that were submitted to NCBI from that work were found for every sample to be among the returns with the best score for the hemagglutinin ProSeq (Supplementary Table B). It should be noted that the previous work based its analysis upon only the results of the hemagglutinin ProSeq. For 8 of 13 Influenza A and 3 of 12 Influenza B cases, the results of ProSeq identification Task(I) and ProSeq grouping Task(II) found that the conventional sequencing was the single best return for the hemagglutinin ProSeq. Owing to the large number of isolate sequences in the database for the hemagglutinin gene it was not surprising that in some cases a single unique entry was not found. In each of the remaining five Influenza A samples, the other sequences returned differed by <0.2% from the conventional sequence. The fewer samples with unique isolate identifications for Influenza B were due to an older reference sequence used for the ProSeq, which allowed less hybridization to occur (18). This also meant that when multiple sequences were returned for a sample they represented greater genetic variation, up to 2%. As a result of the current method of making pathogen determination Task(III) level identification, the final organism reported was less specific (H3N2 or Flu B) for every sample than what was reported as possible in ProSeq grouping Task(II). For HAdV samples, the algorithm also reproduced the finer scale discriminations that had been made previously by manual methods (data not shown). The next example of detection for the Mycoplasma pneumoniae pathogen demonstrated a case where there was only a single ProSeq for the target pathogen. A total of 48 test samples were performed using multiplex PCR (via method in Lin et al., submitted for publication) where for 46 of the samples M.pneumoniae organism was spiked into nasal wash with several other pathogens from 100 to 100 000 colony forming units per ml, the remaining 2 samples were purified with nucleic acid from culture stock at a concentration of 1000 genome copies per reaction volume. This ProSeq was also not optimal for fine discrimination because it was selected from a highly conserved region (345 bp) of the cytadhesin P1 gene. In every case taxonomic database entries for M.pneumoniae or its one recognized distinct strain tied for MaxScore (Supplementary Table 3). To better understand these returns, the database sequences were examined and subdivided into three groups of sequences, A, B and C, based on how well they matched the reference sequence used to make the ProSeq. The placement of the database entries into the three groups was determined from a CLUSTAL alignment of the sequences of this gene. This alignment confirmed that the database entries differed significantly more from each other in regions not represented by the ProSeq and contained sufficient variability that would have allowed finer discrimination. Members of Group A exactly matched the ProSeq and could not be distinguished between on the microarray. Similarly, members of group B matched the ProSeq except at the 199th position where the base called was C rather than T. Group C sequences contained a few database entries that were more variable and might be distinguished from other entries within the ProSeq. For the 48 experimental tests of M.pneumoniae, as much as 80% of the ProSeq hybridized for 19 samples, yet only 5 of these samples had an unambiguous base call at the 199th position. When it was unambiguous, it always matched group B sequences. In the cases where an N base call was made at the 199th location, both groups A and B sequences were returned with the same score. Regardless of this, the target pathogen positively identified was M.pneumoniae for every sample tested. These examples showed how decisions were made independent of whether single or multiple ProSeqs were dedicated to a target pathogen. They also illustrated that the level of discrimination possible was strongly determined by the quality of the selected ProSeq. It is possible that for some pathogens fine level discrimination is not required and the currently tested selections on RPM v.1 would provide satisfactory information. The CIBSI 2.0 algorithm demonstrated its capability to automatically report the maximum level of discrimination that could be supported by the HybSeq information. To demonstrate how the algorithm handled closely related genetic species, a sample of a non-targeted pathogen was considered using multiplex PCR (via method in Lin et al., submitted for publication). For Variola major virus, one of the biothreat pathogens on the RPMv.1, the validation runs demonstrated that Variola major virus purified DNA templates of plasmids were always positively identified when detected (Table 3). Table 4 shows the results when purified Vaccinia genomic DNA was spiked into nasal washes and processed at various concentrations using multiplex PCR. The array has two ProSeqs from hemagglutinin (VMVHA, ∼500 bp) and cytokine response modifier B (VMVcrmB, ∼300 bp) genes for Variola major virus detection. The percentage of the ProSeq that hybridizes is sufficient that if hybridization patterns were only considered one might assume that this tile is identifying the presence of its target. This would indicate that reference sequence selected was not the best choice. However, when our algorithm was applied none of the samples is in fact identified as Variola major or minor virus. Vaccinia was always one of the Orthopoxvirus species listed with the highest scores for VMVcrmB ProSeq, but in only seven cases was it uniquely identified as the probable species detected. In only one sample at the lowest concentration and fraction of VMVcrmB hybridizing, did this ProSeq even identify Variola major and minor virus as one among the Orthopoxvirus species that could be the cause of the hybridization. The lower limit of detection for the amplification method used was between this concentration and the one above it for Variola major itself. The VMVHA ProSeq exhibited much lower sensitivity and made identifications of Orthopoxvirus species in only two experiments and Variola major virus was listed as one of the tied best scoring returns. In both cases, VMVcrmB ProSeq specifically identified Vaccinia virus as the best match. The percentage of the hybridized ProSeq correlated with concentration of the sample. This example demonstrated the importance of the filtering portion of the algorithm by considering the HybSeqs of the ProSeqs for the H1N1 neuraminidase (NA1) and matrix genes from human Influenza A/Puerto Rico/8/34 (H1N1) strain. Filtering was necessary because sending the HybSeq of a ProSeq to BLAST in a single query can bias the scores against strains that have insertions or deletions relative to the ProSeq, especially when using BLAST parameters that maximized the use of base calls. The sliding window test was the portion of the algorithm that controlled filtering. If filtering were turned off, the entire HybSeq would be used in a single subsequence for two influenza ProSeqs that showed significant hybridization. A/Weiss/43 (H1N1) strain was identified as the most likely strain from the HybSeq of the NA1 ProSeq while the HybSeq of the matrix ProSeq correctly identified A/Puerto Rico/8/34. To better understand the source of biasing, CLUSTAL alignment of the NA1 gene of the two strains and the reference sequence used to make the ProSeq are shown in Figure 5. The two strains showed 95% identity (67 mismatches in 1362 aligned bases); however, there was a stretch of 45 bases inserted in both A/Weiss/43 and the NA1 ProSeq compared to A/Puerto Rico/8/34. With the default filtering on, the NA1 ProSeq was split into five SubSeqs as the algorithm encountered large stretches of no calls. In ProSeq identification Task(I), the algorithm determined that three of the SubSeqs, which were shorter, had an identified organism of H1N1 as several isolates including A/Puerto Rico/8/34 tied with the best score, while the two other SubSeqs had an identified organism of only the A/Puerto Rico/8/34 strain as being the closest match. The organism identified by the NA1 ProSeq was A/Puerto Rico/8/34 because one of the SubSeqs had a much higher score. This ProSeq supported the same strain identification made in the matrix ProSeq. The organism identified was A/Puerto Rico/8/34 because two ProSeqs detected only that organism. The correct target pathogen was detected with filtering while without filtering the target pathogen level of identification would have been Influenza A (H1N1 subtype) because two organisms were detected, A/Puerto Rico/8/34 and A/Weiss/43. Splitting the HybSeq into SubSeqs to remove biasing may reduce the level of identification as occurred for the three of the five SubSeqs in this case. The previous example for Vaccinia was another instance where the wrong species (Camel Pox or Callithrix jacchus) identification would occur if the filtering was not used. Clinical samples in Table 2 showed that HybSeqs split into multiple SubSeqs were capable of very specific identification. To illustrate the other filtering performed, when multiplex strategies rather than generic were used for amplification Figure 5 also contains the raw and mask filtered results of this region for the A/Puerto Rico/8/34 sample. It was necessary to perform additional filtering to remove potential biasing caused by the specific primers as described in the methods. In the case shown in Figure 5, a sequence of 18 bases present in the raw result are made N after filtering since they are in a location that interacts with the primers. If these base calls were included in the subsequences constructed, even though the HybSeq would still be split into the same number of SubSeq, the query for the ProSeq would favor an incorrect strain. The algorithm we have developed successfully provided pathogen identification to the maximum level of detail possible (species or strain) depending on the quality of each ProSeq. This identification capability requires minimal input on the identity of the pathogens, making non-expert use feasible. The crucial feature incorporated that allowed complete automation was the use of taxonomic databases, which classify organisms into ordered groups and provide relationships between organism entries, allowing removal of redundancies, comparison of different related prototype sequences and simplification of data presentation. This allows databases, i.e. NCBI, that are redundant and subject to minimal curation but which constantly receive updated and new sequence information to be used with great success. Although we have demonstrated this using only the NCBI databases, other databases or custom made ones could have easily been used, which might improve performance. The algorithm is capable of providing accurate identifications at all analysis levels for pathogens that are less variable or are represented by highly conserved ProSeqs. For more variable or rapidly mutating pathogens, e.g. Influenza A virus, ProSeq identification Task(I) and ProSeq grouping Task(II) still provided accurate detailed identifications, but the pathogen determination Task(III) was unable to report fine scale discrimination. The comparison of the conventionally sequenced Influenza virus gene sequences illustrated that the algorithm is capable of automatically adjusting for updates in databases. The algorithm demonstrated its capability to properly distinguish hybridizations on a ProSeq caused by the specified pathogen from those caused by genetically close (near neighbor) strains and did not make incorrect identifications, eliminating one potential cause of false positives. Filtering the raw hybridization results served to reduce the computation time, accounted for potential primer interference and more importantly reduced potential biasing. This simple integrated algorithm provided sufficient and accurate identification, so that immediate use of the RPM v.1 or similar resequencing arrays and assay is possible. Although not discussed in this paper, the algorithm has successfully detected the presence of simulated multi-infections (Lin et al., submitted for publication). The algorithm as currently developed will detect mixtures when the organisms have sufficient variation; however, detection of a mixture of an organism and its mutation strain in a sample is uncertain in its present phase. In principle it may be possible to detect such mixtures as the resequencing microarray can detect and sequence diploid organisms. Besides demonstrating the success of the CIBSI 2.0 program, the work involved in developing the algorithm allowed insight into the importance of proper ProSeq selection. The RPM v.1 was the first resequencing array designed specifically for multiple pathogen detection using database similarity searching and served as a prototype for this application. We have demonstrated that a single ProSeq with as few as 100 bp, when designed correctly, can be sufficient to unambiguously identify an organism. However, it is clearly indicated that several longer ProSeqs provide better confirmation and more detailed information of a pathogen. Although the algorithm provides accuracy equivalent to manual analysis for determinations of individual ProSeqs, the current algorithm is only partially successful in integrating information from multiple ProSeqs. The emphasis of the design to this point has been on capabilities that are generally applicable to any pathogen. We are incorporating these insights in our newer more comprehensive resequencing array designs. Improving on level of detail reported in pathogen determination Task(III) will require more information about an individual pathogen and may have to be developed for each specific pathogen or class of pathogens. This information is also required for the algorithm to identify which differences between a sample and database entries represent significant mutations. Future work will involve improving the use of the current taxonomic database or potentially developing a new relational database that is specific to our needs and then incorporating more specific information of target pathogens. The hierarchal design of the data analysis makes it easy to incorporate analysis that build upon the analysis already performed. We have met with some success in the current version but want to have increased automated discrimination. We have a well-defined path to completing this aim. The use of properly designed resequencing microarrays and this automated detection algorithm provides a way forward to developing assays that can test for multiple organisms simultaneously while providing fine strain level discrimination giving access to information about detailed strain recognition, antibiotic resistance markers and pathogenicity. This is a capability that other approaches cannot currently provide. In addition, since the design of the original 30 kb RPM microarray, the possible sequence content of the current array has increased 10-fold to 300 kb and further increases in array density are still attainable. This, coupled with our identification algorithms, will allow the analysis of partial sequence information from even more organisms for applications such as differential diagnostics for illnesses with multiple potential causes (i.e. febrile respiratory illness), tracking of emergent pathogens, distinction of biological threats from harmless near genetic neighbors in surveillance applications and for tracking the impact of co-infections or super infections. The concept of categorizing and reporting different degrees of identification depending on the quality of samples and set of target sequences is not limited to resequencing microarrays but is more generally applicable to any platform that is capable of returning sequence level calls that can be used to query a reference DNA database. As the trend for assays that test for multiple pathogens increases, automated analysis tools, such as this one, become more crucial for rapid identification in simple formats useful to the non-expert on a day to day basis. The remaining hurdle to using resequencing microarrays as a routine assay method is now clearly the sample processing methods. Further automating these steps is an important area of future research and development. The program can be obtained free of charge for research purposes by contacting the authors. Supplementary Data are available at NAR Online. \n"
     ]
    }
   ],
   "source": [
    "print(docs_list[0]['content'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1. For both surveillance and diagnostic applications fine scale pathogen identification and near neighbor discrimination is important therefore an assay that monitors at this very specific level is desirable for many types of samples such as clinical and environmental\n",
      "2. To successfully use any method based on DNA or RNA detection these assays must be coupled with large databases of nucleic acid sequence information for assay design to ensure that the desired information is provided and for the interpretation of raw data\n",
      "3. Several well established techniques use PCR to amplify individual target pieces of sequenced genomes to provide detection of organisms\n",
      "4. These methods can roughly be divided into approaches that target individual short sequence lengths or probes bp and methods that examine longer probes\n",
      "5. The advantage of using short probes is that when the uniqueness of the probe has been assured and unique primers are also selected this method gives good specificity\n",
      "6. This approach is capable of providing fine scale identification of several genetically close organisms by selecting sufficient number of probes\n",
      "7. However this can rapidly lead to very large number of total probes being required to detect all organisms of interest\n",
      "8. In addition these selected probes which in the initial selection process were determined to be unique are often later found to be less specific as more organisms are sequenced or are less specific under conditions that differ from the original conditions\n",
      "9. This is particularly problem for organisms belonging to family with high mutation rate and also for pathogens that have relatively few neighboring pathogens sequenced\n",
      "10. In addition PCR approaches focused on short unique probes are not capable of detecting the presence of new significant mutations nor can they easily resolve base sequence details\n",
      "11. Approaches that use longer individual probes avoid many of these issues at the cost of being less specific\n",
      "12. This issue means most of these approaches are not suitable for providing the information desired providing impetus to this work\n",
      "13. High density resequencing microarrays produce variable length segments bp of direct sequence information\n",
      "14. This target sequence falls in the longer target regime of PCR approaches but rather than being hybridized to longer less specific probe on the microarray many shorter specific probes are placed on the microarray to allow more detailed determinations from the entire PCR amplicon\n",
      "15. This also means that the specificity of the primers used can be relaxed\n",
      "16. They have been successfully used to detect single nucleotide polymorphism SNP and genetic variants from viral bacterial and eukaryotic genomes\n",
      "17. Their use for SNP detection has clearly established their ability to provide reliable quality sequence information\n",
      "18. In most cases the microarrays were designed to study limited number of genetically similar target pathogens and for many cases the detection methods relied only on recognizing hybridization patterns for identification\n",
      "19. Taking advantage of the sequential base resolution capability of resequencing microarrays that is required for SNP detection resequencing has recently been successfully adapted recently using different approach for organism identification of multiple bacterial and viral pathogens while allowing for fine detailed discrimination of closely related organisms and tracking mutations within the targeted pathogen\n",
      "20. The new methodology differed from earlier work by using the resolved bases as the query of similarity search of DNA databases to identify the most likely species and variants that match the base calls from the hybridization observed\n",
      "21. The system was capable of testing for pathogens simultaneously and could detect the presence of multiple pathogens\n",
      "22. A software program resequencing pathogen identifier REPI was used to simplify data analysis by performing similarity searches of genetic database using basic local alignment search tool BLAST\n",
      "23. The REPI program used BLAST default settings and would only return sequences that might represent the hybridization if the expect value quantity calculated by the BLAST program that indicates the likelihood that the sequence match found would have occurred by random chance in the database was\n",
      "24. This screened out all cases that had insufficient signal however the final determination of what pathogen was detected and to what degree discrimination was possible required manual examination of the returned results\n",
      "25. This method successfully allowed fine discrimination of various adenoviruses and strain identifications of Flu A and B samples in agreement with conventional sampling results\n",
      "26. Two important advantages of this approach were that the information was always recovered at the most detailed level possible and that it was capable of still recognizing organisms with recent mutations\n",
      "27. This approach also maintained specificity well as it was not dependent on the uniqueness of few individual short probes\n",
      "28. Although this analysis method has utility there are several shortcomings it is time consuming not optimized to maximize sensitivity has complicated results is suitable only for an expert and contains redundant or duplicate information\n",
      "29. The process was time consuming because only the initial screening was handled automatically while the remaining steps required manual interpretation before the detection analysis was complete\n",
      "30. Because simple criterion expect value cutoff of and non optimized BLAST parameters were used to consider pathogen detected the REPI algorithm provided list of candidate organisms but did not make final simple conclusion or relate the results of one prototype sequence to another\n",
      "31. Instead manual process was used to make the final determination but because the REPI program provided all similar results and the use of public nucleic acid databases containing redundant entries large amount of data was presented to user that was not useful\n",
      "32. In addition with manual process it was not possible to establish that the algorithm developed was generally applicable for any organism where nucleic acid base resolved sequence information has been provided\n",
      "33. In this paper we describe new software expert system Computer Implemented Biological Sequence Identifier system CIBSI that successfully uses resolved base sequence information from custom designed Affymetrix resequencing microarrays to provide simple list of organisms that are detected\n",
      "34. This algorithm addresses the most important shortcoming of previous methods by incorporating new features to completely automate pathogen identification\n",
      "35. We have demonstrated the effectiveness of this algorithm for identification via several examples\n",
      "36. The single program is capable of making correct decisions for all pathogens contained on the Respiratory Pathogen Microarray RPM whether detected alone or in combinations with improved sensitivity\n",
      "37. Although the program is currently applied to resequencing microarrays the methodologies developed remain generally applicable\n",
      "38. Only the first portion of the algorithm handles issues specific to microarrays while the remainder deals with sequences that are suitable for use as query by the BLAST algorithm\n",
      "39. In developing the general identification algorithm we have identified and resolved issues specific to resequencing microarrays that complicate their use\n",
      "40. Because the entire decision process for what is detected has been automated it is straightforward to test whether the rules used to make identifications are rigorous and applicable to any pathogen\n",
      "41. With this efficient program resequencing based assays can provide competitive method to test simultaneously for many possible pathogens providing output that can be interpreted by non expert\n",
      "42. The details of the RPM design and the experimental methods have been discussed in previous work Lin et al submitted for publication\n",
      "43. Briefly the RPM chip design includes tiled regions allowing resequencing of kb of sequences from respiratory pathogens and biothreat agents\n",
      "44. These were selected based upon clinical relevance for the population of immediate interest United States military recruit in training\n",
      "45. Partial sequences from the genes containing diagnostic regions were tiled for the detection of these pathogens\n",
      "46. The experimental microarray data used in the present analysis were obtained using variety of purified nucleic acid templates and clinical samples culture throat swabs and nasal washes using random and multiplexed RT PCR amplification schemes for more detail description of amplification methods see Supplementary Data\n",
      "47. Resequencing microarrays provide base call resolution by comparing the intensities between set of four mer probes that differ from each other at the same position th base\n",
      "48. An amplicon or target sequence is represented by numerous overlapping probe sets\n",
      "49. GCOS software Affymetrix Inc Santa Clara CA was used to align and scan hybridized microarrays to determine the intensity of each probe in every probe set\n",
      "50. Base calls were made based on the intensity data of each probe set using GDAS software Affymetrix Inc which used an implementation of the ABACUS algorithm\n",
      "51. The sequences were represented in FASTA format for later analysis steps\n",
      "52. In this paper target pathogens are the organisms the assay was specifically designed to detect\n",
      "53. The sets of probes that represent reference sequence selected from target pathogen genomes are referred to as Prototype Sequence or ProSeq for brevity\n",
      "54. The set of resolved bases that result from hybridization of genomic material to ProSeq is referred to as the hybridized sequence or HybSeq\n",
      "55. The HybSeq is split into possible subsequences or SubSeqs\n",
      "56. The CIBSI program implemented in Perl described in this study handled hierarchy of three tasks Figure I ProSeq identification II ProSeq grouping and III pathogen determination\n",
      "57. The most developed and important portion of the algorithm deals with ProSeq identification Task I and is handled in three important subtasks initial filtering of individual HybSeqs into SubSeqs suitable for sequence similarity comparisons Figure database querying of individual SubSeqs Figure and taxonomic comparison of BLAST returns for each SubSeq Figure\n",
      "58. The NCBI BLAST and taxonomy databases were used for the queries and images were obtained on February\n",
      "59. For the ProSeq grouping Task II ProSeqs were compared to determine if they supported the same identified organism\n",
      "60. In the pathogen determination Task III detected organisms were compared to the list of target pathogens the assay was designed for in order to determine if any were positively detected or were possibly related close genetic near neighbors\n",
      "61. The level of discrimination that particular sample supported was automatically determined\n",
      "62. An initial filtering algorithm REPI was developed previously and the general concepts with revisions were incorporated into the current automated detection algorithm used in the CIBSI program\n",
      "63. Filtering and subsequence selection were used to remove potential biasing caused by reference sequence choice and by other sources e\n",
      "64. primers\n",
      "65. When PCR amplification was used microarrays were hybridized in the presence of only primers to determine locations where they resulted in hybridization\n",
      "66. Any portions of the ProSeqs that hybridized with the primers were masked as N calls so that the HybSeq did not contain biased information\n",
      "67. Normally the primers are designed to be outside the ProSeq region to minimize the interference caused by primers and so minimize the bases to be masked\n",
      "68. There is still the chance that some bases require masking because with the large number of primers used in the multiplex short stretches of ProSeq not corresponding to primer locations may still hybridize with the primers\n",
      "69. Such regions could be removed from the reference sequences and so not appear on the microarray\n",
      "70. However determining such locations are difficult and time consuming task that for most cases is not worth the effort\n",
      "71. The first subtask of ProSeq identification Task I is noted in Figure and shown schematically in detail in Figure\n",
      "72. This subtask uses procedure to examine HybSeq to find the longest possible subsequence of base calls SubSeq that can be submitted as query to BLAST\n",
      "73. It produces group of SubSeq that contain all portions of HybSeq that have chance of producing limited list of returns from BLAST query\n",
      "74. When HybSeq has two regions separated by long stretch of continuous N calls the relational positioning of the two regions cannot be trusted and so must be sent as separate queries\n",
      "75. In addition for shorter subsequences the number of base calls that must be made is dependent on the length\n",
      "76. It was also recognized that for very long sequences longer WORD size in BLAST may be used\n",
      "77. A detailed description of the criteria and process used for each step is contained in Supplementary Data\n",
      "78. Upon completion the algorithm returned to the Task I loop and performed the BLAST subtask\n",
      "79. The database query subtask performed batch similarity search of database using SubSeq as the queries\n",
      "80. The BLAST program used was the NCBI Blastall blastn version with defined set of parameters\n",
      "81. The masking of low complex regions was performed for the seeding phase to speed up the query however low complexity repeats were included in the actual scoring\n",
      "82. The entire nucleotide database from NCBI acquired on February was used as the reference database\n",
      "83. Note that earlier images of the database were used during development but all experiments were rerun with the algorithm as described with the image of the database obtained on this date\n",
      "84. The default gap penalty and nucleotide match score were used\n",
      "85. The nucleotide mismatch penalty parameter was set to rather than the default\n",
      "86. The results of any BLAST query with an expect value were returned in tabular format from the blastall program\n",
      "87. The information about each return bit score expect value mismatches length of match was placed in the Return hash key info hash using the SubSeq identity as the hash key for further analysis\n",
      "88. The next subtask of ProSeq identification Task I carried out was the determination of SubSeq states and is shown in Figure\n",
      "89. The BLAST algorithm gives ranking score which can be reported as accounting for the size of the database expect value or not bit score\n",
      "90. The full taxonomic classifications of every return for SubSeq were retrieved from the NCBI taxonomy database obtained on February\n",
      "91. Using the scores and taxonomy relationships it was possible to find reduced number of returns that had the best match with the HybSeq\n",
      "92. These results were summarized by identifying the taxonomic class to which all the returns belonged to identified organism and parameter that indicated how they are related to each other organism uniqueness\n",
      "93. A detailed description of the steps is contained in Supplementary Data\n",
      "94. After each SubSeq was examined the algorithm moved to the next subtask which was to determine the identified organism of the ProSeq from the SubSeq Figure\n",
      "95. The subsequences from the same ProSeq were only allowed to support single identified organism determination\n",
      "96. The procedure shown in Figure demonstrates the decision method used to arrive at this determination detailed description in Supplementary Data\n",
      "97. After the subtask covered in Figure was completed the ProSeq identification Task I loop continued until all ProSeqs were examined\n",
      "98. A list of ProSeqs that had detected organisms was built up in the Result array\n",
      "99. After the ProSeq identification Task I was completed ProSeq grouping Task II Figure was used to examine the identified organism values listed in Result and grouped them together if they identified the same taxonomic class\n",
      "100. Each entry in Result was examined and new entry was created in Result if the identified organism did not appear in this list\n",
      "101. The entries of Result represented the distinct individual organisms identified but might still contain redundant information\n",
      "102. When the ProSeqs were designed to detect the same organism and they all hybridized well this grouping led to reduction in redundant information being reported\n",
      "103. But when one ProSeq did not hybridize as well for variety of possible reasons multiple entries would appear in Result that actually represent hybridization from the same pathogen\n",
      "104. This is because there is an alternative cause for the ProSeq hybridizing in this manner\n",
      "105. This hybridization could be caused by two different but closely related organisms both being present in sample and hybridizing to the microarray\n",
      "106. Because we have not yet developed methods to distinguish these cases no further reduction of the list of organisms is made for in ProSeq grouping Task II in cases where the level of identification varied on different ProSeq targeted for the same organism\n",
      "107. Although it was difficult to relate results from separate ProSeqs to each other it was desired to have simple final detection decision be made in pathogen determination Task III\n",
      "108. The first task was specifically implemented so that information about what ProSeq was intended to detect was not considered and the second task only minimal consideration of this was taken into account\n",
      "109. This allowed these initial tasks to be capable of recognizing not only just positive and negative identifications of target pathogens but also cases that were indeterminate\n",
      "110. In the final task the algorithm considered whether the identified organisms belonged to the list of organisms the ProSeqs were designed to detect\n",
      "111. The task would group organisms from ProSeq grouping Task II together that belonged to or were child classes of the taxonomic class of target organism\n",
      "112. The taxonomic class reported was the common taxonomic group of all the organisms\n",
      "113. When all the ProSeqs for pathogen hybridized well fine level discrimination was reported\n",
      "114. But if one or more ProSeqs hybridized less well the reported positive target pathogen was only identified at the level of the less detailed level\n",
      "115. This is conservative because methods have not yet been developed to clearly discriminate mixtures of very closely related organisms causing different ProSeqs to hybridize from variable hybridization of single organism on several ProSeqs\n",
      "116. The results of all three tasks were reported and more experienced user can view ProSeq grouping Task II results to clarify some cases\n",
      "117. Note that organisms identified in ProSeq grouping Task II that only belonged to target pathogens were reported as positives\n",
      "118. Clear negative ProSeqs were not mentioned in the output\n",
      "119. ProSeqs that were indeterminate or that detected close genetic organisms were never reported as positives\n",
      "120. These organisms were instead reported as being detected\n",
      "121. Purified Chlamydia pneumoniae nucleic acid samples with genome copies via method in Lin et al submitted for publication were chosen to illustrate how pathogen detection and identification were done when multiple ProSeqs were targeted for the same pathogen\n",
      "122. RPM has three highly conserved ProSeqs selected from the genes encoding for the major outer membrane proteins VD and VD and the DNA directed RNA polymerase rpoB gene\n",
      "123. The HybSeqs from the different samples differed only in the number of unique base calls as shown in Table\n",
      "124. The percentage of the ProSeq called varied from to except for one case at concentration of that had only of the rpoB ProSeq producing unique base calls\n",
      "125. Because the samples at this concentration are not reproducibly generating the same percentage of base calls this is probably the detection limit of this ProSeq of the assay\n",
      "126. Table listed the determinations made for the SubSeq and at the end of each task for the various samples\n",
      "127. The ProSeq from the different cases produced the same number of SubSeqs\n",
      "128. These SubSeqs from different samples reported different bit scores for the same top ranked returns from BLAST\n",
      "129. In fact VD and VD produced exactly the same results\n",
      "130. The NCBI taxonomy database classified the returns into four distinct groups which represented the C pneumoniae taxonomic group and three child strain groups\n",
      "131. AE AE AE and BA appeared in the returns of all the ProSeqs for each sample since they represented database entries of completely sequenced genomes\n",
      "132. One rpoB SubSeq produced for its organism uniqueness SeqUniqu\n",
      "133. All other SubSeqs were TaxAmbig as multiple returns from different taxonomic classes were returned\n",
      "134. Since the VD and VD ProSeq each have single SubSeq Task I assigned the ProSeq the state of the SubSeq\n",
      "135. For the rpoB ProSeq the bit score of one SubSeq was large enough that the algorithm assigned that SubSeq identification to the ProSeq\n",
      "136. Task II of the algorithm grouped all three ProSeqs together since they all had the same identified organism and TaxAmbig was assigned\n",
      "137. The result of Task III was positive for target pathogen C pneumoniae and this decision was straightforward as all the ProSeqs agreed with each other and belonged to the same target pathogen taxonomic class\n",
      "138. Although the rpoB ProSeq was SeqUniqu this was not the final conclusion for Task II as the ProSeq that was SeqUniqu was not the child taxonomic group and other ProSeq were TaxAmbig\n",
      "139. The three recognized strains scored the same which indicated that the sequence selected for the ProSeqs was very conserved and would not allow discrimination between the strains\n",
      "140. Influenza and Human Adenovirus HAdV were the only pathogens that had ProSeq selected that would permit detailed strain level discrimination as discussed in previous work\n",
      "141. This previous work using manual analysis found that the microarray results were in excellent agreement with the conventional sequencing results for clinical samples\n",
      "142. A few of the results of running the CIBSI program using the updated NCBI database on the raw microarray results are presented in Table the results for all samples used in the previous work are presented in Supplementary Table A\n",
      "143. The identified organisms were not identical to the original findings due to the difference in database used and because all ProSeqs were considered rather than only the Flu A and B hemagglutinin\n",
      "144. In fact the conventional sequencing results that were submitted to NCBI from that work were found for every sample to be among the returns with the best score for the hemagglutinin ProSeq Supplementary Table B\n",
      "145. It should be noted that the previous work based its analysis upon only the results of the hemagglutinin ProSeq\n",
      "146. For of Influenza A and of Influenza B cases the results of ProSeq identification Task I and ProSeq grouping Task II found that the conventional sequencing was the single best return for the hemagglutinin ProSeq\n",
      "147. Owing to the large number of isolate sequences in the database for the hemagglutinin gene it was not surprising that in some cases single unique entry was not found\n",
      "148. In each of the remaining five Influenza A samples the other sequences returned differed by from the conventional sequence\n",
      "149. The fewer samples with unique isolate identifications for Influenza B were due to an older reference sequence used for the ProSeq which allowed less hybridization to occur\n",
      "150. This also meant that when multiple sequences were returned for sample they represented greater genetic variation up to\n",
      "151. As result of the current method of making pathogen determination Task III level identification the final organism reported was less specific H N or Flu B for every sample than what was reported as possible in ProSeq grouping Task II\n",
      "152. For HAdV samples the algorithm also reproduced the finer scale discriminations that had been made previously by manual methods data not shown\n",
      "153. The next example of detection for the Mycoplasma pneumoniae pathogen demonstrated case where there was only single ProSeq for the target pathogen\n",
      "154. A total of test samples were performed using multiplex PCR via method in Lin et al submitted for publication where for of the samples M pneumoniae organism was spiked into nasal wash with several other pathogens from to colony forming units per ml the remaining samples were purified with nucleic acid from culture stock at concentration of genome copies per reaction volume\n",
      "155. This ProSeq was also not optimal for fine discrimination because it was selected from highly conserved region bp of the cytadhesin P gene\n",
      "156. In every case taxonomic database entries for M pneumoniae or its one recognized distinct strain tied for MaxScore Supplementary Table\n",
      "157. To better understand these returns the database sequences were examined and subdivided into three groups of sequences A B and C based on how well they matched the reference sequence used to make the ProSeq\n",
      "158. The placement of the database entries into the three groups was determined from CLUSTAL alignment of the sequences of this gene\n",
      "159. This alignment confirmed that the database entries differed significantly more from each other in regions not represented by the ProSeq and contained sufficient variability that would have allowed finer discrimination\n",
      "160. Members of Group A exactly matched the ProSeq and could not be distinguished between on the microarray\n",
      "161. Similarly members of group B matched the ProSeq except at the th position where the base called was C rather than T Group C sequences contained few database entries that were more variable and might be distinguished from other entries within the ProSeq\n",
      "162. For the experimental tests of M pneumoniae as much as of the ProSeq hybridized for samples yet only of these samples had an unambiguous base call at the th position\n",
      "163. When it was unambiguous it always matched group B sequences\n",
      "164. In the cases where an N base call was made at the th location both groups A and B sequences were returned with the same score\n",
      "165. Regardless of this the target pathogen positively identified was M pneumoniae for every sample tested\n",
      "166. These examples showed how decisions were made independent of whether single or multiple ProSeqs were dedicated to target pathogen\n",
      "167. They also illustrated that the level of discrimination possible was strongly determined by the quality of the selected ProSeq\n",
      "168. It is possible that for some pathogens fine level discrimination is not required and the currently tested selections on RPM would provide satisfactory information\n",
      "169. The CIBSI algorithm demonstrated its capability to automatically report the maximum level of discrimination that could be supported by the HybSeq information\n",
      "170. To demonstrate how the algorithm handled closely related genetic species sample of non targeted pathogen was considered using multiplex PCR via method in Lin et al submitted for publication\n",
      "171. For Variola major virus one of the biothreat pathogens on the RPMv the validation runs demonstrated that Variola major virus purified DNA templates of plasmids were always positively identified when detected Table\n",
      "172. Table shows the results when purified Vaccinia genomic DNA was spiked into nasal washes and processed at various concentrations using multiplex PCR\n",
      "173. The array has two ProSeqs from hemagglutinin VMVHA bp and cytokine response modifier B VMVcrmB bp genes for Variola major virus detection\n",
      "174. The percentage of the ProSeq that hybridizes is sufficient that if hybridization patterns were only considered one might assume that this tile is identifying the presence of its target\n",
      "175. This would indicate that reference sequence selected was not the best choice\n",
      "176. However when our algorithm was applied none of the samples is in fact identified as Variola major or minor virus\n",
      "177. Vaccinia was always one of the Orthopoxvirus species listed with the highest scores for VMVcrmB ProSeq but in only seven cases was it uniquely identified as the probable species detected\n",
      "178. In only one sample at the lowest concentration and fraction of VMVcrmB hybridizing did this ProSeq even identify Variola major and minor virus as one among the Orthopoxvirus species that could be the cause of the hybridization\n",
      "179. The lower limit of detection for the amplification method used was between this concentration and the one above it for Variola major itself\n",
      "180. The VMVHA ProSeq exhibited much lower sensitivity and made identifications of Orthopoxvirus species in only two experiments and Variola major virus was listed as one of the tied best scoring returns\n",
      "181. In both cases VMVcrmB ProSeq specifically identified Vaccinia virus as the best match\n",
      "182. The percentage of the hybridized ProSeq correlated with concentration of the sample\n",
      "183. This example demonstrated the importance of the filtering portion of the algorithm by considering the HybSeqs of the ProSeqs for the H N neuraminidase NA and matrix genes from human Influenza A Puerto Rico H N strain\n",
      "184. Filtering was necessary because sending the HybSeq of ProSeq to BLAST in single query can bias the scores against strains that have insertions or deletions relative to the ProSeq especially when using BLAST parameters that maximized the use of base calls\n",
      "185. The sliding window test was the portion of the algorithm that controlled filtering\n",
      "186. If filtering were turned off the entire HybSeq would be used in single subsequence for two influenza ProSeqs that showed significant hybridization\n",
      "187. A Weiss H N strain was identified as the most likely strain from the HybSeq of the NA ProSeq while the HybSeq of the matrix ProSeq correctly identified A Puerto Rico\n",
      "188. To better understand the source of biasing CLUSTAL alignment of the NA gene of the two strains and the reference sequence used to make the ProSeq are shown in Figure\n",
      "189. The two strains showed identity mismatches in aligned bases however there was stretch of bases inserted in both A Weiss and the NA ProSeq compared to A Puerto Rico\n",
      "190. With the default filtering on the NA ProSeq was split into five SubSeqs as the algorithm encountered large stretches of no calls\n",
      "191. In ProSeq identification Task I the algorithm determined that three of the SubSeqs which were shorter had an identified organism of H N as several isolates including A Puerto Rico tied with the best score while the two other SubSeqs had an identified organism of only the A Puerto Rico strain as being the closest match\n",
      "192. The organism identified by the NA ProSeq was A Puerto Rico because one of the SubSeqs had much higher score\n",
      "193. This ProSeq supported the same strain identification made in the matrix ProSeq\n",
      "194. The organism identified was A Puerto Rico because two ProSeqs detected only that organism\n",
      "195. The correct target pathogen was detected with filtering while without filtering the target pathogen level of identification would have been Influenza A H N subtype because two organisms were detected A Puerto Rico and A Weiss\n",
      "196. Splitting the HybSeq into SubSeqs to remove biasing may reduce the level of identification as occurred for the three of the five SubSeqs in this case\n",
      "197. The previous example for Vaccinia was another instance where the wrong species Camel Pox or Callithrix jacchus identification would occur if the filtering was not used\n",
      "198. Clinical samples in Table showed that HybSeqs split into multiple SubSeqs were capable of very specific identification\n",
      "199. To illustrate the other filtering performed when multiplex strategies rather than generic were used for amplification Figure also contains the raw and mask filtered results of this region for the A Puerto Rico sample\n",
      "200. It was necessary to perform additional filtering to remove potential biasing caused by the specific primers as described in the methods\n",
      "201. In the case shown in Figure sequence of bases present in the raw result are made N after filtering since they are in location that interacts with the primers\n",
      "202. If these base calls were included in the subsequences constructed even though the HybSeq would still be split into the same number of SubSeq the query for the ProSeq would favor an incorrect strain\n",
      "203. The algorithm we have developed successfully provided pathogen identification to the maximum level of detail possible species or strain depending on the quality of each ProSeq\n",
      "204. This identification capability requires minimal input on the identity of the pathogens making non expert use feasible\n",
      "205. The crucial feature incorporated that allowed complete automation was the use of taxonomic databases which classify organisms into ordered groups and provide relationships between organism entries allowing removal of redundancies comparison of different related prototype sequences and simplification of data presentation\n",
      "206. This allows databases e\n",
      "207. NCBI that are redundant and subject to minimal curation but which constantly receive updated and new sequence information to be used with great success\n",
      "208. Although we have demonstrated this using only the NCBI databases other databases or custom made ones could have easily been used which might improve performance\n",
      "209. The algorithm is capable of providing accurate identifications at all analysis levels for pathogens that are less variable or are represented by highly conserved ProSeqs\n",
      "210. For more variable or rapidly mutating pathogens g\n",
      "211. Influenza A virus ProSeq identification Task I and ProSeq grouping Task II still provided accurate detailed identifications but the pathogen determination Task III was unable to report fine scale discrimination\n",
      "212. The comparison of the conventionally sequenced Influenza virus gene sequences illustrated that the algorithm is capable of automatically adjusting for updates in databases\n",
      "213. The algorithm demonstrated its capability to properly distinguish hybridizations on ProSeq caused by the specified pathogen from those caused by genetically close near neighbor strains and did not make incorrect identifications eliminating one potential cause of false positives\n",
      "214. Filtering the raw hybridization results served to reduce the computation time accounted for potential primer interference and more importantly reduced potential biasing\n",
      "215. This simple integrated algorithm provided sufficient and accurate identification so that immediate use of the RPM or similar resequencing arrays and assay is possible\n",
      "216. Although not discussed in this paper the algorithm has successfully detected the presence of simulated multi infections Lin et al submitted for publication\n",
      "217. The algorithm as currently developed will detect mixtures when the organisms have sufficient variation however detection of mixture of an organism and its mutation strain in sample is uncertain in its present phase\n",
      "218. In principle it may be possible to detect such mixtures as the resequencing microarray can detect and sequence diploid organisms\n",
      "219. Besides demonstrating the success of the CIBSI program the work involved in developing the algorithm allowed insight into the importance of proper ProSeq selection\n",
      "220. The RPM was the first resequencing array designed specifically for multiple pathogen detection using database similarity searching and served as prototype for this application\n",
      "221. We have demonstrated that single ProSeq with as few as bp when designed correctly can be sufficient to unambiguously identify an organism\n",
      "222. However it is clearly indicated that several longer ProSeqs provide better confirmation and more detailed information of pathogen\n",
      "223. Although the algorithm provides accuracy equivalent to manual analysis for determinations of individual ProSeqs the current algorithm is only partially successful in integrating information from multiple ProSeqs\n",
      "224. The emphasis of the design to this point has been on capabilities that are generally applicable to any pathogen\n",
      "225. We are incorporating these insights in our newer more comprehensive resequencing array designs\n",
      "226. Improving on level of detail reported in pathogen determination Task III will require more information about an individual pathogen and may have to be developed for each specific pathogen or class of pathogens\n",
      "227. This information is also required for the algorithm to identify which differences between sample and database entries represent significant mutations\n",
      "228. Future work will involve improving the use of the current taxonomic database or potentially developing new relational database that is specific to our needs and then incorporating more specific information of target pathogens\n",
      "229. The hierarchal design of the data analysis makes it easy to incorporate analysis that build upon the analysis already performed\n",
      "230. We have met with some success in the current version but want to have increased automated discrimination\n",
      "231. We have well defined path to completing this aim\n",
      "232. The use of properly designed resequencing microarrays and this automated detection algorithm provides way forward to developing assays that can test for multiple organisms simultaneously while providing fine strain level discrimination giving access to information about detailed strain recognition antibiotic resistance markers and pathogenicity\n",
      "233. This is capability that other approaches cannot currently provide\n",
      "234. In addition since the design of the original kb RPM microarray the possible sequence content of the current array has increased fold to kb and further increases in array density are still attainable\n",
      "235. This coupled with our identification algorithms will allow the analysis of partial sequence information from even more organisms for applications such as differential diagnostics for illnesses with multiple potential causes e\n",
      "236. febrile respiratory illness tracking of emergent pathogens distinction of biological threats from harmless near genetic neighbors in surveillance applications and for tracking the impact of co infections or super infections\n",
      "237. The concept of categorizing and reporting different degrees of identification depending on the quality of samples and set of target sequences is not limited to resequencing microarrays but is more generally applicable to any platform that is capable of returning sequence level calls that can be used to query reference DNA database\n",
      "238. As the trend for assays that test for multiple pathogens increases automated analysis tools such as this one become more crucial for rapid identification in simple formats useful to the non expert on day to day basis\n",
      "239. The remaining hurdle to using resequencing microarrays as routine assay method is now clearly the sample processing methods\n",
      "240. Further automating these steps is an important area of future research and development\n",
      "241. The program can be obtained free of charge for research purposes by contacting the authors\n",
      "242. Supplementary Data are available at NAR Online\n"
     ]
    }
   ],
   "source": [
    "for number, sentence in enumerate(docs_list[0]['clean_content_sent_tokenized']):\n",
    "    print(str(number+1)+'.', sentence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1. ['surveillance', 'diagnostic', 'application', 'fine', 'scale', 'pathogen', 'identification', 'near', 'neighbor', 'discrimination', 'important', 'therefore', 'assay', 'monitor', 'specific', 'level', 'desirable', 'many', 'type', 'sample', 'clinical', 'environmental']\n",
      "2. ['successfully', 'use', 'method', 'base', 'DNA', 'RNA', 'detection', 'assay', 'must', 'couple', 'large', 'database', 'nucleic', 'acid', 'sequence', 'information', 'assay', 'design', 'ensure', 'desire', 'information', 'provide', 'interpretation', 'raw', 'data']\n",
      "3. ['Several', 'well', 'establish', 'technique', 'use', 'PCR', 'amplify', 'individual', 'target', 'piece', 'sequence', 'genome', 'provide', 'detection', 'organism']\n",
      "4. ['method', 'roughly', 'divide', 'approach', 'target', 'individual', 'short', 'sequence', 'length', 'probe', 'bp', 'method', 'examine', 'longer', 'probe']\n",
      "5. ['advantage', 'use', 'short', 'probe', 'uniqueness', 'probe', 'assure', 'unique', 'primer', 'also', 'select', 'method', 'give', 'good', 'specificity']\n",
      "6. ['approach', 'capable', 'provide', 'fine', 'scale', 'identification', 'several', 'genetically', 'close', 'organism', 'select', 'sufficient', 'number', 'probe']\n",
      "7. ['However', 'rapidly', 'lead', 'large', 'number', 'total', 'probe', 'require', 'detect', 'organism', 'interest']\n",
      "8. ['addition', 'select', 'probe', 'initial', 'selection', 'process', 'determine', 'unique', 'often', 'later', 'found', 'less', 'specific', 'organism', 'sequence', 'less', 'specific', 'condition', 'differ', 'original', 'condition']\n",
      "9. ['particularly', 'problem', 'organism', 'belonging', 'family', 'high', 'mutation', 'rate', 'also', 'pathogen', 'relatively', 'neighbor', 'pathogen', 'sequence']\n",
      "10. ['addition', 'PCR', 'approach', 'focus', 'short', 'unique', 'probe', 'capable', 'detect', 'presence', 'new', 'significant', 'mutation', 'easily', 'resolve', 'base', 'sequence', 'detail']\n",
      "11. ['Approaches', 'use', 'longer', 'individual', 'probe', 'avoid', 'many', 'issue', 'cost', 'less', 'specific']\n",
      "12. ['issue', 'mean', 'approach', 'suitable', 'provide', 'information', 'desire', 'provide', 'impetus', 'work']\n",
      "13. ['High', 'density', 'resequencing', 'microarrays', 'produce', 'variable', 'length', 'segment', 'bp', 'direct', 'sequence', 'information']\n",
      "14. ['target', 'sequence', 'fall', 'longer', 'target', 'regime', 'PCR', 'approach', 'rather', 'hybridize', 'longer', 'less', 'specific', 'probe', 'microarray', 'many', 'shorter', 'specific', 'probe', 'place', 'microarray', 'allow', 'detailed', 'determination', 'entire', 'PCR', 'amplicon']\n",
      "15. ['also', 'mean', 'specificity', 'primer', 'use', 'relaxed']\n",
      "16. ['successfully', 'use', 'detect', 'single', 'nucleotide', 'polymorphism', 'SNP', 'genetic', 'variant', 'viral', 'bacterial', 'eukaryotic', 'genome']\n",
      "17. ['use', 'SNP', 'detection', 'clearly', 'establish', 'ability', 'provide', 'reliable', 'quality', 'sequence', 'information']\n",
      "18. ['case', 'microarrays', 'design', 'study', 'limited', 'number', 'genetically', 'similar', 'target', 'pathogen', 'many', 'case', 'detection', 'method', 'rely', 'recognize', 'hybridization', 'pattern', 'identification']\n",
      "19. ['Taking', 'advantage', 'sequential', 'base', 'resolution', 'capability', 'resequencing', 'microarrays', 'require', 'SNP', 'detection', 'resequencing', 'recently', 'successfully', 'adapt', 'recently', 'use', 'different', 'approach', 'organism', 'identification', 'multiple', 'bacterial', 'viral', 'pathogen', 'allow', 'fine', 'detailed', 'discrimination', 'closely', 'related', 'organism', 'track', 'mutation', 'within', 'target', 'pathogen']\n",
      "20. ['new', 'methodology', 'differ', 'earlier', 'work', 'use', 'resolve', 'base', 'query', 'similarity', 'search', 'DNA', 'database', 'identify', 'likely', 'specie', 'variant', 'match', 'base', 'call', 'hybridization', 'observe']\n",
      "21. ['system', 'capable', 'test', 'pathogen', 'simultaneously', 'could', 'detect', 'presence', 'multiple', 'pathogen']\n",
      "22. ['software', 'program', 'resequencing', 'pathogen', 'identifier', 'REPI', 'use', 'simplify', 'data', 'analysis', 'perform', 'similarity', 'search', 'genetic', 'database', 'use', 'basic', 'local', 'alignment', 'search', 'tool', 'BLAST']\n",
      "23. ['REPI', 'program', 'use', 'BLAST', 'default', 'setting', 'would', 'return', 'sequence', 'might', 'represent', 'hybridization', 'expect', 'value', 'quantity', 'calculate', 'BLAST', 'program', 'indicates', 'likelihood', 'sequence', 'match', 'found', 'would', 'occur', 'random', 'chance', 'database']\n",
      "24. ['screen', 'case', 'insufficient', 'signal', 'however', 'final', 'determination', 'pathogen', 'detect', 'degree', 'discrimination', 'possible', 'require', 'manual', 'examination', 'return', 'result']\n",
      "25. ['method', 'successfully', 'allow', 'fine', 'discrimination', 'various', 'adenovirus', 'strain', 'identification', 'Flu', 'B', 'sample', 'agreement', 'conventional', 'sample', 'result']\n",
      "26. ['Two', 'important', 'advantage', 'approach', 'information', 'always', 'recover', 'detailed', 'level', 'possible', 'capable', 'still', 'recognize', 'organism', 'recent', 'mutation']\n",
      "27. ['approach', 'also', 'maintain', 'specificity', 'well', 'dependent', 'uniqueness', 'individual', 'short', 'probe']\n",
      "28. ['Although', 'analysis', 'method', 'utility', 'several', 'shortcoming', 'time', 'consume', 'optimize', 'maximize', 'sensitivity', 'complicate', 'result', 'suitable', 'expert', 'contains', 'redundant', 'duplicate', 'information']\n",
      "29. ['process', 'time', 'consume', 'initial', 'screen', 'handle', 'automatically', 'remain', 'step', 'require', 'manual', 'interpretation', 'detection', 'analysis', 'complete']\n",
      "30. ['simple', 'criterion', 'expect', 'value', 'cutoff', 'non', 'optimize', 'BLAST', 'parameter', 'use', 'consider', 'pathogen', 'detect', 'REPI', 'algorithm', 'provide', 'list', 'candidate', 'organism', 'make', 'final', 'simple', 'conclusion', 'relate', 'result', 'one', 'prototype', 'sequence', 'another']\n",
      "31. ['Instead', 'manual', 'process', 'use', 'make', 'final', 'determination', 'REPI', 'program', 'provide', 'similar', 'result', 'use', 'public', 'nucleic', 'acid', 'database', 'contain', 'redundant', 'entry', 'large', 'amount', 'data', 'present', 'user', 'useful']\n",
      "32. ['addition', 'manual', 'process', 'possible', 'establish', 'algorithm', 'developed', 'generally', 'applicable', 'organism', 'nucleic', 'acid', 'base', 'resolve', 'sequence', 'information', 'provide']\n",
      "33. ['paper', 'describe', 'new', 'software', 'expert', 'system', 'Computer', 'Implemented', 'Biological', 'Sequence', 'Identifier', 'system', 'CIBSI', 'successfully', 'us', 'resolve', 'base', 'sequence', 'information', 'custom', 'design', 'Affymetrix', 'resequencing', 'microarrays', 'provide', 'simple', 'list', 'organism', 'detect']\n",
      "34. ['algorithm', 'address', 'important', 'shortcoming', 'previous', 'method', 'incorporate', 'new', 'feature', 'completely', 'automate', 'pathogen', 'identification']\n",
      "35. ['demonstrate', 'effectiveness', 'algorithm', 'identification', 'via', 'several', 'example']\n",
      "36. ['single', 'program', 'capable', 'make', 'correct', 'decision', 'pathogen', 'contain', 'Respiratory', 'Pathogen', 'Microarray', 'RPM', 'whether', 'detect', 'alone', 'combination', 'improve', 'sensitivity']\n",
      "37. ['Although', 'program', 'currently', 'apply', 'resequencing', 'microarrays', 'methodology', 'developed', 'remain', 'generally', 'applicable']\n",
      "38. ['first', 'portion', 'algorithm', 'handle', 'issue', 'specific', 'microarrays', 'remainder', 'deal', 'sequence', 'suitable', 'use', 'query', 'BLAST', 'algorithm']\n",
      "39. ['develop', 'general', 'identification', 'algorithm', 'identify', 'resolve', 'issue', 'specific', 'resequencing', 'microarrays', 'complicate', 'use']\n",
      "40. ['entire', 'decision', 'process', 'detect', 'automate', 'straightforward', 'test', 'whether', 'rule', 'use', 'make', 'identification', 'rigorous', 'applicable', 'pathogen']\n",
      "41. ['efficient', 'program', 'resequencing', 'base', 'assay', 'provide', 'competitive', 'method', 'test', 'simultaneously', 'many', 'possible', 'pathogen', 'provide', 'output', 'interpret', 'non', 'expert']\n",
      "42. ['detail', 'RPM', 'design', 'experimental', 'method', 'discuss', 'previous', 'work', 'Lin', 'et', 'al', 'submit', 'publication']\n",
      "43. ['Briefly', 'RPM', 'chip', 'design', 'include', 'tile', 'region', 'allow', 'resequencing', 'kb', 'sequence', 'respiratory', 'pathogen', 'biothreat', 'agent']\n",
      "44. ['select', 'base', 'upon', 'clinical', 'relevance', 'population', 'immediate', 'interest', 'United', 'States', 'military', 'recruit', 'training']\n",
      "45. ['Partial', 'sequence', 'gene', 'contain', 'diagnostic', 'region', 'tile', 'detection', 'pathogen']\n",
      "46. ['experimental', 'microarray', 'data', 'use', 'present', 'analysis', 'obtain', 'use', 'variety', 'purify', 'nucleic', 'acid', 'template', 'clinical', 'sample', 'culture', 'throat', 'swab', 'nasal', 'wash', 'use', 'random', 'multiplexed', 'RT', 'PCR', 'amplification', 'scheme', 'detail', 'description', 'amplification', 'method', 'see', 'Supplementary', 'Data']\n",
      "47. ['Resequencing', 'microarrays', 'provide', 'base', 'call', 'resolution', 'compare', 'intensity', 'set', 'four', 'mer', 'probe', 'differ', 'position', 'th', 'base']\n",
      "48. ['amplicon', 'target', 'sequence', 'represent', 'numerous', 'overlap', 'probe', 'set']\n",
      "49. ['GCOS', 'software', 'Affymetrix', 'Inc', 'Santa', 'Clara', 'CA', 'use', 'align', 'scan', 'hybridize', 'microarrays', 'determine', 'intensity', 'probe', 'every', 'probe', 'set']\n",
      "50. ['Base', 'call', 'make', 'base', 'intensity', 'data', 'probe', 'set', 'use', 'GDAS', 'software', 'Affymetrix', 'Inc', 'use', 'implementation', 'ABACUS', 'algorithm']\n",
      "51. ['sequence', 'represent', 'FASTA', 'format', 'later', 'analysis', 'step']\n",
      "52. ['paper', 'target', 'pathogen', 'organism', 'assay', 'specifically', 'design', 'detect']\n",
      "53. ['set', 'probe', 'represent', 'reference', 'sequence', 'select', 'target', 'pathogen', 'genome', 'refer', 'Prototype', 'Sequence', 'ProSeq', 'brevity']\n",
      "54. ['set', 'resolve', 'base', 'result', 'hybridization', 'genomic', 'material', 'ProSeq', 'refer', 'hybridize', 'sequence', 'HybSeq']\n",
      "55. ['HybSeq', 'split', 'possible', 'subsequence', 'SubSeqs']\n",
      "56. ['CIBSI', 'program', 'implement', 'Perl', 'described', 'study', 'handle', 'hierarchy', 'three', 'task', 'Figure', 'ProSeq', 'identification', 'II', 'ProSeq', 'group', 'III', 'pathogen', 'determination']\n",
      "57. ['developed', 'important', 'portion', 'algorithm', 'deal', 'ProSeq', 'identification', 'Task', 'handle', 'three', 'important', 'subtasks', 'initial', 'filter', 'individual', 'HybSeqs', 'SubSeqs', 'suitable', 'sequence', 'similarity', 'comparison', 'Figure', 'database', 'query', 'individual', 'SubSeqs', 'Figure', 'taxonomic', 'comparison', 'BLAST', 'return', 'SubSeq', 'Figure']\n",
      "58. ['NCBI', 'BLAST', 'taxonomy', 'database', 'use', 'query', 'image', 'obtain', 'February']\n",
      "59. ['ProSeq', 'group', 'Task', 'II', 'ProSeqs', 'compare', 'determine', 'support', 'identify', 'organism']\n",
      "60. ['pathogen', 'determination', 'Task', 'III', 'detect', 'organism', 'compare', 'list', 'target', 'pathogen', 'assay', 'design', 'order', 'determine', 'positively', 'detect', 'possibly', 'related', 'close', 'genetic', 'near', 'neighbor']\n",
      "61. ['level', 'discrimination', 'particular', 'sample', 'support', 'automatically', 'determine']\n",
      "62. ['initial', 'filter', 'algorithm', 'REPI', 'developed', 'previously', 'general', 'concept', 'revision', 'incorporate', 'current', 'automate', 'detection', 'algorithm', 'use', 'CIBSI', 'program']\n",
      "63. ['Filtering', 'subsequence', 'selection', 'use', 'remove', 'potential', 'biasing', 'cause', 'reference', 'sequence', 'choice', 'source', 'e']\n",
      "64. ['primer']\n",
      "65. ['PCR', 'amplification', 'use', 'microarrays', 'hybridize', 'presence', 'primer', 'determine', 'location', 'result', 'hybridization']\n",
      "66. ['portion', 'ProSeqs', 'hybridize', 'primer', 'masked', 'N', 'call', 'HybSeq', 'contain', 'bias', 'information']\n",
      "67. ['Normally', 'primer', 'design', 'outside', 'ProSeq', 'region', 'minimize', 'interference', 'cause', 'primer', 'minimize', 'base', 'masked']\n",
      "68. ['still', 'chance', 'base', 'require', 'mask', 'large', 'number', 'primer', 'use', 'multiplex', 'short', 'stretch', 'ProSeq', 'correspond', 'primer', 'location', 'may', 'still', 'hybridize', 'primer']\n",
      "69. ['region', 'could', 'remove', 'reference', 'sequence', 'appear', 'microarray']\n",
      "70. ['However', 'determine', 'location', 'difficult', 'time', 'consume', 'task', 'case', 'worth', 'effort']\n",
      "71. ['first', 'subtask', 'ProSeq', 'identification', 'Task', 'note', 'Figure', 'show', 'schematically', 'detail', 'Figure']\n",
      "72. ['subtask', 'us', 'procedure', 'examine', 'HybSeq', 'find', 'long', 'possible', 'subsequence', 'base', 'call', 'SubSeq', 'submit', 'query', 'BLAST']\n",
      "73. ['produce', 'group', 'SubSeq', 'contain', 'portion', 'HybSeq', 'chance', 'produce', 'limited', 'list', 'return', 'BLAST', 'query']\n",
      "74. ['HybSeq', 'two', 'region', 'separate', 'long', 'stretch', 'continuous', 'N', 'call', 'relational', 'position', 'two', 'region', 'trust', 'must', 'sent', 'separate', 'query']\n",
      "75. ['addition', 'shorter', 'subsequence', 'number', 'base', 'call', 'must', 'make', 'dependent', 'length']\n",
      "76. ['also', 'recognize', 'long', 'sequence', 'longer', 'WORD', 'size', 'BLAST', 'may', 'use']\n",
      "77. ['detailed', 'description', 'criterion', 'process', 'use', 'step', 'contain', 'Supplementary', 'Data']\n",
      "78. ['Upon', 'completion', 'algorithm', 'return', 'Task', 'loop', 'perform', 'BLAST', 'subtask']\n",
      "79. ['database', 'query', 'subtask', 'perform', 'batch', 'similarity', 'search', 'database', 'use', 'SubSeq', 'query']\n",
      "80. ['BLAST', 'program', 'use', 'NCBI', 'Blastall', 'blastn', 'version', 'define', 'set', 'parameter']\n",
      "81. ['mask', 'low', 'complex', 'region', 'perform', 'seed', 'phase', 'speed', 'query', 'however', 'low', 'complexity', 'repeat', 'include', 'actual', 'score']\n",
      "82. ['entire', 'nucleotide', 'database', 'NCBI', 'acquire', 'February', 'use', 'reference', 'database']\n",
      "83. ['Note', 'earlier', 'image', 'database', 'use', 'development', 'experiment', 'rerun', 'algorithm', 'described', 'image', 'database', 'obtain', 'date']\n",
      "84. ['default', 'gap', 'penalty', 'nucleotide', 'match', 'score', 'use']\n",
      "85. ['nucleotide', 'mismatch', 'penalty', 'parameter', 'set', 'rather', 'default']\n",
      "86. ['result', 'BLAST', 'query', 'expect', 'value', 'return', 'tabular', 'format', 'blastall', 'program']\n",
      "87. ['information', 'return', 'bit', 'score', 'expect', 'value', 'mismatch', 'length', 'match', 'place', 'Return', 'hash', 'key', 'info', 'hash', 'use', 'SubSeq', 'identity', 'hash', 'key', 'far', 'analysis']\n",
      "88. ['next', 'subtask', 'ProSeq', 'identification', 'Task', 'carry', 'determination', 'SubSeq', 'state', 'show', 'Figure']\n",
      "89. ['BLAST', 'algorithm', 'give', 'rank', 'score', 'report', 'accounting', 'size', 'database', 'expect', 'value', 'bit', 'score']\n",
      "90. ['full', 'taxonomic', 'classification', 'every', 'return', 'SubSeq', 'retrieve', 'NCBI', 'taxonomy', 'database', 'obtain', 'February']\n",
      "91. ['Using', 'score', 'taxonomy', 'relationship', 'possible', 'find', 'reduce', 'number', 'return', 'best', 'match', 'HybSeq']\n",
      "92. ['result', 'summarize', 'identify', 'taxonomic', 'class', 'return', 'belong', 'identify', 'organism', 'parameter', 'indicate', 'related', 'organism', 'uniqueness']\n",
      "93. ['detailed', 'description', 'step', 'contain', 'Supplementary', 'Data']\n",
      "94. ['SubSeq', 'examine', 'algorithm', 'move', 'next', 'subtask', 'determine', 'identify', 'organism', 'ProSeq', 'SubSeq', 'Figure']\n",
      "95. ['subsequence', 'ProSeq', 'allow', 'support', 'single', 'identify', 'organism', 'determination']\n",
      "96. ['procedure', 'show', 'Figure', 'demonstrates', 'decision', 'method', 'use', 'arrive', 'determination', 'detailed', 'description', 'Supplementary', 'Data']\n",
      "97. ['subtask', 'cover', 'Figure', 'complete', 'ProSeq', 'identification', 'Task', 'loop', 'continued', 'ProSeqs', 'examine']\n",
      "98. ['list', 'ProSeqs', 'detect', 'organism', 'built', 'Result', 'array']\n",
      "99. ['ProSeq', 'identification', 'Task', 'complete', 'ProSeq', 'group', 'Task', 'II', 'Figure', 'use', 'examine', 'identify', 'organism', 'value', 'list', 'Result', 'grouped', 'together', 'identify', 'taxonomic', 'class']\n",
      "100. ['entry', 'Result', 'examine', 'new', 'entry', 'create', 'Result', 'identify', 'organism', 'appear', 'list']\n",
      "101. ['entry', 'Result', 'represent', 'distinct', 'individual', 'organism', 'identify', 'might', 'still', 'contain', 'redundant', 'information']\n",
      "102. ['ProSeqs', 'design', 'detect', 'organism', 'hybridize', 'well', 'group', 'lead', 'reduction', 'redundant', 'information', 'report']\n",
      "103. ['one', 'ProSeq', 'hybridize', 'well', 'variety', 'possible', 'reason', 'multiple', 'entry', 'would', 'appear', 'Result', 'actually', 'represent', 'hybridization', 'pathogen']\n",
      "104. ['alternative', 'cause', 'ProSeq', 'hybridize', 'manner']\n",
      "105. ['hybridization', 'could', 'cause', 'two', 'different', 'closely', 'related', 'organism', 'present', 'sample', 'hybridize', 'microarray']\n",
      "106. ['yet', 'developed', 'method', 'distinguish', 'case', 'far', 'reduction', 'list', 'organism', 'make', 'ProSeq', 'group', 'Task', 'II', 'case', 'level', 'identification', 'varied', 'different', 'ProSeq', 'target', 'organism']\n",
      "107. ['Although', 'difficult', 'relate', 'result', 'separate', 'ProSeqs', 'desire', 'simple', 'final', 'detection', 'decision', 'make', 'pathogen', 'determination', 'Task', 'III']\n",
      "108. ['first', 'task', 'specifically', 'implement', 'information', 'ProSeq', 'intend', 'detect', 'consider', 'second', 'task', 'minimal', 'consideration', 'take', 'account']\n",
      "109. ['allow', 'initial', 'task', 'capable', 'recognize', 'positive', 'negative', 'identification', 'target', 'pathogen', 'also', 'case', 'indeterminate']\n",
      "110. ['final', 'task', 'algorithm', 'consider', 'whether', 'identify', 'organism', 'belong', 'list', 'organism', 'ProSeqs', 'design', 'detect']\n",
      "111. ['task', 'would', 'group', 'organism', 'ProSeq', 'group', 'Task', 'II', 'together', 'belong', 'child', 'class', 'taxonomic', 'class', 'target', 'organism']\n",
      "112. ['taxonomic', 'class', 'report', 'common', 'taxonomic', 'group', 'organism']\n",
      "113. ['ProSeqs', 'pathogen', 'hybridize', 'well', 'fine', 'level', 'discrimination', 'report']\n",
      "114. ['one', 'ProSeqs', 'hybridize', 'less', 'well', 'report', 'positive', 'target', 'pathogen', 'identify', 'level', 'less', 'detailed', 'level']\n",
      "115. ['conservative', 'method', 'yet', 'developed', 'clearly', 'discriminate', 'mixture', 'closely', 'related', 'organism', 'cause', 'different', 'ProSeqs', 'hybridize', 'variable', 'hybridization', 'single', 'organism', 'several', 'ProSeqs']\n",
      "116. ['result', 'three', 'task', 'report', 'experienced', 'user', 'view', 'ProSeq', 'group', 'Task', 'II', 'result', 'clarify', 'case']\n",
      "117. ['Note', 'organism', 'identify', 'ProSeq', 'group', 'Task', 'II', 'belong', 'target', 'pathogen', 'report', 'positive']\n",
      "118. ['Clear', 'negative', 'ProSeqs', 'mention', 'output']\n",
      "119. ['ProSeqs', 'indeterminate', 'detect', 'close', 'genetic', 'organism', 'never', 'report', 'positive']\n",
      "120. ['organism', 'instead', 'report', 'detect']\n",
      "121. ['Purified', 'Chlamydia', 'pneumoniae', 'nucleic', 'acid', 'sample', 'genome', 'copy', 'via', 'method', 'Lin', 'et', 'al', 'submit', 'publication', 'chosen', 'illustrate', 'pathogen', 'detection', 'identification', 'multiple', 'ProSeqs', 'target', 'pathogen']\n",
      "122. ['RPM', 'three', 'highly', 'conserve', 'ProSeqs', 'select', 'gene', 'encode', 'major', 'outer', 'membrane', 'protein', 'VD', 'VD', 'DNA', 'direct', 'RNA', 'polymerase', 'rpoB', 'gene']\n",
      "123. ['HybSeqs', 'different', 'sample', 'differ', 'number', 'unique', 'base', 'call', 'show', 'Table']\n",
      "124. ['percentage', 'ProSeq', 'call', 'varied', 'except', 'one', 'case', 'concentration', 'rpoB', 'ProSeq', 'produce', 'unique', 'base', 'call']\n",
      "125. ['sample', 'concentration', 'reproducibly', 'generate', 'percentage', 'base', 'call', 'probably', 'detection', 'limit', 'ProSeq', 'assay']\n",
      "126. ['Table', 'list', 'determination', 'make', 'SubSeq', 'end', 'task', 'various', 'sample']\n",
      "127. ['ProSeq', 'different', 'case', 'produce', 'number', 'SubSeqs']\n",
      "128. ['SubSeqs', 'different', 'sample', 'report', 'different', 'bit', 'score', 'top', 'ranked', 'return', 'BLAST']\n",
      "129. ['fact', 'VD', 'VD', 'produce', 'exactly', 'result']\n",
      "130. ['NCBI', 'taxonomy', 'database', 'classify', 'return', 'four', 'distinct', 'group', 'represent', 'C', 'pneumoniae', 'taxonomic', 'group', 'three', 'child', 'strain', 'group']\n",
      "131. ['AE', 'AE', 'AE', 'BA', 'appear', 'return', 'ProSeqs', 'sample', 'since', 'represent', 'database', 'entry', 'completely', 'sequence', 'genome']\n",
      "132. ['One', 'rpoB', 'SubSeq', 'produce', 'organism', 'uniqueness', 'SeqUniqu']\n",
      "133. ['SubSeqs', 'TaxAmbig', 'multiple', 'return', 'different', 'taxonomic', 'class', 'return']\n",
      "134. ['Since', 'VD', 'VD', 'ProSeq', 'single', 'SubSeq', 'Task', 'assign', 'ProSeq', 'state', 'SubSeq']\n",
      "135. ['rpoB', 'ProSeq', 'bit', 'score', 'one', 'SubSeq', 'large', 'enough', 'algorithm', 'assign', 'SubSeq', 'identification', 'ProSeq']\n",
      "136. ['Task', 'II', 'algorithm', 'grouped', 'three', 'ProSeqs', 'together', 'since', 'identify', 'organism', 'TaxAmbig', 'assign']\n",
      "137. ['result', 'Task', 'III', 'positive', 'target', 'pathogen', 'C', 'pneumoniae', 'decision', 'straightforward', 'ProSeqs', 'agree', 'belong', 'target', 'pathogen', 'taxonomic', 'class']\n",
      "138. ['Although', 'rpoB', 'ProSeq', 'SeqUniqu', 'final', 'conclusion', 'Task', 'II', 'ProSeq', 'SeqUniqu', 'child', 'taxonomic', 'group', 'ProSeq', 'TaxAmbig']\n",
      "139. ['three', 'recognize', 'strain', 'score', 'indicate', 'sequence', 'select', 'ProSeqs', 'conserve', 'would', 'allow', 'discrimination', 'strain']\n",
      "140. ['Influenza', 'Human', 'Adenovirus', 'HAdV', 'pathogen', 'ProSeq', 'select', 'would', 'permit', 'detailed', 'strain', 'level', 'discrimination', 'discuss', 'previous', 'work']\n",
      "141. ['previous', 'work', 'use', 'manual', 'analysis', 'found', 'microarray', 'result', 'excellent', 'agreement', 'conventional', 'sequence', 'result', 'clinical', 'sample']\n",
      "142. ['result', 'run', 'CIBSI', 'program', 'use', 'update', 'NCBI', 'database', 'raw', 'microarray', 'result', 'present', 'Table', 'result', 'sample', 'use', 'previous', 'work', 'present', 'Supplementary', 'Table']\n",
      "143. ['identify', 'organism', 'identical', 'original', 'finding', 'due', 'difference', 'database', 'use', 'ProSeqs', 'consider', 'rather', 'Flu', 'B', 'hemagglutinin']\n",
      "144. ['fact', 'conventional', 'sequence', 'result', 'submit', 'NCBI', 'work', 'found', 'every', 'sample', 'among', 'return', 'best', 'score', 'hemagglutinin', 'ProSeq', 'Supplementary', 'Table', 'B']\n",
      "145. ['note', 'previous', 'work', 'base', 'analysis', 'upon', 'result', 'hemagglutinin', 'ProSeq']\n",
      "146. ['Influenza', 'Influenza', 'B', 'case', 'result', 'ProSeq', 'identification', 'Task', 'ProSeq', 'group', 'Task', 'II', 'found', 'conventional', 'sequence', 'single', 'best', 'return', 'hemagglutinin', 'ProSeq']\n",
      "147. ['Owing', 'large', 'number', 'isolate', 'sequence', 'database', 'hemagglutinin', 'gene', 'surprising', 'case', 'single', 'unique', 'entry', 'found']\n",
      "148. ['remain', 'five', 'Influenza', 'sample', 'sequence', 'return', 'differ', 'conventional', 'sequence']\n",
      "149. ['sample', 'unique', 'isolate', 'identification', 'Influenza', 'B', 'due', 'old', 'reference', 'sequence', 'use', 'ProSeq', 'allow', 'less', 'hybridization', 'occur']\n",
      "150. ['also', 'meant', 'multiple', 'sequence', 'return', 'sample', 'represent', 'great', 'genetic', 'variation']\n",
      "151. ['result', 'current', 'method', 'make', 'pathogen', 'determination', 'Task', 'III', 'level', 'identification', 'final', 'organism', 'report', 'less', 'specific', 'H', 'N', 'Flu', 'B', 'every', 'sample', 'report', 'possible', 'ProSeq', 'group', 'Task', 'II']\n",
      "152. ['HAdV', 'sample', 'algorithm', 'also', 'reproduce', 'finer', 'scale', 'discrimination', 'make', 'previously', 'manual', 'method', 'data', 'show']\n",
      "153. ['next', 'example', 'detection', 'Mycoplasma', 'pneumoniae', 'pathogen', 'demonstrate', 'case', 'single', 'ProSeq', 'target', 'pathogen']\n",
      "154. ['total', 'test', 'sample', 'perform', 'use', 'multiplex', 'PCR', 'via', 'method', 'Lin', 'et', 'al', 'submit', 'publication', 'sample', 'pneumoniae', 'organism', 'spike', 'nasal', 'wash', 'several', 'pathogen', 'colony', 'form', 'unit', 'per', 'ml', 'remain', 'sample', 'purify', 'nucleic', 'acid', 'culture', 'stock', 'concentration', 'genome', 'copy', 'per', 'reaction', 'volume']\n",
      "155. ['ProSeq', 'also', 'optimal', 'fine', 'discrimination', 'select', 'highly', 'conserve', 'region', 'bp', 'cytadhesin', 'P', 'gene']\n",
      "156. ['every', 'case', 'taxonomic', 'database', 'entry', 'pneumoniae', 'one', 'recognize', 'distinct', 'strain', 'tie', 'MaxScore', 'Supplementary', 'Table']\n",
      "157. ['well', 'understand', 'return', 'database', 'sequence', 'examine', 'subdivide', 'three', 'group', 'sequence', 'B', 'C', 'base', 'well', 'match', 'reference', 'sequence', 'use', 'make', 'ProSeq']\n",
      "158. ['placement', 'database', 'entry', 'three', 'group', 'determine', 'CLUSTAL', 'alignment', 'sequence', 'gene']\n",
      "159. ['alignment', 'confirm', 'database', 'entry', 'differ', 'significantly', 'region', 'represent', 'ProSeq', 'contain', 'sufficient', 'variability', 'would', 'allow', 'finer', 'discrimination']\n",
      "160. ['Members', 'Group', 'exactly', 'match', 'ProSeq', 'could', 'distinguish', 'microarray']\n",
      "161. ['Similarly', 'member', 'group', 'B', 'match', 'ProSeq', 'except', 'th', 'position', 'base', 'call', 'C', 'rather', 'Group', 'C', 'sequence', 'contain', 'database', 'entry', 'variable', 'might', 'distinguish', 'entry', 'within', 'ProSeq']\n",
      "162. ['experimental', 'test', 'pneumoniae', 'much', 'ProSeq', 'hybridize', 'sample', 'yet', 'sample', 'unambiguous', 'base', 'call', 'th', 'position']\n",
      "163. ['unambiguous', 'always', 'match', 'group', 'B', 'sequence']\n",
      "164. ['case', 'N', 'base', 'call', 'make', 'th', 'location', 'group', 'B', 'sequence', 'return', 'score']\n",
      "165. ['Regardless', 'target', 'pathogen', 'positively', 'identify', 'pneumoniae', 'every', 'sample', 'test']\n",
      "166. ['example', 'show', 'decision', 'make', 'independent', 'whether', 'single', 'multiple', 'ProSeqs', 'dedicate', 'target', 'pathogen']\n",
      "167. ['also', 'illustrate', 'level', 'discrimination', 'possible', 'strongly', 'determine', 'quality', 'select', 'ProSeq']\n",
      "168. ['possible', 'pathogen', 'fine', 'level', 'discrimination', 'require', 'currently', 'test', 'selection', 'RPM', 'would', 'provide', 'satisfactory', 'information']\n",
      "169. ['CIBSI', 'algorithm', 'demonstrate', 'capability', 'automatically', 'report', 'maximum', 'level', 'discrimination', 'could', 'support', 'HybSeq', 'information']\n",
      "170. ['demonstrate', 'algorithm', 'handle', 'closely', 'related', 'genetic', 'specie', 'sample', 'non', 'target', 'pathogen', 'consider', 'use', 'multiplex', 'PCR', 'via', 'method', 'Lin', 'et', 'al', 'submit', 'publication']\n",
      "171. ['Variola', 'major', 'virus', 'one', 'biothreat', 'pathogen', 'RPMv', 'validation', 'run', 'demonstrate', 'Variola', 'major', 'virus', 'purify', 'DNA', 'template', 'plasmid', 'always', 'positively', 'identify', 'detect', 'Table']\n",
      "172. ['Table', 'show', 'result', 'purify', 'Vaccinia', 'genomic', 'DNA', 'spike', 'nasal', 'wash', 'process', 'various', 'concentration', 'use', 'multiplex', 'PCR']\n",
      "173. ['array', 'two', 'ProSeqs', 'hemagglutinin', 'VMVHA', 'bp', 'cytokine', 'response', 'modifier', 'B', 'VMVcrmB', 'bp', 'gene', 'Variola', 'major', 'virus', 'detection']\n",
      "174. ['percentage', 'ProSeq', 'hybridizes', 'sufficient', 'hybridization', 'pattern', 'consider', 'one', 'might', 'assume', 'tile', 'identify', 'presence', 'target']\n",
      "175. ['would', 'indicate', 'reference', 'sequence', 'select', 'best', 'choice']\n",
      "176. ['However', 'algorithm', 'apply', 'none', 'sample', 'fact', 'identify', 'Variola', 'major', 'minor', 'virus']\n",
      "177. ['Vaccinia', 'always', 'one', 'Orthopoxvirus', 'specie', 'list', 'high', 'score', 'VMVcrmB', 'ProSeq', 'seven', 'case', 'uniquely', 'identify', 'probable', 'specie', 'detect']\n",
      "178. ['one', 'sample', 'low', 'concentration', 'fraction', 'VMVcrmB', 'hybridize', 'ProSeq', 'even', 'identify', 'Variola', 'major', 'minor', 'virus', 'one', 'among', 'Orthopoxvirus', 'specie', 'could', 'cause', 'hybridization']\n",
      "179. ['low', 'limit', 'detection', 'amplification', 'method', 'use', 'concentration', 'one', 'Variola', 'major']\n",
      "180. ['VMVHA', 'ProSeq', 'exhibit', 'much', 'low', 'sensitivity', 'make', 'identification', 'Orthopoxvirus', 'specie', 'two', 'experiment', 'Variola', 'major', 'virus', 'list', 'one', 'tie', 'best', 'score', 'return']\n",
      "181. ['case', 'VMVcrmB', 'ProSeq', 'specifically', 'identify', 'Vaccinia', 'virus', 'best', 'match']\n",
      "182. ['percentage', 'hybridize', 'ProSeq', 'correlate', 'concentration', 'sample']\n",
      "183. ['example', 'demonstrate', 'importance', 'filter', 'portion', 'algorithm', 'consider', 'HybSeqs', 'ProSeqs', 'H', 'N', 'neuraminidase', 'NA', 'matrix', 'gene', 'human', 'Influenza', 'Puerto', 'Rico', 'H', 'N', 'strain']\n",
      "184. ['Filtering', 'necessary', 'send', 'HybSeq', 'ProSeq', 'BLAST', 'single', 'query', 'bias', 'score', 'strain', 'insertion', 'deletion', 'relative', 'ProSeq', 'especially', 'use', 'BLAST', 'parameter', 'maximize', 'use', 'base', 'call']\n",
      "185. ['slide', 'window', 'test', 'portion', 'algorithm', 'control', 'filter']\n",
      "186. ['filter', 'turn', 'entire', 'HybSeq', 'would', 'use', 'single', 'subsequence', 'two', 'influenza', 'ProSeqs', 'show', 'significant', 'hybridization']\n",
      "187. ['Weiss', 'H', 'N', 'strain', 'identify', 'likely', 'strain', 'HybSeq', 'NA', 'ProSeq', 'HybSeq', 'matrix', 'ProSeq', 'correctly', 'identify', 'Puerto', 'Rico']\n",
      "188. ['well', 'understand', 'source', 'biasing', 'CLUSTAL', 'alignment', 'NA', 'gene', 'two', 'strain', 'reference', 'sequence', 'use', 'make', 'ProSeq', 'show', 'Figure']\n",
      "189. ['two', 'strain', 'show', 'identity', 'mismatch', 'align', 'base', 'however', 'stretch', 'base', 'insert', 'Weiss', 'NA', 'ProSeq', 'compare', 'Puerto', 'Rico']\n",
      "190. ['default', 'filter', 'NA', 'ProSeq', 'split', 'five', 'SubSeqs', 'algorithm', 'encounter', 'large', 'stretch', 'call']\n",
      "191. ['ProSeq', 'identification', 'Task', 'algorithm', 'determine', 'three', 'SubSeqs', 'shorter', 'identify', 'organism', 'H', 'N', 'several', 'isolates', 'include', 'Puerto', 'Rico', 'tie', 'best', 'score', 'two', 'SubSeqs', 'identify', 'organism', 'Puerto', 'Rico', 'strain', 'closest', 'match']\n",
      "192. ['organism', 'identify', 'NA', 'ProSeq', 'Puerto', 'Rico', 'one', 'SubSeqs', 'much', 'high', 'score']\n",
      "193. ['ProSeq', 'support', 'strain', 'identification', 'make', 'matrix', 'ProSeq']\n",
      "194. ['organism', 'identify', 'Puerto', 'Rico', 'two', 'ProSeqs', 'detect', 'organism']\n",
      "195. ['correct', 'target', 'pathogen', 'detect', 'filter', 'without', 'filter', 'target', 'pathogen', 'level', 'identification', 'would', 'Influenza', 'H', 'N', 'subtype', 'two', 'organism', 'detect', 'Puerto', 'Rico', 'Weiss']\n",
      "196. ['Splitting', 'HybSeq', 'SubSeqs', 'remove', 'biasing', 'may', 'reduce', 'level', 'identification', 'occur', 'three', 'five', 'SubSeqs', 'case']\n",
      "197. ['previous', 'example', 'Vaccinia', 'another', 'instance', 'wrong', 'specie', 'Camel', 'Pox', 'Callithrix', 'jacchus', 'identification', 'would', 'occur', 'filter', 'use']\n",
      "198. ['Clinical', 'sample', 'Table', 'show', 'HybSeqs', 'split', 'multiple', 'SubSeqs', 'capable', 'specific', 'identification']\n",
      "199. ['illustrate', 'filter', 'perform', 'multiplex', 'strategy', 'rather', 'generic', 'use', 'amplification', 'Figure', 'also', 'contains', 'raw', 'mask', 'filter', 'result', 'region', 'Puerto', 'Rico', 'sample']\n",
      "200. ['necessary', 'perform', 'additional', 'filter', 'remove', 'potential', 'biasing', 'cause', 'specific', 'primer', 'described', 'method']\n",
      "201. ['case', 'show', 'Figure', 'sequence', 'base', 'present', 'raw', 'result', 'make', 'N', 'filter', 'since', 'location', 'interacts', 'primer']\n",
      "202. ['base', 'call', 'include', 'subsequence', 'construct', 'even', 'though', 'HybSeq', 'would', 'still', 'split', 'number', 'SubSeq', 'query', 'ProSeq', 'would', 'favor', 'incorrect', 'strain']\n",
      "203. ['algorithm', 'developed', 'successfully', 'provide', 'pathogen', 'identification', 'maximum', 'level', 'detail', 'possible', 'specie', 'strain', 'depend', 'quality', 'ProSeq']\n",
      "204. ['identification', 'capability', 'require', 'minimal', 'input', 'identity', 'pathogen', 'make', 'non', 'expert', 'use', 'feasible']\n",
      "205. ['crucial', 'feature', 'incorporate', 'allow', 'complete', 'automation', 'use', 'taxonomic', 'database', 'classify', 'organism', 'order', 'group', 'provide', 'relationship', 'organism', 'entry', 'allow', 'removal', 'redundancy', 'comparison', 'different', 'related', 'prototype', 'sequence', 'simplification', 'data', 'presentation']\n",
      "206. ['allows', 'database', 'e']\n",
      "207. ['NCBI', 'redundant', 'subject', 'minimal', 'curation', 'constantly', 'receive', 'update', 'new', 'sequence', 'information', 'use', 'great', 'success']\n",
      "208. ['Although', 'demonstrate', 'use', 'NCBI', 'database', 'database', 'custom', 'make', 'one', 'could', 'easily', 'use', 'might', 'improve', 'performance']\n",
      "209. ['algorithm', 'capable', 'provide', 'accurate', 'identification', 'analysis', 'level', 'pathogen', 'less', 'variable', 'represent', 'highly', 'conserve', 'ProSeqs']\n",
      "210. ['variable', 'rapidly', 'mutate', 'pathogen', 'g']\n",
      "211. ['Influenza', 'virus', 'ProSeq', 'identification', 'Task', 'ProSeq', 'group', 'Task', 'II', 'still', 'provide', 'accurate', 'detailed', 'identification', 'pathogen', 'determination', 'Task', 'III', 'unable', 'report', 'fine', 'scale', 'discrimination']\n",
      "212. ['comparison', 'conventionally', 'sequence', 'Influenza', 'virus', 'gene', 'sequence', 'illustrate', 'algorithm', 'capable', 'automatically', 'adjust', 'update', 'database']\n",
      "213. ['algorithm', 'demonstrate', 'capability', 'properly', 'distinguish', 'hybridization', 'ProSeq', 'cause', 'specify', 'pathogen', 'cause', 'genetically', 'close', 'near', 'neighbor', 'strain', 'make', 'incorrect', 'identification', 'eliminate', 'one', 'potential', 'cause', 'false', 'positive']\n",
      "214. ['Filtering', 'raw', 'hybridization', 'result', 'serve', 'reduce', 'computation', 'time', 'account', 'potential', 'primer', 'interference', 'importantly', 'reduce', 'potential', 'biasing']\n",
      "215. ['simple', 'integrate', 'algorithm', 'provide', 'sufficient', 'accurate', 'identification', 'immediate', 'use', 'RPM', 'similar', 'resequencing', 'array', 'assay', 'possible']\n",
      "216. ['Although', 'discuss', 'paper', 'algorithm', 'successfully', 'detect', 'presence', 'simulated', 'multi', 'infection', 'Lin', 'et', 'al', 'submit', 'publication']\n",
      "217. ['algorithm', 'currently', 'developed', 'detect', 'mixture', 'organism', 'sufficient', 'variation', 'however', 'detection', 'mixture', 'organism', 'mutation', 'strain', 'sample', 'uncertain', 'present', 'phase']\n",
      "218. ['principle', 'may', 'possible', 'detect', 'mixture', 'resequencing', 'microarray', 'detect', 'sequence', 'diploid', 'organism']\n",
      "219. ['Besides', 'demonstrate', 'success', 'CIBSI', 'program', 'work', 'involve', 'develop', 'algorithm', 'allow', 'insight', 'importance', 'proper', 'ProSeq', 'selection']\n",
      "220. ['RPM', 'first', 'resequencing', 'array', 'design', 'specifically', 'multiple', 'pathogen', 'detection', 'use', 'database', 'similarity', 'search', 'serve', 'prototype', 'application']\n",
      "221. ['demonstrate', 'single', 'ProSeq', 'bp', 'design', 'correctly', 'sufficient', 'unambiguously', 'identify', 'organism']\n",
      "222. ['However', 'clearly', 'indicate', 'several', 'longer', 'ProSeqs', 'provide', 'well', 'confirmation', 'detailed', 'information', 'pathogen']\n",
      "223. ['Although', 'algorithm', 'provide', 'accuracy', 'equivalent', 'manual', 'analysis', 'determination', 'individual', 'ProSeqs', 'current', 'algorithm', 'partially', 'successful', 'integrate', 'information', 'multiple', 'ProSeqs']\n",
      "224. ['emphasis', 'design', 'point', 'capability', 'generally', 'applicable', 'pathogen']\n",
      "225. ['incorporate', 'insight', 'newer', 'comprehensive', 'resequencing', 'array', 'design']\n",
      "226. ['Improving', 'level', 'detail', 'report', 'pathogen', 'determination', 'Task', 'III', 'require', 'information', 'individual', 'pathogen', 'may', 'developed', 'specific', 'pathogen', 'class', 'pathogen']\n",
      "227. ['information', 'also', 'require', 'algorithm', 'identify', 'difference', 'sample', 'database', 'entry', 'represent', 'significant', 'mutation']\n",
      "228. ['Future', 'work', 'involve', 'improve', 'use', 'current', 'taxonomic', 'database', 'potentially', 'develop', 'new', 'relational', 'database', 'specific', 'need', 'incorporate', 'specific', 'information', 'target', 'pathogen']\n",
      "229. ['hierarchal', 'design', 'data', 'analysis', 'make', 'easy', 'incorporate', 'analysis', 'build', 'upon', 'analysis', 'already', 'perform']\n",
      "230. ['met', 'success', 'current', 'version', 'want', 'increase', 'automate', 'discrimination']\n",
      "231. ['well', 'define', 'path', 'complete', 'aim']\n",
      "232. ['use', 'properly', 'design', 'resequencing', 'microarrays', 'automate', 'detection', 'algorithm', 'provide', 'way', 'forward', 'develop', 'assay', 'test', 'multiple', 'organism', 'simultaneously', 'provide', 'fine', 'strain', 'level', 'discrimination', 'give', 'access', 'information', 'detailed', 'strain', 'recognition', 'antibiotic', 'resistance', 'marker', 'pathogenicity']\n",
      "233. ['capability', 'approach', 'currently', 'provide']\n",
      "234. ['addition', 'since', 'design', 'original', 'kb', 'RPM', 'microarray', 'possible', 'sequence', 'content', 'current', 'array', 'increase', 'fold', 'kb', 'far', 'increase', 'array', 'density', 'still', 'attainable']\n",
      "235. ['couple', 'identification', 'algorithm', 'allow', 'analysis', 'partial', 'sequence', 'information', 'even', 'organism', 'application', 'differential', 'diagnostics', 'illness', 'multiple', 'potential', 'cause', 'e']\n",
      "236. ['febrile', 'respiratory', 'illness', 'track', 'emergent', 'pathogen', 'distinction', 'biological', 'threat', 'harmless', 'near', 'genetic', 'neighbor', 'surveillance', 'application', 'track', 'impact', 'co', 'infection', 'super', 'infection']\n",
      "237. ['concept', 'categorize', 'reporting', 'different', 'degree', 'identification', 'depend', 'quality', 'sample', 'set', 'target', 'sequence', 'limited', 'resequencing', 'microarrays', 'generally', 'applicable', 'platform', 'capable', 'return', 'sequence', 'level', 'call', 'use', 'query', 'reference', 'DNA', 'database']\n",
      "238. ['trend', 'assay', 'test', 'multiple', 'pathogen', 'increase', 'automate', 'analysis', 'tool', 'one', 'become', 'crucial', 'rapid', 'identification', 'simple', 'format', 'useful', 'non', 'expert', 'day', 'day', 'basis']\n",
      "239. ['remain', 'hurdle', 'use', 'resequencing', 'microarrays', 'routine', 'assay', 'method', 'clearly', 'sample', 'processing', 'method']\n",
      "240. ['automate', 'step', 'important', 'area', 'future', 'research', 'development']\n",
      "241. ['program', 'obtain', 'free', 'charge', 'research', 'purpose', 'contact', 'author']\n",
      "242. ['Supplementary', 'Data', 'available', 'NAR', 'Online']\n"
     ]
    }
   ],
   "source": [
    "for number, sentence in enumerate(docs_list[0]['no_stopwords']):\n",
    "    print(str(number+1)+'.', sentence)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Corpus of documents"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "class CorpusCreator:\n",
    "    \n",
    "    def __init__(self,docs_list):\n",
    "        self.docs_list = docs_list\n",
    "        \n",
    "    def createCorpus(self, doc_field_key):\n",
    "        corpus = []\n",
    "        sentence_dict = {}\n",
    "        doc_nr = 0\n",
    "        sentence_nr = 0\n",
    "        for doc in docs_list:    \n",
    "            for sentence in doc[doc_field_key]: \n",
    "                sentence_dict[sentence_nr] = doc_nr  \n",
    "                corpus.append(' '.join(sentence))\n",
    "                sentence_nr += 1\n",
    "            doc_nr += 1\n",
    "        return corpus, sentence_dict\n",
    "    \n",
    "corpus_creator = CorpusCreator(docs_list)\n",
    "\n",
    "corpus_sentences, sentence_dict = corpus_creator.createCorpus('no_stopwords')\n",
    "\n",
    "corpus = []\n",
    "for doc in docs_list:\n",
    "    doc_content = \"\"\n",
    "    for sentence_list in doc['no_stopwords']:\n",
    "        doc_content += \" \".join(sentence_list)+ \" \"\n",
    "    corpus.append(doc_content.strip())    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "surveillance diagnostic application fine scale pathogen identification near neighbor discrimination important therefore assay monitor specific level desirable many type sample clinical environmental successfully use method base DNA RNA detection assay must couple large database nucleic acid sequence information assay design ensure desire information provide interpretation raw data Several well establish technique use PCR amplify individual target piece sequence genome provide detection organism method roughly divide approach target individual short sequence length probe bp method examine longer probe advantage use short probe uniqueness probe assure unique primer also select method give good specificity approach capable provide fine scale identification several genetically close organism select sufficient number probe However rapidly lead large number total probe require detect organism interest addition select probe initial selection process determine unique often later found less specific organism sequence less specific condition differ original condition particularly problem organism belonging family high mutation rate also pathogen relatively neighbor pathogen sequence addition PCR approach focus short unique probe capable detect presence new significant mutation easily resolve base sequence detail Approaches use longer individual probe avoid many issue cost less specific issue mean approach suitable provide information desire provide impetus work High density resequencing microarrays produce variable length segment bp direct sequence information target sequence fall longer target regime PCR approach rather hybridize longer less specific probe microarray many shorter specific probe place microarray allow detailed determination entire PCR amplicon also mean specificity primer use relaxed successfully use detect single nucleotide polymorphism SNP genetic variant viral bacterial eukaryotic genome use SNP detection clearly establish ability provide reliable quality sequence information case microarrays design study limited number genetically similar target pathogen many case detection method rely recognize hybridization pattern identification Taking advantage sequential base resolution capability resequencing microarrays require SNP detection resequencing recently successfully adapt recently use different approach organism identification multiple bacterial viral pathogen allow fine detailed discrimination closely related organism track mutation within target pathogen new methodology differ earlier work use resolve base query similarity search DNA database identify likely specie variant match base call hybridization observe system capable test pathogen simultaneously could detect presence multiple pathogen software program resequencing pathogen identifier REPI use simplify data analysis perform similarity search genetic database use basic local alignment search tool BLAST REPI program use BLAST default setting would return sequence might represent hybridization expect value quantity calculate BLAST program indicates likelihood sequence match found would occur random chance database screen case insufficient signal however final determination pathogen detect degree discrimination possible require manual examination return result method successfully allow fine discrimination various adenovirus strain identification Flu B sample agreement conventional sample result Two important advantage approach information always recover detailed level possible capable still recognize organism recent mutation approach also maintain specificity well dependent uniqueness individual short probe Although analysis method utility several shortcoming time consume optimize maximize sensitivity complicate result suitable expert contains redundant duplicate information process time consume initial screen handle automatically remain step require manual interpretation detection analysis complete simple criterion expect value cutoff non optimize BLAST parameter use consider pathogen detect REPI algorithm provide list candidate organism make final simple conclusion relate result one prototype sequence another Instead manual process use make final determination REPI program provide similar result use public nucleic acid database contain redundant entry large amount data present user useful addition manual process possible establish algorithm developed generally applicable organism nucleic acid base resolve sequence information provide paper describe new software expert system Computer Implemented Biological Sequence Identifier system CIBSI successfully us resolve base sequence information custom design Affymetrix resequencing microarrays provide simple list organism detect algorithm address important shortcoming previous method incorporate new feature completely automate pathogen identification demonstrate effectiveness algorithm identification via several example single program capable make correct decision pathogen contain Respiratory Pathogen Microarray RPM whether detect alone combination improve sensitivity Although program currently apply resequencing microarrays methodology developed remain generally applicable first portion algorithm handle issue specific microarrays remainder deal sequence suitable use query BLAST algorithm develop general identification algorithm identify resolve issue specific resequencing microarrays complicate use entire decision process detect automate straightforward test whether rule use make identification rigorous applicable pathogen efficient program resequencing base assay provide competitive method test simultaneously many possible pathogen provide output interpret non expert detail RPM design experimental method discuss previous work Lin et al submit publication Briefly RPM chip design include tile region allow resequencing kb sequence respiratory pathogen biothreat agent select base upon clinical relevance population immediate interest United States military recruit training Partial sequence gene contain diagnostic region tile detection pathogen experimental microarray data use present analysis obtain use variety purify nucleic acid template clinical sample culture throat swab nasal wash use random multiplexed RT PCR amplification scheme detail description amplification method see Supplementary Data Resequencing microarrays provide base call resolution compare intensity set four mer probe differ position th base amplicon target sequence represent numerous overlap probe set GCOS software Affymetrix Inc Santa Clara CA use align scan hybridize microarrays determine intensity probe every probe set Base call make base intensity data probe set use GDAS software Affymetrix Inc use implementation ABACUS algorithm sequence represent FASTA format later analysis step paper target pathogen organism assay specifically design detect set probe represent reference sequence select target pathogen genome refer Prototype Sequence ProSeq brevity set resolve base result hybridization genomic material ProSeq refer hybridize sequence HybSeq HybSeq split possible subsequence SubSeqs CIBSI program implement Perl described study handle hierarchy three task Figure ProSeq identification II ProSeq group III pathogen determination developed important portion algorithm deal ProSeq identification Task handle three important subtasks initial filter individual HybSeqs SubSeqs suitable sequence similarity comparison Figure database query individual SubSeqs Figure taxonomic comparison BLAST return SubSeq Figure NCBI BLAST taxonomy database use query image obtain February ProSeq group Task II ProSeqs compare determine support identify organism pathogen determination Task III detect organism compare list target pathogen assay design order determine positively detect possibly related close genetic near neighbor level discrimination particular sample support automatically determine initial filter algorithm REPI developed previously general concept revision incorporate current automate detection algorithm use CIBSI program Filtering subsequence selection use remove potential biasing cause reference sequence choice source e primer PCR amplification use microarrays hybridize presence primer determine location result hybridization portion ProSeqs hybridize primer masked N call HybSeq contain bias information Normally primer design outside ProSeq region minimize interference cause primer minimize base masked still chance base require mask large number primer use multiplex short stretch ProSeq correspond primer location may still hybridize primer region could remove reference sequence appear microarray However determine location difficult time consume task case worth effort first subtask ProSeq identification Task note Figure show schematically detail Figure subtask us procedure examine HybSeq find long possible subsequence base call SubSeq submit query BLAST produce group SubSeq contain portion HybSeq chance produce limited list return BLAST query HybSeq two region separate long stretch continuous N call relational position two region trust must sent separate query addition shorter subsequence number base call must make dependent length also recognize long sequence longer WORD size BLAST may use detailed description criterion process use step contain Supplementary Data Upon completion algorithm return Task loop perform BLAST subtask database query subtask perform batch similarity search database use SubSeq query BLAST program use NCBI Blastall blastn version define set parameter mask low complex region perform seed phase speed query however low complexity repeat include actual score entire nucleotide database NCBI acquire February use reference database Note earlier image database use development experiment rerun algorithm described image database obtain date default gap penalty nucleotide match score use nucleotide mismatch penalty parameter set rather default result BLAST query expect value return tabular format blastall program information return bit score expect value mismatch length match place Return hash key info hash use SubSeq identity hash key far analysis next subtask ProSeq identification Task carry determination SubSeq state show Figure BLAST algorithm give rank score report accounting size database expect value bit score full taxonomic classification every return SubSeq retrieve NCBI taxonomy database obtain February Using score taxonomy relationship possible find reduce number return best match HybSeq result summarize identify taxonomic class return belong identify organism parameter indicate related organism uniqueness detailed description step contain Supplementary Data SubSeq examine algorithm move next subtask determine identify organism ProSeq SubSeq Figure subsequence ProSeq allow support single identify organism determination procedure show Figure demonstrates decision method use arrive determination detailed description Supplementary Data subtask cover Figure complete ProSeq identification Task loop continued ProSeqs examine list ProSeqs detect organism built Result array ProSeq identification Task complete ProSeq group Task II Figure use examine identify organism value list Result grouped together identify taxonomic class entry Result examine new entry create Result identify organism appear list entry Result represent distinct individual organism identify might still contain redundant information ProSeqs design detect organism hybridize well group lead reduction redundant information report one ProSeq hybridize well variety possible reason multiple entry would appear Result actually represent hybridization pathogen alternative cause ProSeq hybridize manner hybridization could cause two different closely related organism present sample hybridize microarray yet developed method distinguish case far reduction list organism make ProSeq group Task II case level identification varied different ProSeq target organism Although difficult relate result separate ProSeqs desire simple final detection decision make pathogen determination Task III first task specifically implement information ProSeq intend detect consider second task minimal consideration take account allow initial task capable recognize positive negative identification target pathogen also case indeterminate final task algorithm consider whether identify organism belong list organism ProSeqs design detect task would group organism ProSeq group Task II together belong child class taxonomic class target organism taxonomic class report common taxonomic group organism ProSeqs pathogen hybridize well fine level discrimination report one ProSeqs hybridize less well report positive target pathogen identify level less detailed level conservative method yet developed clearly discriminate mixture closely related organism cause different ProSeqs hybridize variable hybridization single organism several ProSeqs result three task report experienced user view ProSeq group Task II result clarify case Note organism identify ProSeq group Task II belong target pathogen report positive Clear negative ProSeqs mention output ProSeqs indeterminate detect close genetic organism never report positive organism instead report detect Purified Chlamydia pneumoniae nucleic acid sample genome copy via method Lin et al submit publication chosen illustrate pathogen detection identification multiple ProSeqs target pathogen RPM three highly conserve ProSeqs select gene encode major outer membrane protein VD VD DNA direct RNA polymerase rpoB gene HybSeqs different sample differ number unique base call show Table percentage ProSeq call varied except one case concentration rpoB ProSeq produce unique base call sample concentration reproducibly generate percentage base call probably detection limit ProSeq assay Table list determination make SubSeq end task various sample ProSeq different case produce number SubSeqs SubSeqs different sample report different bit score top ranked return BLAST fact VD VD produce exactly result NCBI taxonomy database classify return four distinct group represent C pneumoniae taxonomic group three child strain group AE AE AE BA appear return ProSeqs sample since represent database entry completely sequence genome One rpoB SubSeq produce organism uniqueness SeqUniqu SubSeqs TaxAmbig multiple return different taxonomic class return Since VD VD ProSeq single SubSeq Task assign ProSeq state SubSeq rpoB ProSeq bit score one SubSeq large enough algorithm assign SubSeq identification ProSeq Task II algorithm grouped three ProSeqs together since identify organism TaxAmbig assign result Task III positive target pathogen C pneumoniae decision straightforward ProSeqs agree belong target pathogen taxonomic class Although rpoB ProSeq SeqUniqu final conclusion Task II ProSeq SeqUniqu child taxonomic group ProSeq TaxAmbig three recognize strain score indicate sequence select ProSeqs conserve would allow discrimination strain Influenza Human Adenovirus HAdV pathogen ProSeq select would permit detailed strain level discrimination discuss previous work previous work use manual analysis found microarray result excellent agreement conventional sequence result clinical sample result run CIBSI program use update NCBI database raw microarray result present Table result sample use previous work present Supplementary Table identify organism identical original finding due difference database use ProSeqs consider rather Flu B hemagglutinin fact conventional sequence result submit NCBI work found every sample among return best score hemagglutinin ProSeq Supplementary Table B note previous work base analysis upon result hemagglutinin ProSeq Influenza Influenza B case result ProSeq identification Task ProSeq group Task II found conventional sequence single best return hemagglutinin ProSeq Owing large number isolate sequence database hemagglutinin gene surprising case single unique entry found remain five Influenza sample sequence return differ conventional sequence sample unique isolate identification Influenza B due old reference sequence use ProSeq allow less hybridization occur also meant multiple sequence return sample represent great genetic variation result current method make pathogen determination Task III level identification final organism report less specific H N Flu B every sample report possible ProSeq group Task II HAdV sample algorithm also reproduce finer scale discrimination make previously manual method data show next example detection Mycoplasma pneumoniae pathogen demonstrate case single ProSeq target pathogen total test sample perform use multiplex PCR via method Lin et al submit publication sample pneumoniae organism spike nasal wash several pathogen colony form unit per ml remain sample purify nucleic acid culture stock concentration genome copy per reaction volume ProSeq also optimal fine discrimination select highly conserve region bp cytadhesin P gene every case taxonomic database entry pneumoniae one recognize distinct strain tie MaxScore Supplementary Table well understand return database sequence examine subdivide three group sequence B C base well match reference sequence use make ProSeq placement database entry three group determine CLUSTAL alignment sequence gene alignment confirm database entry differ significantly region represent ProSeq contain sufficient variability would allow finer discrimination Members Group exactly match ProSeq could distinguish microarray Similarly member group B match ProSeq except th position base call C rather Group C sequence contain database entry variable might distinguish entry within ProSeq experimental test pneumoniae much ProSeq hybridize sample yet sample unambiguous base call th position unambiguous always match group B sequence case N base call make th location group B sequence return score Regardless target pathogen positively identify pneumoniae every sample test example show decision make independent whether single multiple ProSeqs dedicate target pathogen also illustrate level discrimination possible strongly determine quality select ProSeq possible pathogen fine level discrimination require currently test selection RPM would provide satisfactory information CIBSI algorithm demonstrate capability automatically report maximum level discrimination could support HybSeq information demonstrate algorithm handle closely related genetic specie sample non target pathogen consider use multiplex PCR via method Lin et al submit publication Variola major virus one biothreat pathogen RPMv validation run demonstrate Variola major virus purify DNA template plasmid always positively identify detect Table Table show result purify Vaccinia genomic DNA spike nasal wash process various concentration use multiplex PCR array two ProSeqs hemagglutinin VMVHA bp cytokine response modifier B VMVcrmB bp gene Variola major virus detection percentage ProSeq hybridizes sufficient hybridization pattern consider one might assume tile identify presence target would indicate reference sequence select best choice However algorithm apply none sample fact identify Variola major minor virus Vaccinia always one Orthopoxvirus specie list high score VMVcrmB ProSeq seven case uniquely identify probable specie detect one sample low concentration fraction VMVcrmB hybridize ProSeq even identify Variola major minor virus one among Orthopoxvirus specie could cause hybridization low limit detection amplification method use concentration one Variola major VMVHA ProSeq exhibit much low sensitivity make identification Orthopoxvirus specie two experiment Variola major virus list one tie best score return case VMVcrmB ProSeq specifically identify Vaccinia virus best match percentage hybridize ProSeq correlate concentration sample example demonstrate importance filter portion algorithm consider HybSeqs ProSeqs H N neuraminidase NA matrix gene human Influenza Puerto Rico H N strain Filtering necessary send HybSeq ProSeq BLAST single query bias score strain insertion deletion relative ProSeq especially use BLAST parameter maximize use base call slide window test portion algorithm control filter filter turn entire HybSeq would use single subsequence two influenza ProSeqs show significant hybridization Weiss H N strain identify likely strain HybSeq NA ProSeq HybSeq matrix ProSeq correctly identify Puerto Rico well understand source biasing CLUSTAL alignment NA gene two strain reference sequence use make ProSeq show Figure two strain show identity mismatch align base however stretch base insert Weiss NA ProSeq compare Puerto Rico default filter NA ProSeq split five SubSeqs algorithm encounter large stretch call ProSeq identification Task algorithm determine three SubSeqs shorter identify organism H N several isolates include Puerto Rico tie best score two SubSeqs identify organism Puerto Rico strain closest match organism identify NA ProSeq Puerto Rico one SubSeqs much high score ProSeq support strain identification make matrix ProSeq organism identify Puerto Rico two ProSeqs detect organism correct target pathogen detect filter without filter target pathogen level identification would Influenza H N subtype two organism detect Puerto Rico Weiss Splitting HybSeq SubSeqs remove biasing may reduce level identification occur three five SubSeqs case previous example Vaccinia another instance wrong specie Camel Pox Callithrix jacchus identification would occur filter use Clinical sample Table show HybSeqs split multiple SubSeqs capable specific identification illustrate filter perform multiplex strategy rather generic use amplification Figure also contains raw mask filter result region Puerto Rico sample necessary perform additional filter remove potential biasing cause specific primer described method case show Figure sequence base present raw result make N filter since location interacts primer base call include subsequence construct even though HybSeq would still split number SubSeq query ProSeq would favor incorrect strain algorithm developed successfully provide pathogen identification maximum level detail possible specie strain depend quality ProSeq identification capability require minimal input identity pathogen make non expert use feasible crucial feature incorporate allow complete automation use taxonomic database classify organism order group provide relationship organism entry allow removal redundancy comparison different related prototype sequence simplification data presentation allows database e NCBI redundant subject minimal curation constantly receive update new sequence information use great success Although demonstrate use NCBI database database custom make one could easily use might improve performance algorithm capable provide accurate identification analysis level pathogen less variable represent highly conserve ProSeqs variable rapidly mutate pathogen g Influenza virus ProSeq identification Task ProSeq group Task II still provide accurate detailed identification pathogen determination Task III unable report fine scale discrimination comparison conventionally sequence Influenza virus gene sequence illustrate algorithm capable automatically adjust update database algorithm demonstrate capability properly distinguish hybridization ProSeq cause specify pathogen cause genetically close near neighbor strain make incorrect identification eliminate one potential cause false positive Filtering raw hybridization result serve reduce computation time account potential primer interference importantly reduce potential biasing simple integrate algorithm provide sufficient accurate identification immediate use RPM similar resequencing array assay possible Although discuss paper algorithm successfully detect presence simulated multi infection Lin et al submit publication algorithm currently developed detect mixture organism sufficient variation however detection mixture organism mutation strain sample uncertain present phase principle may possible detect mixture resequencing microarray detect sequence diploid organism Besides demonstrate success CIBSI program work involve develop algorithm allow insight importance proper ProSeq selection RPM first resequencing array design specifically multiple pathogen detection use database similarity search serve prototype application demonstrate single ProSeq bp design correctly sufficient unambiguously identify organism However clearly indicate several longer ProSeqs provide well confirmation detailed information pathogen Although algorithm provide accuracy equivalent manual analysis determination individual ProSeqs current algorithm partially successful integrate information multiple ProSeqs emphasis design point capability generally applicable pathogen incorporate insight newer comprehensive resequencing array design Improving level detail report pathogen determination Task III require information individual pathogen may developed specific pathogen class pathogen information also require algorithm identify difference sample database entry represent significant mutation Future work involve improve use current taxonomic database potentially develop new relational database specific need incorporate specific information target pathogen hierarchal design data analysis make easy incorporate analysis build upon analysis already perform met success current version want increase automate discrimination well define path complete aim use properly design resequencing microarrays automate detection algorithm provide way forward develop assay test multiple organism simultaneously provide fine strain level discrimination give access information detailed strain recognition antibiotic resistance marker pathogenicity capability approach currently provide addition since design original kb RPM microarray possible sequence content current array increase fold kb far increase array density still attainable couple identification algorithm allow analysis partial sequence information even organism application differential diagnostics illness multiple potential cause e febrile respiratory illness track emergent pathogen distinction biological threat harmless near genetic neighbor surveillance application track impact co infection super infection concept categorize reporting different degree identification depend quality sample set target sequence limited resequencing microarrays generally applicable platform capable return sequence level call use query reference DNA database trend assay test multiple pathogen increase automate analysis tool one become crucial rapid identification simple format useful non expert day day basis remain hurdle use resequencing microarrays routine assay method clearly sample processing method automate step important area future research development program obtain free charge research purpose contact author Supplementary Data available NAR Online\n"
     ]
    }
   ],
   "source": [
    "print(corpus[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2075 <-- programmed ribosomal frameshift PRF non conventional translation phenomenon pertains particular change reading frame messenger RNA mRNA induced stimulatory signal\n",
      "2076 <-- strategy mainly use virus synthesize precursor enzyme maintain specific ratio structural enzymatic protein\n",
      "2077 <-- addition PRF use translation prokaryotic eukaryotic mRNAs\n",
      "2078 <-- One best know example PRF occurs ribosome translate full length mRNA human immunodeficiency virus type HIV\n",
      "2079 <-- PRF induced two ci element within mRNA slippery heptanucleotide X XXY YYZ X nucleotide either U Z G eukaryote space indicate initial reading frame PRF occurs follow specific RNA secondary structure call stimulatory signal\n",
      "2080 <-- RNA structure often pseudoknot also stem loop case HIV three way stem loop found bacterial insertion sequence\n",
      "2081 <-- stimulatory signal control PRF efficiency make ribosome pause slippery sequence\n",
      "2082 <-- However pause sufficient promote PRF propose stimulatory signal specific interaction ribosome\n",
      "2083 <-- Altering PRF efficiency impairs viral replication observe even small change PRF efficiency substantially handicap replication capacity HIV\n",
      "2084 <-- indicates PRF event could serve target design development new antiviral drug\n",
      "2085 <-- thus important fully understand PRF mechanism\n",
      "2086 <-- Several mechanistic model discuss literature account PRF\n",
      "2087 <-- Initially Jacks et al\n",
      "2088 <-- propose model simultaneous slippage peptidyl tRNA pept tRNA aminoacyl tRNA aa tRNA\n",
      "2089 <-- model Figure pept tRNA aa tRNA bound respectively XXY YYZ codon ribosomal P P site unpair mRNA first second letter represent respectively binding site small large ribosomal subunit\n",
      "2090 <-- tRNAs ribosome shift towards direction pair mRNA new reading frame\n",
      "2091 <-- Jacks et al\n",
      "2092 <-- model criticize ignores fact peptide bond formation occurs rapidly leaf much time shift occur accommodation aa tRNA site\n",
      "2093 <-- refinement Jacks et al\n",
      "2094 <-- model provide Plant et al\n",
      "2095 <-- suggest movement Å anticodon loop aa tRNA upon occupancy site creates tension mRNA resistance unwind stimulatory signal\n",
      "2096 <-- tension would relieve unpairing tRNAs slippage mRNA one base direction pair tRNAs new reading frame\n",
      "2097 <-- However propose Å displacement anticodon loop aa tRNA support structure analysis large scale molecular dynamic\n",
      "2098 <-- second model propose Weiss et al\n",
      "2099 <-- PRF occurs translocational step elongation cycle\n",
      "2100 <-- Translocation proceeds stepwise manner require conformational change within ribosome\n",
      "2101 <-- simplify way peptide bond formation acceptor stem newly deacylated tRNA deac tRNA pept tRNA move respectively P E site P site large ribosomal subunit\n",
      "2102 <-- result position tRNAs thus described P E P site respectively\n",
      "2103 <-- next step anticodon stem loop tRNAs move E P site small ribosomal subunit drag mRNA one codon\n",
      "2104 <-- Weiss et al\n",
      "2105 <-- suggest tRNAs occupy P E P site unpair mRNA move direction ribosome pair new reading frame\n",
      "2106 <-- anticodon stem loop tRNAs move mRNA E P site small ribosomal subunit Figure B\n",
      "2107 <-- analysis electron cryo microscopy cryo EM ribosome mRNA pseudoknot complex stall process PRF support hypothesis event occurs translocation show elongation factor EF eEF eukaryotic homologue EF G bound stall complex\n",
      "2108 <-- However observe PRF affected mutation small subunit rRNA alter accommodation aa tRNA site well mutation elongation factor eEF eukaryotic homologue EF Tu contributes accommodation process\n",
      "2109 <-- observation explain two previous model take account third model initially propose Farabaugh far support experimental data group\n",
      "2110 <-- model proposes PRF occurs aa tRNA pept tRNA locate respectively entry site P P site Figure C\n",
      "2111 <-- condition PRF time take place compare model Jacks et al\n",
      "2112 <-- Still observation related PRF explain model\n",
      "2113 <-- particular swap fragment eight nucleotide locate immediately slippery sequence HIV frameshift region correspond fragment human cell leukaemia virus type HTLV found decrease PRF efficiency\n",
      "2114 <-- study reveal additional aspect important PRF use HIV frameshift region model\n",
      "2115 <-- HIV slippery sequence U UUU UUA stimulatory signal two stem helix internal three nucleotide bulge separate low stem upper stem loop Figure\n",
      "2116 <-- signal separate slippery sequence one nucleotide\n",
      "2117 <-- previously suggest low stem favour specific interaction upper stem loop ribosome stimulatory signal first encounter ribosome\n",
      "2118 <-- first contact ribosome progress along mRNA unwinds low stem\n",
      "2119 <-- slippery sequence occupies decode site upper stem loop distance eight nucleotide slippery sequence act effective frameshift stimulatory signal\n",
      "2120 <-- allows PRF occur involve tRNAs interact P site codon\n",
      "2121 <-- use dual luciferase reporter system HIV frameshift region insert cod sequence Renilla Rluc firefly luciferase Fluc Fluc expression require PRF HIV demonstrate mutation E site codon alter PRF efficiency\n",
      "2122 <-- observation make viral slippery sequence\n",
      "2123 <-- addition since recently show PRF HIV recapitulate Escherichia coli prokaryotic ribosome respond manner eukaryotic ribosome mutation HIV frameshift stimulatory signal use PRF system harbour E coli undertake random search mutation rRNA affect PRF\n",
      "2124 <-- found three mutation kind locate platform region small ribosomal subunit\n",
      "2125 <-- region position close binding site E site tRNA know involve conformational change ribosome translocation\n",
      "2126 <-- Based finding propose novel model PRF trigger incomplete translocation require slippage tRNAs interact P site codon also tRNA occupy E site\n",
      "2127 <-- primer use generate construct purchase Integrated DNA Technologies Inc Coralville IA USA\n",
      "2128 <-- Plasmids use frameshift assay mammalian culture cell derivative pDual HIV plasmid described Dulude et al\n",
      "2129 <-- \n",
      "2130 <-- pDual HIV plasmid contains HIV group subtype B frameshift region insert sequence cod Rluc Fluc Fluc cod sequence reading frame relative Rluc initiator codon\n",
      "2131 <-- pDual HIV Fluc frame Rluc initiator codon addition adenine adjacent slippery sequence inactivate mutagenesis C UUC CUC\n",
      "2132 <-- Rluc expression use normalize Fluc expression\n",
      "2133 <-- PRF efficiency construct obtain divide Fluc Rluc ratio correspond ratio frame construct Figure\n",
      "2134 <-- mutant C C construct generate amplification mutate DNA pDual HIV use standard PCR procedure primer span KpnI site GCAGGGGGTACCTGGAAAGGAAGGACACCAAATGAAAGATTGTTCGAGAGACAG GCNNNNNNNNNNGGGAAGATCTGG N corresponds mutate nucleotide show Figures B primer span Pfl II site GCCAACCGA ACGGACATTTCG forward reverse reaction respectively\n",
      "2135 <-- mutate DNA subcloned pDual HIV previously digest KpnI Pfl II restriction enzyme\n",
      "2136 <-- Mutations rRNA ΔG iC G select specialized bacterial ribosome system use RGFP HIV plasmid described detail elsewhere\n",
      "2137 <-- Briefly RGFP HIV contains E coli rrnB operon control inducible lacUV promoter well two reporter gene cod red RFP green GFP fluorescent protein\n",
      "2138 <-- DsRed plasmid cod RFP generous gift Dr B\n",
      "2139 <-- Glick University Chicago GFP cod sequence obtain pGFPemd N plasmid kind gift Dr Bouvier Université de Montréal\n",
      "2140 <-- RFP express conventional translation whereas HIV frameshift region insert begin cod sequence GFP expression require PRF\n",
      "2141 <-- ribosome binding site RBS reporter change AUCCC messenger binding site MBS rRNA change GGGAU reporter exclusively translate ribosome contain plasmid encode rRNA GFP RFP system\n",
      "2142 <-- Mutations iG ΔC introduce rRNA amplification mutate DNA fragment RGFP HIV two step PCR approach encompass restriction site BlnI DraIII use overlap extension procedure\n",
      "2143 <-- resultant PCR fragment subcloned RGFP HIV previously digest enzyme\n",
      "2144 <-- specialized bacterial dual luciferase system create follow Rluc Fluc cod sequence encompass HIV frameshift region pDual HIV construct amplify PCR use primer span NsiI site CTAGAGCCACC ATGCATACCAGCAAGG primer span Pfl II site GTTTCATAGCTTCTGCCAACCGAACG forward reverse reaction respectively\n",
      "2145 <-- resultant PCR fragment subcloned pRNAluc plasmid digest NsiI Pfl II described Bélanger et al\n",
      "2146 <-- generate pDual HIV P plasmid P stand prokaryote plasmid\n",
      "2147 <-- pRNAluc plasmid contains E coli rrnB operon control inducible lacUV promoter reporter gene cod Fluc\n",
      "2148 <-- GFP RFP system RBS dual luciferase reporter MBS rRNA mutate dual luciferase reporter exclusively translate ribosome contain plasmid encode rRNA\n",
      "2149 <-- rRNA mutation RGFP HIV ΔG iC G iG ΔC clone pDual HIV P use two ApaI restriction site\n",
      "2150 <-- Random mutation introduce rRNA fragment use high copy plasmid pUC Fermentas\n",
      "2151 <-- BamHI SacI fragment encompass rRNA RGFP HIV clone pUC digest enzyme generate pUC\n",
      "2152 <-- E coli XL Red mutator strain Stratagene use produce random mutation rRNA\n",
      "2153 <-- Escherichia coli XL Red competent cell transform pUC direct manufacturer obtain randomly mutate plasmid pUC R R stand randomize rRNA\n",
      "2154 <-- procedure repeat seven time clone analyse contain mutation rRNA fragment\n",
      "2155 <-- second BlnI restriction site locate begin gene cod rRNA RGFP HIV mutate AflII restriction site use standard mutagenesis procedure conserve unique BlnI restriction site rRNA gene\n",
      "2156 <-- rRNA random mutant library pUC R clone RGFP HIV use BlnI DraIII restriction site generate RGFP HIV R\n",
      "2157 <-- isolated rRNA mutation introduce RGFP HIV verify phenotype conserve\n",
      "2158 <-- Frameshift assay eukaryote monitor transient transfection dual luciferase plasmid pDual HIV derivative human embryonic kidney cell HEK maintain Dulbecco modify Eagle medium DMEM supplement v FBS Wisent cell well seed well plate\n",
      "2159 <-- Two μg plasmid dilute initial culture volume fresh serum free medium mixed equivalent volume contain μg PEI polyethylenimine Polysciences Inc\n",
      "2160 <-- mixture incubate min room temperature prior addition cell culture\n",
      "2161 <-- Cells culture wash twice ml PBS lysed μl Cell Passive Lysis Buffer X Promega\n",
      "2162 <-- Fluc versus Rluc activity construct measure relative light unit RLU EG G Berthold Lumat LB luminometer use non commercial dual luciferase enzyme assay system\n",
      "2163 <-- Dual luciferase GFP RFP assay E coli conduct follow overnight culture E coli Top strain Invitrogen transform pDual HIV P RGFP HIV derivative\n",
      "2164 <-- Plasmids grown LB contain μg ml ampicillin C\n",
      "2165 <-- culture dilute absorbance nm incubate C\n",
      "2166 <-- plasmid encode rRNA expression induced mM isopropyl β thiogalactopyra noside IPTG dual luciferase assay GFP RFP assay C\n",
      "2167 <-- dual luciferase assay carry luciferase assay described Bélanger et al\n",
      "2168 <-- except cell lysed μl Cell Passive Lysis Buffer X incubate min room temperature\n",
      "2169 <-- Fluc Rluc activity measure described\n",
      "2170 <-- GFP RFP assay ml culture cell centrifuge\n",
      "2171 <-- pellet wash three time µl PBS suspend µl PBS\n",
      "2172 <-- Fluorescence measure Fusion Universal Microplate Analyser Fusion α FP Packard nm excitation wavelength bandpass nm GFP RFP nm bandpass nm nm bandpass nm emission wavelength GFP RFP respectively\n",
      "2173 <-- mention earlier mutation within eight nucleotide immediately precede slippery sequence affect efficiency PRF\n",
      "2174 <-- may expect closer nucleotide region slippery heptamer X XXY YYZ high influence PRF\n",
      "2175 <-- study focus three nucleotide position immediately upstream slippery heptamer name B C Together slippery heptamer nucleotide form decamer BCX XXY YYZ call extend slippery sequence\n",
      "2176 <-- last model describe PRF Figure C XXY YYZ codon classic slippery sequence X XXY YYZ locate P site respectively PRF occurs\n",
      "2177 <-- BC nucleotide directly upstream slippery sequence position BCX ABC codon locate respectively E site shift\n",
      "2178 <-- investigate whether identity nucleotide BC affect PRF study seven mutant frameshift region HIV BC nucleotide sequence U AAU UUU UUA replace correspond sequence found virus\n",
      "2179 <-- use dual luciferase reporter system mammalian culture cell\n",
      "2180 <-- system HIV frameshift region position Rluc Fluc cod sequence Fluc expression require PRF HIV\n",
      "2181 <-- Figure one see extend slippery sequence wild type HIV WT U AAU UUU UUA seven mutant numbered\n",
      "2182 <-- mutant corresponds variant isolated protease inhibitor expose patient natural variant HIV\n",
      "2183 <-- mutant nucleotide BC take HIV type HIV giardia virus severe acute respiratory syndrome coronavirus SARS CoV human cell lymphotropic virus type HTLV equine infectious anaemia virus EIAV\n",
      "2184 <-- absolute value WT PRF efficiency\n",
      "2185 <-- Figure B value arbitrarily assign efficiency efficiency mutant show relative WT\n",
      "2186 <-- Among investigate mutant characterize low level PRF efficiency WT level\n",
      "2187 <-- drop efficiency explain fact variant pair pept tRNA new reading frame allow variant likely poorly infectious\n",
      "2188 <-- mutant PRF efficiency increase\n",
      "2189 <-- contrast PRF efficiency decrease slightly wild type value frameshift level remain unchanged\n",
      "2190 <-- Although effect dramatic significant well reproducible mutant P value mutant P value determine Student test\n",
      "2191 <-- verify whether PRF could also influence mutation upstream BC nucleotide mutate GCU codon immediately upstream BCX codon GCU exchange AAU UCU GGU CAU position remain unchanged\n",
      "2192 <-- mutation alter PRF efficiency data show\n",
      "2193 <-- conclude mutation position BC immediately upstream classic slippery sequence noticeably affect PRF efficiency HIV therefore nucleotide locate position BC could involve PRF HIV\n",
      "2194 <-- case nucleotide upstream BC\n",
      "2195 <-- next investigate whether identity BC nucleotide also important PRF viral slippery sequence\n",
      "2196 <-- test create various construct extend slippery sequence HIV U AAU UUU UUA replace extend viral slippery sequence\n",
      "2197 <-- two stem helix HIV use stimulatory signal different slippery sequence assayed eukaryotic dual luciferase system described show Figure\n",
      "2198 <-- C C C construct contain extend slippery sequence found respectively giardia virus C AUC CCU UUA EIAV U CCA AAA AAC SARS CoV G UUU UUA AAC\n",
      "2199 <-- PRF efficiency obtain three construct vary HIV PRF efficiency Figure B\n",
      "2200 <-- nucleotide position BC C C C construct mutate UAA nucleotide found position BC upstream classic slippery sequence HIV generate C C C chimeric construct\n",
      "2201 <-- decrease PRF efficiency compare C C C construct confirms observation make HIV slippery sequence concern importance nucleotide BC PRF\n",
      "2202 <-- previously demonstrate exchange slippery sequence different frameshift region alters PRF efficiency\n",
      "2203 <-- However study heptanucleotide slippery sequence exchange\n",
      "2204 <-- Figure B also compare effect exchange either classic extend HIV slippery sequence viral slippery sequence PRF efficiency\n",
      "2205 <-- classic slippery sequence HIV replace classic slippery sequence giardia virus EIAV SARS CoV PRF efficiency drop compare HIV PRF efficiency\n",
      "2206 <-- HIV extend slippery sequence replace extend slippery sequence found giardia virus EIAV SARS CoV PRF efficiency also decrease much less substitution involves classic heptamer HIV PRF efficiency\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2207 <-- result confirm exchange slippery sequence alters PRF efficiency addition show extent decrease depends upon whether extend slippery sequence classic slippery sequence exchange\n",
      "2208 <-- result also confirm importance position BC PRF\n",
      "2209 <-- also attempt identify rRNA mutant could interfere PRF HIV\n",
      "2210 <-- far rRNA mutation found influence PRF perturb accommodation process\n",
      "2211 <-- aim select mutation influence PRF affect E site tRNA binding since found mutation E site codon alter PRF\n",
      "2212 <-- strategy consist use bacterial ribosome introduce random mutation rRNA high mutation rate E coli XL Red mutator strain see Materials Methods section detail\n",
      "2213 <-- pool randomly mutate plasmid obtain clone analyse contain mutation rRNA\n",
      "2214 <-- rRNA random library clone plasmid contain reporter gene cod RFP GFP RFP express conventional translation whereas GFP expression require PRF HIV\n",
      "2215 <-- RBS reporter MBS site rRNA random library mutate remain complementary reporter exclusively translate ribosome contain plasmid encode rRNA\n",
      "2216 <-- small screen individually perform clone use GFP RFP assay select clone GFP expression affected conventional translation moderately decrease\n",
      "2217 <-- Three mutant select mutant ΔG guanosine delete stretch guanosine position mutant iC cytosine insert stretch cytosine position mutant G guanosine replace adenosine\n",
      "2218 <-- Figure show variation HIV PRF efficiency obtain select mutant relative wild type rRNA set arbitrarily base GFP RFP assay\n",
      "2219 <-- ΔG decrease PRF efficiency mildly whereas G iC increase PRF efficiency respectively\n",
      "2220 <-- also create bacterial dual luciferase vector similar one use mammalian culture cell dual luciferase reporter exclusively translate ribosome contain plasmid encode rRNA\n",
      "2221 <-- rRNA mutant introduce vector PRF efficiency assess measurement luciferase activity\n",
      "2222 <-- mutant G iC PRF efficiency increase\n",
      "2223 <-- However surprisingly PRF efficiency also increase mutant ΔG despite fact decrease GFP RFP assay\n",
      "2224 <-- discrepancy GFP RFP bacterial dual luciferase assay could result fact HIV frameshift region locate short distance AUG initiator codon GFP reporter GFP RFP system two reporter translate separately\n",
      "2225 <-- dual luciferase assay HIV frameshift region insert two reporter gene Fluc reporter whose expression require PRF HIV synthesize Rluc fusion protein\n",
      "2226 <-- Although entirely hypothetical one propose change PRF efficiency trigger ΔG mutant could cause drop short peptide result PRF GFP RFP system\n",
      "2227 <-- could cause apparent decrease PRF efficiency situation occur dual luciferase system\n",
      "2228 <-- ΔG iC mutant locate either strand rRNA helix use standard PCR procedure create inverse mutation iG ΔC\n",
      "2229 <-- iG effect PRF efficiency GFP RFP dual luciferase assay data show ΔC increase PRF efficiency respectively assay\n",
      "2230 <-- far investigate whether rRNA mutation also increase PRF efficiency combine slippery sequence\n",
      "2231 <-- iC mutation investigate C construct contains extend slippery sequence found SARS CoV C chimeric construct classic slippery sequence HIV replace one found SARS CoV whereas BC nucleotide correspond found HIV described Figure\n",
      "2232 <-- HIV frameshift region iC mutation increase PRF efficiency frameshift region compare wild type rRNA\n",
      "2233 <-- location mutation select random library show secondary tertiary structure rRNA Figure\n",
      "2234 <-- ΔG iG ΔC iC locate helix whereas G locate helix rRNA\n",
      "2235 <-- Helices part central domain rRNA region form subunit platform participates E tRNA binding see subsequently\n",
      "2236 <-- Using dual luciferase assay mammalian culture cell demonstrate mutate BCX codon plus precede base immediately upstream HIV classic heptanucleotide slippery sequence affect PRF efficiency\n",
      "2237 <-- BCX ABC codon occupy respectively E site PRF\n",
      "2238 <-- found mutate ABC nucleotide locate position upstream viral slippery sequence also change PRF efficiency\n",
      "2239 <-- Interestingly analysis eukaryotic virus slippery sequence retrieve RECODE database Table observe bias identity BCX codon locate upstream various slippery sequence may different codon locate immediately upstream give classic slippery sequence appear exclusive slippery sequence g\n",
      "2240 <-- CCU UUU codon found upstream slippery sequence U UUA AAC found upstream viral slippery sequence Table\n",
      "2241 <-- bias observe codon immediately upstream codon\n",
      "2242 <-- observation agreement finding mutate E site codon change PRF efficiency case change codon E site codon\n",
      "2243 <-- Using specialized bacterial ribosome system also show mutation locate helix rRNA platform region small ribosomal subunit increase HIV PRF efficiency Figure\n",
      "2244 <-- analysis ribosome crystal structure infer mutation could influence structure region involve binding tRNA E site Figure\n",
      "2245 <-- Indeed mutation locate helix close nucleotide loop contact anticodon stem tRNA E site\n",
      "2246 <-- interaction require kink turn motif maintain coaxial helix\n",
      "2247 <-- Therefore mutation alter PRF efficiency could related E site\n",
      "2248 <-- Interestingly mutation small subunit ribosomal protein locate proximity E site found increase spontaneous frameshift far support relationship region frameshift\n",
      "2249 <-- rRNA mutation assess different combination classic slippery sequence upstream triplet ABC position PRF efficiency increase extent\n",
      "2250 <-- Moreover platform region involve conformational change translocation\n",
      "2251 <-- rRNA mutant could alter PRF efficiency influence binding tRNA E site also interfere translocation\n",
      "2252 <-- worth mention rRNA C G mutation affect translocation increase spontaneous frameshift also support relationship translocation frameshift\n",
      "2253 <-- Also cycloheximide translocation inhibitor found increase HIV PRF efficiency fold unpublished data support involvement translocation PRF\n",
      "2254 <-- previous observation result propose refinement last model describe PRF\n",
      "2255 <-- novel model Figure sequence event lead PRF start BCX XXY codon AAU UUU HIV frameshift region occupy respectively P site\n",
      "2256 <-- peptide bond formation newly deac tRNA pept tRNA P site become engage translocation process\n",
      "2257 <-- Translocation begin binding EF G eEF associate GTP molecule ribosome\n",
      "2258 <-- GTP hydrolysis acceptor stem tRNAs move towards E P site respectively large subunit\n",
      "2259 <-- tRNAs occupy hybrid site acceptor stem tRNAs large ribosomal subunit intermediate position post pre translocational state P E P star refers transition state\n",
      "2260 <-- intermediate recently identify use single turnover rapid kinetics assay Figure\n",
      "2261 <-- tRNA anticodon stem loop move towards E P site small ribosomal subunit occupy first intermediate E E P P site\n",
      "2262 <-- next step movement E E P P site drag mRNA distance three nucleotide e\n",
      "2263 <-- one codon release EF G GDP ribosome\n",
      "2264 <-- propose fraction ribosome two tRNAs drag mRNA three nucleotide two due presence stimulatory signal resistant unwind\n",
      "2265 <-- consequence two tRNAs block intermediate E E P P site translocation incomplete Figure B\n",
      "2266 <-- hypothesis mainly relies cryo EM structure Namy et al\n",
      "2267 <-- show presence frameshift stimulatory signal stall ribosome translocation process\n",
      "2268 <-- next step translation arrival incoming aa tRNA bound EF Tu eEF associate GTP molecule\n",
      "2269 <-- incomplete translocation incoming aa tRNA occupies entry site differs standard entry site Figure C\n",
      "2270 <-- Codon anticodon interaction dynamic break form\n",
      "2271 <-- propose tRNAs locate intermediate site respective standard high affinity site ribosome prone shift standard site shift pair mRNA reading frame Figure\n",
      "2272 <-- drive force change reading frame therefore tendency tRNAs occupy standard binding site\n",
      "2273 <-- slippage tRNAs would start tRNA locate E E site follow successive slippage tRNAs locate P P site towards E E P P site see ref\n",
      "2274 <-- concept successive tRNA slippage\n",
      "2275 <-- PRF would follow aa tRNA accommodation site couple E site tRNA ejection ribosome Figure E conventional translation would resume unfold frameshift stimulatory signal\n",
      "2276 <-- model two step ribosome pause block translocation three tRNAs shift standard site\n",
      "2277 <-- mention Introduction section Weiss et al\n",
      "2278 <-- first suggest PRF link translocation model PRF occurs XXY YYZ codon occupy P site Figure B\n",
      "2279 <-- suggest translocation anomaly lead PRF occurs precede elongation cycle BCX XXY codon occupy P site take account fact E site codon participates PRF\n",
      "2280 <-- model three tRNAs participate PRF\n",
      "2281 <-- event trigger blockade ribosome transition state follow incomplete translocation due presence stimulatory signal\n",
      "2282 <-- explain identity BCX codon influence PRF\n",
      "2283 <-- well know mutate classic X XXY YYZ motif prevents slippage tRNAs locate P site since shift tRNAs pair mRNA new reading frame\n",
      "2284 <-- However whether codon anticodon interaction E site still matter debate\n",
      "2285 <-- analysis crystal structure codon anticodon interaction observe mRNA deac tRNA E site\n",
      "2286 <-- contrast vitro translational experiment support codon anticodon interaction E site\n",
      "2287 <-- slippery sequence analysis Table observe PRF one standard base pair possible tRNA ABC nucleotide case\n",
      "2288 <-- codon anticodon interaction E site could suggest base pair Watson Crick G U wobble pair tolerate E site\n",
      "2289 <-- relaxed condition major consequence mutate BCX codon plus precede base likely alter PRF efficiency influence codon anticodon interaction E site tRNA shift frame although minor effect exclude\n",
      "2290 <-- propose main consequence make change modify identity E site tRNA\n",
      "2291 <-- hypothesis relationship structural peculiarity tRNA capacity shift intermediate classic site since movement tRNA precedes thus control movement two tRNAs\n",
      "2292 <-- group Rousset propose PRF E site tRNA carry pseudouridine modification position couple high PRF efficiency\n",
      "2293 <-- However result could reproduce Rousset J P personal communication\n",
      "2294 <-- Also search use tRNA Compilation database present result support relationship presence tRNA modification position PRF efficiency data show\n",
      "2295 <-- conclusion propose novel model describe PRF three tRNAs involve\n",
      "2296 <-- detailed understand PRF mechanism contribute development novel anti frameshift agent affect replication capacity virus HIV\n"
     ]
    }
   ],
   "source": [
    "for nr, setentence in enumerate(corpus_sentences):\n",
    "    if sentence_dict[nr] == 11:\n",
    "        print(str(nr)+' <--', setentence)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Words embeddings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.feature_extraction.text import CountVectorizer\n",
    "count_vectorizer = CountVectorizer()\n",
    "word_count_vector = count_vectorizer.fit_transform(corpus)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Vocabulary:\n",
      " ['aa' 'aaa' 'aaaccatggtactgcgatc' 'aac' 'aag' 'aat' 'aatacggtccctact'\n",
      " 'aau' 'ab' 'abacus' 'abandon' 'abasic' 'abbreviate' 'abc' 'abcc'\n",
      " 'aberrant' 'abi' 'ability' 'ablation' 'able' 'abnormal' 'abolish'\n",
      " 'abrogate' 'abruptly' 'absence' 'absent' 'absolute' 'absolutely' 'absorb'\n",
      " 'absorbance' 'absorbed' 'absorption' 'abstract' 'abundance' 'abundant'\n",
      " 'aca' 'acaacacgtttagcctgacc' 'acc' 'accelerate' 'accept' 'accepted'\n",
      " 'acceptor' 'access' 'accessibility' 'accessible' 'accession' 'accessory'\n",
      " 'accommodate' 'accommodation' 'accompany' 'accomplish' 'accomplishment'\n",
      " 'accord' 'accordance' 'accordingly' 'account' 'accounting' 'accumulate'\n",
      " 'accumulates' 'accumulating' 'accumulation' 'accuracy' 'accurate'\n",
      " 'accurately' 'ace' 'acetate' 'acetic' 'acetone' 'acetonitrile'\n",
      " 'acggacatttcg' 'achievable' 'achieve' 'achievement' 'achieving' 'acid'\n",
      " 'acidic' 'acon' 'acp' 'acquire' 'acquisition' 'across' 'acrylamide' 'act'\n",
      " 'action' 'activate' 'activation' 'activator' 'active' 'actively'\n",
      " 'activity' 'acttaatttgattccat' 'actual' 'actually' 'actuate' 'acute' 'ad'\n",
      " 'adapt' 'adapter' 'adaptive']\n",
      "\n",
      "Shape of vocabulary: (6370,)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Extract the vocabulary and display it\n",
    "vocabulary = np.array(count_vectorizer.get_feature_names())\n",
    "print(\"\\nVocabulary:\\n\", vocabulary[:99])\n",
    "print()\n",
    "print('Shape of vocabulary:',vocabulary.shape)\n",
    "print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Full vector: \n",
      "[[0 0 0 ... 0 0 0]\n",
      " [0 1 0 ... 0 0 0]\n",
      " [0 0 0 ... 0 0 0]\n",
      " ...\n",
      " [0 0 0 ... 0 0 0]\n",
      " [0 0 0 ... 0 0 0]\n",
      " [0 0 0 ... 0 0 0]]\n",
      "Shape of vector: (25, 6370)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Create a vector by passing the text into the vectorizer to get back counts\n",
    "word_count_vector\n",
    "\n",
    "# Our final vector:\n",
    "print('Full vector: ')\n",
    "print(word_count_vector.toarray())\n",
    "print('Shape of vector:',word_count_vector.toarray().shape)\n",
    "print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([10,  1,  0,  3,  0,  0,  0,  5,  0,  0,  0,  0,  0,  5,  0,  0,  0,\n",
       "        0,  0,  0,  0,  0,  0,  0,  0,  0,  1,  0,  0,  1,  0,  0,  0,  0,\n",
       "        0,  0,  0,  0,  0,  0,  0,  3,  0,  0,  0,  0,  0,  0,  5,  0,  0,\n",
       "        0,  0,  0,  0,  3,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,\n",
       "        0,  1,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  1,  0,  0,\n",
       "        0,  0,  0,  0,  3,  0,  0,  0,  0,  1,  0,  0,  0,  0,  0,  5,  1,\n",
       "        0,  0,  0,  0,  1,  0,  0,  1,  0,  0,  0,  1,  0,  0,  0,  0,  0,\n",
       "        0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  8,  2,\n",
       "        0,  1,  0,  0,  0,  1,  0,  0,  0,  1,  0,  0,  0,  0,  0,  0,  0,\n",
       "        0,  0,  1,  0,  0,  0,  0,  0,  0,  0,  1,  0,  0,  0,  0, 12,  0,\n",
       "        0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,\n",
       "        0,  0,  0,  0,  1,  1,  0,  0,  0,  1,  0,  0,  0, 17,  0,  7,  0,\n",
       "        1,  0,  0,  0,  2,  3,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,\n",
       "        0,  0,  0,  0,  0,  1,  0,  0,  0,  0,  0,  0,  0,  0,  1,  0,  0,\n",
       "        1,  0,  0,  0,  2,  1,  1,  0,  0,  0,  0,  2,  1,  6,  0,  0,  0,\n",
       "        0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,\n",
       "        0,  0,  0,  0,  1,  0,  0,  0,  0,  0,  1,  0,  0,  0, 12,  0,  0,\n",
       "        0,  0,  0,  0,  1,  0,  0,  0,  0,  0,  0,  1,  0,  0,  0,  0,  0,\n",
       "        0,  0,  1,  0,  1,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  1,\n",
       "        0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  2,  0,  0,  0,  0,  0,\n",
       "        0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,\n",
       "        1,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,\n",
       "        0,  0,  1,  0,  0, 18,  1,  0,  0,  0,  2,  0,  1,  0,  2,  0,  0,\n",
       "        0,  0,  0,  0,  0,  0,  0,  1,  0,  0,  0,  0,  0,  0,  0,  0,  0,\n",
       "        0,  0,  0,  0,  1,  0,  0,  0,  0,  0,  0,  0,  0,  1,  1,  1,  0,\n",
       "        0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,\n",
       "        0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,\n",
       "        0,  0,  0,  0,  0,  0,  0,  0,  7,  0,  0,  0,  0,  0,  1,  0,  3,\n",
       "        0,  0,  0,  0,  0,  0,  0,  0,  6,  1,  0,  0,  0,  0,  0,  0,  0,\n",
       "       14,  0,  0,  0, 10,  0,  0,  0,  0,  0,  0,  1,  0,  0,  3,  0,  0])"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "word_count_vector.toarray()[11][0:510]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Word \" virus \" appears 12 times in the doc 11\n"
     ]
    }
   ],
   "source": [
    "word = 'virus'\n",
    "word_index = np.where(vocabulary==word)[0][0]\n",
    "doc_nr = 11\n",
    "print('Word \\\"',word,'\\\" appears',word_count_vector.toarray()[doc_nr][word_index],'times in the doc',doc_nr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ 9,  1, 40,  0, 39,  0,  0,  6,  9,  5, 27, 12,  1, 15,  1,  2,  1,\n",
       "        0, 30,  2,  0,  0,  6,  4,  0])"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Count how many times word \"virus\" appears in the documents\n",
    "word_count_vector.toarray()[:,word_index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PMC1635287.xml.json\n",
      "\n",
      "RNA viruses, including coronaviruses, encode a group of structural phosphoproteins with basic patches on their surface. As a pivotal structural component of the virion, this group of proteins plays essential roles in packaging the RNA genome to form a ribonucleoprotein (RNP) complex resulting from assembly of the viral RNA and multiple copies of the (nucleo)capsid protein. A detailed dissection and characterization of the intrinsic RNA-binding properties of these proteins are thus essential for understanding several important processes in the life cycle of RNA viruses, including assembly of the nucleocapsid, the specific encapsidation of viral RNA and morphogenesis of virions. In this study, amino acid residues in the N-terminal domain of the nucleocapsid (N) protein of a coronavirus that are critical for its RNA-binding activity are identified and their roles in viral infectivity are analyzed, using site-directed mutagenesis based on our previous structural studies. Coronaviruses are enveloped viruses containing the largest known single-stranded, positive-sense RNA genome of ∼30 kb (1). During the coronavirus life cycle, the N protein is synthesized in large amounts and is thought to play an important role by specifically packaging the viral genome into a filamentous nucleocapsid of ∼10 to 15 nm in diameter and several 100 nm in length, a macromolecular structure that is visible by using electron microscopy (2). In addition to its structural role, the N protein also participates in viral RNA transcription, replication and in modulating the metabolism of host cells (3–13). Using X-ray crystallography techniques, we recently determined a 3D structure of the N protein from coronavirus infectious bronchitis virus (IBV), a prototype coronavirus, and showed that the basic building block for nucleocapsid assembly was a dimer of the N protein (14). The 409 amino acids long IBV N protein is composed of two globular domains that are resistant to proteolysis. Amino acids 29 to 160 form the N-terminal RNA binding domain and residues 218 to 329 form its C-terminal dimerization domain (5,14). Even though the structure of the N-terminal domain was determined in the absence of a nucleic acid ligand, its overall shape and the distribution of electrostatic charges suggest a plausible model for RNA binding. The outer surface of the N-terminal domain is enriched in aromatic and basic residues, a common feature for RNA-binding proteins (15). The shape of this domain is reminiscent of a hand having basic fingers, a hydrophobic palm and an acidic ‘wrist’. The positively charged ‘fingers-like’ β-hairpin extension could neutralize the phosphate groups emanating from RNA, whilst the base moieties could make contact with exposed aromatic residues from the hydrophobic palm (14). Based on this structural hypothesis, we chose to mutate a number of evolutionarily conserved residues in the N-terminal domain of the IBV N protein in order to assess their roles in viral genomic RNA binding and in viral replication. We used purified recombinant wild type and mutant proteins containing the N-terminal domain expressed in Escherichia coli for RNA binding assays. These studies led to the identification of a number of amino acid residues essential for the RNA-binding activity of the domain. Subsequently, we introduced mutations that either severely or mildly impair the RNA binding activities of the N-terminal domain into an infectious cDNA clone system derived from the genomic RNA of IBV to assess their effects on viral replication and infectivity. We identified several residues exposed to the solvent whose substitutions to alanine yield a decrease of RNA binding and a concomitant reduction in virus replication. Interestingly, residues Arg-76 and Tyr-94, located at the base of the positively charged flexible hairpin loop and on the hydrophobic platform, respectively, were critical for RNA binding and viral infectivity. His-tagged wild type and mutant N-terminal domains of IBV N protein were expressed in E.coli BL-21 by induction with 1 mM isopropyl-β-d-thiogalactopyranoside (IPTG). Cells were lysed by sonication and purified by metal affinity chromatography with Protino-Ni 150 kit (Macherey Nagel). PCR fragments covering the IBV genome from 27 100 to 27 608 nt were cloned into a plasmid in either forward or reverse orientation under the control of a T7 promoter. The Dig-labeled sense (+) and anti-sense (−) RNA probes were made in vitro using the DIG RNA labeling kit according to the instructions of the manufacturer (Roche). Five micrograms of purified proteins were resolved on an SDS-15% polyacrylamide gel and transferred onto nitrocellulose membranes (Hybond C-Extra, Amersham Biosciences) using a semi-dry transfer apparatus. Membranes were washed for 10 min with the probe buffer [1× Denhardt's Reagent, 1 mM EDTA, 10 mM Tris–HCl (pH 7.5) and 50 mM NaCl], blocked for 1 h with 25 µg/ml yeast tRNA (Ambion) and subsequently incubated with 10 µg of DIG-labeled RNA probe in the same probe buffer for 1 h. Membranes were washed three times for 15 min each with the probe buffer, before proceeding to detection with CDP-Star (Roche) according to the manufacturer's instructions. Construction of an infectious IBV clone was carried out essentially as described (16,17). Briefly, five fragments spanning the entire IBV genome were obtained by RT–PCR from Vero cells infected with the Vero cell-adapted IBV p65. The PCR products were purified from agarose gels and cloned into pCR-XL-TOPO (Invitrogen) or pGEM-T Easy (Promega) vectors. Subsequently, fragment A was removed from pCR-XL-TOPO by digestion with NheI and EcoRI, and subcloned into pKT0 vector. Plasmids were digested with either BsmBI (fragment A) or BsaI (fragments B, C, D and E). The digested plasmids were separated on 0.8% agarose gels containing crystal violet. Bands corresponding to each of the fragments were cut from the gels and purified with QIAquick gel extraction kit (QIAGEN Inc.). Fragments A and B, and fragments C, D and E were first ligated with T4 DNA ligase at 4°C overnight. The two reaction mixtures were then mixed and further ligated at 4°C overnight. The final ligation products were extracted with phenol/chloroform/isoamyl alcohol (25:24:1), precipitated with ethanol and detected by electrophoresis on 0.4% agarose gels. Full-length transcripts were generated in vitro using the mMessage mMachine T7 kit (Ambion, Austin, TX). The N transcripts were generated by using a linearized pKT0-IBVN containing IBV N gene and the 3′-untranslated region (3′-UTR) as templates. The in vitro synthesized full-length and N transcripts were treated with DNase I and purified with phenol/chloroform. Vero cells were grown to 90% confluence, trypsinized, washed twice with cold phosphate-buffered saline (PBS), and resuspended in PBS. RNA transcripts were added to 400 µl of Vero cell suspension in an electroporation cuvette, and electroporated with one pulse at 450 V, 50 µF with a Bio-Rad Gene Pulser II electroporator. The transfected Vero cells were cultured overnight in 1% FBS-containing MEM in a 60 mm dish or a 6-well plate and further cultured in MEM without FBS. Mutations were introduced into the corresponding fragments by using QuickChange site-directed mutagenesis kit (Stratagene), and confirmed by sequencing of the whole fragments. Confluent monolayers of Vero cells on 6-well plates were infected with wild-type and mutant viruses at a multiplicity of ∼1 PFU/cell. After 1 h of incubation at 37°C, cells were washed twice with PBS and cultured in 3 ml of MEM containing 0.5% carboxymethy cellulose for 3 days. The cells were fixed and stained with 0.1% toluidine. Vero cells were infected with wild-type and recombinant IBV, and harvested at different times post-infection. Viral stocks were prepared by freezing/thawing of the cells three times. The plaque-forming units per ml of each sample were determined by infecting Vero cells on 6-well plates in duplicate with 10-fold serial dilution of each viral stock. Vero cells were infected with wild type and Q74A mutant virus at a multiplicity of ∼1 PFU/cell. Total RNA was extracted from cells infected with wild type or mutant viruses. Ten micrograms of RNA were added to a mixture of 1× MOPS, 37% formaldehyde and formamide and incubated at 65°C for 20 min before subjected to gel electrophoresis. The segregated RNA bands were transferred onto a Hybond N+ membrane (Amersham Biosciences) via capillary action overnight and fixed by ultraviolet (UV) crosslinking (Stratalinker). Hybridization of Dig-labeled DNA probes was carried out at 50°C in hybridization oven overnight. Membranes were washed three times for 15 min each with the probe buffer, before proceeding to detection with CDP-Star (Roche) according to the manufacturer's instructions. Vero cells were infected with wild type and Q74A mutant virus at a multiplicity of ∼1 PFU/cell. Total proteins extracted from Vero cells were lysed with 2× SDS loading buffer in the presence of 200 mM DTT plus 10 mM of iodoacetamide and subjected to SDS–PAGE. Proteins were transferred to PVDF membrane (Stratagene) and blocked overnight at 4°C in blocking buffer (5% fat free milk powder in PBST buffer). The membrane was incubated with 1:2000 diluted primary antibodies in blocking buffer for 2 h at room temperature. After washing three times with PBST, the membrane was incubated with 1:2000 diluted anti-mouse or anti-rabbit IgG antibodies conjugated with horseradish peroxidase (HRP) (DAKO) in blocking buffer for 1 h at room temperature. After washing three times with PBST, the polypeptides were detected with a chemiluminescence detection kit (ECL, Amersham Biosciences) according to the manufacturer's instructions. PCR products covering the IBV sequence from 25 957–26 352 nt were amplified by using the forward primer 5′-CGGGCATATGTCTTCTGGAAATGCATCTTGG-3′ and the reverse primer 5′-CGGGATCCTTACAGGGGAATGAAGTCCCAAC-3′. The PCR fragments were digested with NdeI and BamHI, and ligated into NdeI- and BamHI-digested pET-16b (Novagen). Each mutation was introduced by two-round PCR and the mutation introduced was confirmed by automated nucleotide sequencing. The N-terminal RNA-binding domain of N protein from both IBV and SARS-CoV contains a flexible and positively charged hairpin loop that extends much beyond the protein core and that could grasp a nucleic acid substrate by neutralizing its phosphate groups (Figure 1) (14,18). Positively charged lysine and arginine residues in the N-terminal domain of SARS-CoV N protein were proposed to bind a 32 nt stem–loop structure located at the 3′ end of the SARS-CoV RNA genome (18). Indeed, the structure of this domain possesses some features reminiscent of RNA-binding proteins sharing the RNP fold (19–22). These RNA-binding proteins, including the U1A spliceosomal protein (23) and the coat protein from MS2 bacteriophage (24), bind the RNA ligand with residues emanating from the surface of a four-stranded anti-parallel β sheet. Presumably, strands β2 and β3 as well as the flexible β-hairpin from the IBV nucleocapsid protein could fulfill a similar role by interacting with phosphate groups from an RNA ligand (Figure 1). The β3 strand is the longest in the N-terminal domain of IBV N protein and several evolutionarily conserved residues project from it (Figure 1). Nucleotide bases in single-stranded RNA have a tendency to stack either on adjacent bases or with aromatic side chains of the protein (20). The above structural model was partially supported by the demonstration that the N-terminal domain of IBV N protein could interact with RNA fragments corresponding to the 3′ end of the viral genome (14). To more vigorously test this model, systematic mutagenesis of residues essential for the RNA-binding activity was carried out. We targeted the exposed hydrophobic residues Tyr-92 and Tyr-94 (strand β3) that could form stacking interactions with the nucleotide bases as well as several basic or polar residues emanating from the β-hairpin (Figure 2a). As summarized in Figure 2a, 10 amino acids were mutated to alanine. These residues are either exposed to the solvent (Gln-74, Arg-76, Lys-78, Lys-81, Tyr-92 and Tyr-94) or have been conserved during evolution (Tyr-70 and Arg-73) and are thus likely to be involved in RNA binding or to play an undefined functional role. In addition, mutation of Asp-111 to an alanine was included as an additional control for the RNA-binding assay (Figure 2a). The wild type and mutant constructs, spanning 29 to 160 amino acids with a hexa-histidine tag at their N-termini, were cloned into a bacterial expression vector, expressed in E.coli and purified to near homogeneity. Approximately equal amounts of the purified proteins were separated on SDS–PAGE (Figure 2b, upper panels). After transfer to a Hybond C extra membrane, the RNA-binding activities of wild type and mutant proteins were analyzed by northwestern blot with two probes corresponding to the positive- and negative-sense 3′-UTR of IBV, respectively. A moderate decrease in binding to both the positive and negative sense RNA probes was consistently observed with most mutant proteins, and representative gels are shown in Figure 2b (Figure 2b, middle and lower panels). Among all mutants, mutation of Tyr-94 to an alanine residue showed the most significant reduction in binding of the domain to both positive and negative probes. The RNA binding activities of this mutant to positive and negative probes were reduced to 27 and 25%, respectively, of the wild type were observed (Figure 2b, lane 9). Relatively less reduction in the RNA-binding activity was observed for the R76A, K78A, K81A and Y92A mutant proteins which retain between 50–74% of the wild-type binding activity (Figure 2b, lanes 5–8). With the exception of the Q74A and D111A mutants, which displayed a slight increase in the RNA binding activity to both the positive- and negative-sense RNAs, (Figure 2b, lanes 4 and 10), all other mutants showed a certain degree of reduction in their binding activity to either RNA probes (Figure 2b). As a control, a circular dichroism analysis of the two corresponding R76A and Y94A single mutants was performed to check whether the mutant proteins were properly folded. The spectra obtained for these two mutants were similar to the wild-type protein, ruling out any drastic conformational change introduced by the mutations (data not shown). As none of the single mutations introduced could totally abrogate the RNA-binding activity of the IBV N-terminal domain, two mutant constructs carrying double mutations were made. As can be seen in Figure 2b (lanes 13 and 14), mutations of both Arg-73 and Lys-78 to alanine residues (R73/K78A) reduced the RNA-binding activities to the positive- and negative-sense RNA probes between 44 and 10% of wild-type, respectively. Mutations of both Arg-76 and Tyr-94 (R76/Y94A) to alanine reduced the RNA-binding to the positive- and negative-sense RNA probes to only 7 and 19% of wild-type, respectively (Figure 2b, lanes 15 and 16). Taken together, these results are consistent with the presence of an extended RNA binding site covering a large part of the accessible surface of the N-terminal domain of the coronavirus N protein. They are also in agreement with the assumption that removal of a single positive charge at the protein surface by mutation of a lysine or arginine residue to an alanine is insufficient to totally disrupt the formation of the protein–RNA complex. The R76A and Y94A mutations were first introduced into an infectious clone of IBV, and in vitro transcribed full-length RNA molecules derived from wild type and mutant constructs were generated by in vitro transcription using the T7 RNA polymerase in the presence of a cap analog. Electroporation of wild-type RNA transcripts together with an RNA fragment covering the N protein region into Vero cells showed the formation of massive CPE at 2 days post-electroporation (Figure 3, panel rIBV+N). In contrast, no CPE formation was observed in cells transfected with transcripts derived from either mutant up to 5 days post-electroporation, suggesting that no infectious virus could be rescued from the two mutant transcripts (Figure 3, panels R76A+N and Y94A+N). These results indicate that R76A and Y94A mutations, which severely reduce the RNA-binding activity of the N-terminal domain of the N protein, may abolish the infectivity of the in vitro synthesized full-length IBV transcripts. Introduction of Q74A and Y92A mutations into the full-length infectious clone of IBV was then carried out. Typical CPEs were observed in cells transfected with Q74A mutant transcripts at 3 days post-electroporation and recombinant viruses were recovered (Figure 3, panel Q74A). In cells electroporated with Y92A transcripts, a typical CPE was observed at 2 days post-electroporation [Figure 3, panel Y92A(1)]. Incubation of the transfected cells for one more day saw the appearance of more CPEs [Figure 3, panel Y92A(2)]. However, prolonged incubation of the transfected cells up to 6 days saw no spread of the infection from the two initiated infected loci to neighboring cells. Re-infection of fresh Vero cells with either medium collected from these cells or total cell lysates prepared by freezing and thawing of the transfected cells showed no detection of the CPE formation, indicating that the minute amount of viruses rescued from the initially transfected cells could not maintain infectivity in subsequent passage. The R76A and Y94A mutant N constructs were then used to test if they could facilitate the rescue of the in vitro synthesized full-length IBV transcripts in trans. As mentioned, efficient recovery of infectious virus from the in vitro transcribed full-length IBV RNA requires co-transfection of these in vitro transcripts together with the in vitro synthesized RNA covering the N gene. Electroporation of wild-type full-length IBV transcripts together with RNAs synthesized in vitro from wild-type and the two mutant N constructs, respectively, showed efficient recovery of the infectious virus with CPE observed on almost the whole monolayer at 3 days post-electroporation (Figure 3, panels rIBV+N-R76A and rIBV+N-Y94A). In contrast, cells transfected with the full-length IBV transcripts only (without co-electroporation with the N transcripts) showed the formation of a single CPE on the whole monolayer at 3 days post-electroporation (Figure 3, panel rIBV-N). As no infectious virus was recovered from cells transfected with R76A and Y94A mutant transcripts, total RNA was extracted from cells electroporated with wild-type and mutant full-length transcripts and RT–PCR amplification of subgenomic mRNAs was carried out to check whether a low level of RNA replication and subgenomic mRNA transcription occurred in cells transfected with the mutant transcripts. The forward primer used in this reaction corresponds to the leader sequence from 26–46 nt in the genomic RNA and the downstream primers covers IBV sequences from 24 784 to 24 803 nt. If transcription of subgenomic mRNAs did occur, a 415 bp PCR product corresponding to the 5′-terminal region of the subgenomic mRNA4 and a 1010 bp fragment corresponding to the 5′-terminal region of the subgenomic mRNA3 would be expected. As shown in Figure 4a, a dominant 415 bp band and a weak 1010 bp band were observed in cells electroporated with wild-type full-length transcripts at 2 days post-electroporation (lane 2). Sequencing of the PCR fragments confirmed that they represent the correct sequences of the corresponding regions of the subgenomic mRNAs 3 and 4, respectively. The same PCR products were not detected in cells electroporated with the R76A and Y94A mutant transcripts, respectively, at 24, 48 and 72 h post-electroporation (Figure 4a, lanes 3–8). In cells electroporated with Q74A mutants at 3 days post-electroporation, RT–PCR amplification of the subgenomic mRNA 3 and 4 showed the detection of the 415 and 1010 bp bands (Figure 4a, lane 9). The same PCR products were not evident in cells electroporated with the Y92A mutant transcripts (Figure 4a, lane 10).\n",
      " RT–PCR amplification of the negative strand RNA was performed to check if RNA replication occurred in these transfected cells. The primer pair was chosen so that the IBV sequence from 14 931 to 15 600 nt would be amplified by the RT–PCR. If replication of viral RNA occurred, a 670 bp PCR fragment would be expected. As shown in Figure 4b, RT–PCR fragments amplified from both positive (lanes 4, 7 and 9) and negative (lanes 3, 6 and 10) strand RNA templates were obtained from cells transfected with the mutant transcripts. The amount of the negative strand RNA was approximately half of that of the positive strand RNA (Figure 4b). As positive control, the same RT–PCR fragment for the negative strand RNA was observed in cells transfected with wild-type transcripts (Figure 4b, lanes 8 and 11). In the negative controls, total RNA extracted from cells transfected with Y94A (Figure 4b, lane 2) and R76A (Figure 4b, lane 5) was analyzed by PCR directly using the primer set for the negative strand RNA. No corresponding PCR fragment was detected from these samples, demonstrating that the detection of the negative strand RNA in cells transfected with wild-type and mutant constructs by RT–PCR is due to the replication of viral RNA. These results confirm that transcription of the negative strand RNA has taken place in cells transfected with the mutant transcripts. The transcription of negative strand RNA in cells electroporated with wild-type, R76A, Y92A and Y94A transcripts was then quantitatively determined and compared by real time RT–PCR at 24 and 48 h post-electroporation, respectively. Compared to cells transfected with wild-type transcripts, the relative amounts of negative strand RNA in cells transfected with R76A, Y92A and Y94A were 48, 46 and 50%, respectively, at 24 h post-electroporation (Figure 4c, lanes 1–4). These figures were dropped to 41, 45 and 23%, respectively, at 48 h post-electroporation (Figure 4c, lanes 5–6). Taken together, these results demonstrate that transcription of the negative strand RNA has taken place in cells transfected with the mutant transcripts. As the level of negative strand RNA was not increased in cells transfected with the mutant transcripts over a time-course experiment, it suggests that transcription of these RNA species may occur only in the initially transfected cells using the in vitro transcripts as templates. The growth properties of the Q74A mutant virus on Vero cells were tested by analysis of plaque sizes and growth curves of passage 3 mutant virus. Compared to cells infected with wild-type recombinant virus (rIBV), slightly smaller sized plaques were observed in cells infected with the Q74A mutant virus (Figure 5a). Analysis of the growth curves of wild-type and the mutant virus demonstrated that the mutant virus grew slightly more slowly than the wild type recombinant virus. When 10-fold more mutant virus was used, 2- to 10-fold more mutant virus was produced at 8–24 h post-infection (Figure 5a). The titers of the mutant virus dropped more rapidly than wild-type virus at 36 and 48 h post-infection (Figure 5a).\n",
      " Further characterization of the mutant virus was subsequently carried out by analysis of the genomic and subgenomic RNA, and the expression of viral structural proteins S and N. Northern blot analysis of cells infected with the 5th passage of rIBV and Q74A mutant viruses showed that similar amounts of genomic and subgenomic RNAs were detected (Figure 5b). Quantitative comparison of the subgenomic RNA synthesis by densitometry analysis of the bands shown in Figure 5b demonstrated that in cells infected with Q74A mutant virus (p5), the relative amounts of mRNA 2, 3, 4, 5 and 6 are ∼98, 101, 121, 104, 106 and 101%, respectively, of those in cells infected with rIBV. However, Western blot analysis of cells infected with rIBV and Q74A mutant virus showed that more S and N proteins were detected in cells infected with wild-type virus (Figure 5c, lanes 1–14). To confirm if the Q74A mutation is genetically stable during passage of the mutant virus in cells, viral RNA was isolated from cells infected with passage 2 and 5 virus and amplified by RT–PCR. Sequencing analysis showed that the two mutated nucleotides (CA-GC) remained stable (Figure 5d). Furthermore, no compensatory mutation was found in other region of the N protein, confirming that Q74A mutant virus is stable in cell culture system. The intrinsic RNA-binding activity of the nucleocapsid protein of RNA viruses is essential for packaging the RNA genome into a RNP complex. In previous structural studies of the N protein from two coronaviruses, the N-terminal domain was shown to be resistant to proteolysis and to fold independently (14,15). It serves as a functional unit critical for the RNA-binding activity of the protein (14). Based on these structural results, the importance of individual residues in RNA-binding was studied. Here we showed that Tyr-94 in strand β3 binds to the RNA ligand. The positively charged Arg-76 is located in the immediate vicinity of this aromatic residue at the base of the extended flexible hairpin loop, within a cluster of positively charged residues. Thus, the interaction between Arg-76 and Tyr-94 with RNA appears to play a crucial role during virus assembly. Since no single mutation could totally disrupt nucleic acid binding, other aromatic or basic residues located close to these two residues at the surface of the N-terminal domain are likely to contribute to RNA binding by creating a broad surface that makes contact with the genomic RNA. Using a simple docking experiment, the RNA binding site seems capable of accommodating up to 8–9 nt bases per N-terminal domain monomer. The Arg-76 and Tyr-94 residues identified in this paper thus provide more precise mapping of the location of the RNA-binding site on the IBV N protein. It is interesting to note that the main effect of these mutations on viral life cycles was on subgenomic RNA transcription. However, the replication of negative strand RNA is nearly normal in cells transfected with these mutant constructs. Is it possible that these mutations will directly impair an undefined function of the N protein that is essential for subgenomic RNA transcription? Two possibilities were considered. First, the coronavirus N protein is directly involved in subgenomic RNA replication. The MHV N protein was shown to be involved in coronaviral RNA transcription (4), but the underlying mechanism was unrevealed. Alternatively, the observed phenotypic changes in subgenomic RNA replication caused by these mutations may be due to a generally reduced viral replication rate and infectivity. In fact, several other mutations that affect viral replication cycles were found to have defects in subgenomic RNA transcription (25–27). In this study, we have repeatedly observed that detection of subgenomic RNA replication in cells infected with low multiplicity of infectivity of IBV by RT–PCR is technically challenging. Positive results can be obtained only if transcription of subgenomic RNA reaches a certainly high level. When in vitro transcripts containing Y92A mutation were introduced into cells, one or two CPEs were always observable in a monolayer, suggesting the presence of low level viral RNA replication. However, detection of subgenomic RNA was consistently unsuccessful in cells electroporated with this mutant transcript. Coronavirus N protein is a multi-functional protein. The protein was shown to be able to facilitate rescue of infectious virus from several coronaviruses by co-electroporation of the in vitro synthesized N transcripts with the in vitro synthesized full-length transcripts (17,28–30). The mechanism for this trans-effect is currently unknown. In this study, the R76A and Y94A mutant N constructs could promote rescue of the infectivity of the in vitro synthesized IBV RNA as efficiently as did the wild type N transcripts, suggesting that the reduced RNA-binding activity in the N-terminal domain does not affect this function of the N protein. It also suggests that RNA-binding per se is not a major factor that contributes to this important function of the N protein. Are the reduced replication rate and loss of the infectivity of the mutant transcripts caused by the effects of these mutations on the functionality of the N protein or on the integrity of the genomic RNA itself? The nice correlation between the reduced RNA-binding activity of the N-terminal domain from the mutant proteins and the loss of infectivity of the corresponding mutant transcripts would support that the failure to rescue the viral infectivity is due to the specific mutation that impairs the functionality of the N protein. In addition, in the cases of Q74A and R76A mutants, the two mutations are physically adjacent in the genome, but yet totally different phenotypes were observed from the two mutants. It would support that the failure to recover infectious viruses from the mutant transcripts is due to the fact that these mutations reduce/impair the RNA-binding activity of the N protein. If this were the case, however, supplement of wild type N protein in trans would rescue the infectivity of the mutant transcripts. Two experiments were conducted to test this possibility, but none of them were successful. First, co-transfection of the mutant full-length transcripts with wild-type N transcripts into cells failed to recover infectious virus from the mutant transcripts, suggesting that wild-type N RNA cannot rescue the infectivity of the mutant transcripts in trans. The second experiment was to introduce the mutant full-length transcripts into cells stably expressing a six-His-tagged IBV N protein. Once again, it failed to rescue the infectivity of the mutant transcripts. At present, it is unclear if this was due to the low-level expression of N protein in these cells or because the N protein expressed in these cells contains the six-His tag at its C-terminus. Residues Arg-76 and Tyr-94 in the IBV N protein are well conserved across the whole family of coronaviruses. Structurally, these two residues may correspond to the Arg-94 and Tyr-112 residues in the SARS-CoV N protein (see Figure 1a). It is possible that introduction of similar alanine mutations into the SARS-CoV genome may yield a similar phenotype observed here for IBV. The fact that substitutions of Gln-74, Arg-76, Tyr-72 and Tyr-94 in the IBV genome by using an infectious clone of IBV showed nice correlation between the RNA-binding activity and viral infectivity would suggest potential strategies for intervention of viral replication. A compound which could specifically bind to this region of the N protein might interfere with RNA binding and thus inhibit virus replication. In addition, this study points to a number of residues on the surface of the N protein that contribute less to the RNA-binding activity and viral infectivity for the design of attenuated coronavirus strains as potential vaccine candidates. \n"
     ]
    }
   ],
   "source": [
    "# Doc nr 2 (we start counting from \"0\"!) is full of \"viruses\" ;-) Check yourself\n",
    "print(docs_list[2]['title'])\n",
    "print()\n",
    "print(docs_list[2]['content'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Cosine similarity\n",
    "- [What is a cosine similarity matrix?](https://medium.com/acing-ai/what-is-cosine-similarity-matrix-f0819e674ad1)\n",
    "- [Machine Learning :: Cosine Similarity for Vector Space Models (Part III)](http://blog.christianperone.com/2013/09/machine-learning-cosine-similarity-for-vector-space-models-part-iii/)\n",
    "- [Plot of cos x](https://pl.wikipedia.org/wiki/Funkcje_trygonometryczne#/media/Plik:Cos_proportional.svg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0.15094335],\n",
       "       [0.15315763],\n",
       "       [1.        ],\n",
       "       [0.153413  ],\n",
       "       [0.24265409],\n",
       "       [0.15069377],\n",
       "       [0.21260077],\n",
       "       [0.16038297],\n",
       "       [0.16880613],\n",
       "       [0.25720911],\n",
       "       [0.13794949],\n",
       "       [0.21928365],\n",
       "       [0.29726208],\n",
       "       [0.20027931],\n",
       "       [0.30528945],\n",
       "       [0.21981989],\n",
       "       [0.18471829],\n",
       "       [0.16147375],\n",
       "       [0.49377355],\n",
       "       [0.21076306],\n",
       "       [0.28578879],\n",
       "       [0.47563728],\n",
       "       [0.20259825],\n",
       "       [0.18226817],\n",
       "       [0.19567734]])"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Let's find docs similar to the doc no. 2\n",
    "from sklearn.metrics.pairwise import cosine_similarity\n",
    "import heapq\n",
    "\n",
    "cosine_similarity(word_count_vector,word_count_vector[2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYUAAAEWCAYAAACJ0YulAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjMsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+AADFEAAAgAElEQVR4nO3dd3hUVfrA8e+bQAg1BAi9VwELJYJgQ0VRBBu6yipi+enqWlZdV111BXVtqIAFewFdBF07gqKsCqKCgDSpBqRJC5CEkp55f3/MJA4hZcqdksz7eZ48mbn3nnPfyUzmvfece88RVcUYY4wBiIt0AMYYY6KHJQVjjDElLCkYY4wpYUnBGGNMCUsKxhhjSlhSMMYYU8KSgimXiNwrIq8FWPZyEfnS67mKSGeH4jpZRNY5UM9VIjLfiZhinYiMFZH/hHF/9t6FiCWFKCIim0RkcKllEfvwq+qjqvp/AZadqqpnOR2Tp+7vVLVbKOqOViJSS0ReF5HNInJARJaKyDkVbB+yz42IDBKRbaGo20SeJQUT1USkRqRjiBI1gK3AqUAS8C/gPRFpH8GYQsLe88iypFDFiEh3EflWRDJFZJWInOdZ3sGzLM7z/DUR2e1V7j8icls5dd4tIr97jkDXicgZnuUlTQIi0t7TBHS1iGwVkQwRuUFEjheRFZ59P+9VZ7lHqiJyrudId7+nrrFe64r3c62IbAG+LqP8YUeq5cVfRrnGIvKpZ78/AZ1KrR8oIotEJMvze6DXukYi8qaIbPe89o/L2cdVIjJfRJ7ybPeb9xG9iLT0xLBPRNJE5Lqy6ilNVQ+p6lhV3aSqLlX9DPgN6FtGDN2Bl4ABInJQRDI9y5NE5C0RSfeccdxf/Hkpo45aIjLR83q3ex7XEpG6wOdAS0/dB0WkpadYgqf+A57PZmqp1/2BZ9+/icitXuvGisj7ns/ofuCqMuJx/L0TkSYi8pnns7tPRL4r7+8RU1TVfqLkB9gEDC617CpgvudxTSANuBdIAE4HDgDdPOu3AH09j9cBG4HuXut6l7HPbriPQFt6nrcHOnkejwX+47VccX/ZJAJnAbnAx0BToBWwGzi1dNye5wp09jweBByD+6DkWGAXcEGp/bwF1AVqlxHzIGBbZfGXUW468J6n3qOB373+to2ADGAU7qPykZ7njT3rZwLvAsme9+HUcvZxFVAAXAfEAzcC2wHxrJ8LvOD5G/YC0oEzAvisNPP8/Y+qII75pZa9BXwC1Pf8ndYD15ZT/iFggee9TQF+AB4u/ff32n6sJ56hntf9GLDAsy4OWAI8gPtz2xH3Z3OIV9kC4ALPtmW9546/d54YX/IsqwmcXPw+xfJPxAOwH683w50UDgKZXj/ZXh/+k4GdQJxXmWnAWM/jt4E7gOa4k8I44Aagg6euuDL22Rn3l/lgoGapdWM5Mim08lq/F7jU6/kHwG2ex4d9KeGVFMqIYSIwodR+Olbwdyr5Uqoo/lJl4j1fPEd5LXvU6287CvipVJkfPa+jBeACkn14D68C0rye1/G8nuZAG6AIqO+1/jFgsp+fk5rAHODlSuLw/vvHA3lAD69lfwG+Laf8BmCo1/MhwKbSf/9Sn5U5Xs97ADmex/2BLaW2/yfwplfZeeF+73Anvk/K+1zG6o+dKkWfC1S1YfEP8FevdS2Brarq8lq2GfdROriPQgcBpwDzgG9xt0GfCnxXqhwAqpoG3Ib7H3O3iEz3ag4oyy6vxzllPK9X2QsUkf4i8o2nKSELd+JqUmqzrZXV42f8KfzRLl9ss9fjlqWeF69vhfvLfJ+qZvgSE+7EXRxftudhPc8+9qnqgTL24RNP88bbQD5ws6/lcP99Ezj8NVa079J/j82eZRXZ6fU4G0gUd/9AO9zNTZnFP7jPdpt5bV/R+x2q9+5J3GfeX4rIRhG5p4IYYoYlhaplO9CmVLtnW9yn0uBOCifjTgxzgfnAibiTwtzyKlXVd1T1JNz/vAo84Xjkh3sH+BRoo6pJuE/hpXRYvlbmY/zpQCHuL4libb0eb/eUp9T633F/GTUSkYa+xlSO7Z566pexj0qJiACv4/4yHaGqBRVsXvrvtwf30bb3a6xo36X/Hm09y8qquzJbgd+8D3ZUtb6qDq0gXm8hee9U9YCq/l1VOwLDgTvK64+KJZYUqpaFwCHgLhGpKSKDcH+YpwOo6q+4j9avwH06vh/3kfwIykkKItJNRE4XkVq424RzcDdxhFJ93EdvuSLSD/hzoBX5Gr+qFgEfAmNFpI6I9ABGe20yC+gqIn8WkRoicinuJpDPVHUH7s7VF0Qk2fO3P8XfWFV1K+62+cdEJFFEjgWuBab6WMWLQHdguKrmVLLtLqC1iCR49l2Eu03+ERGpLyLtcDc1lndvwTTgfhFJEZEmuPsDirfdBTQWkSQf4/4J2C/uCwJqi0i8iBwtIsf7UjhU752IDBORzp5kux/35ybUn/2oZ0mhClHVfOA84BzcR34vAFeq6lqvzeYCe1V1i9dzAZaWU20t4HFPfTtxdyze63z0h/kr8JCIHMD9ZfNeEHX5E//NuJtxdgKTgTeLV6jqXmAY8HfcfSV3AcNUdY9nk1G4j7TX4u7DKPNKLh+MxN1vsh34CBijql9ByQ1/q8oq5PkS/wvuzumdXlf+XF7Ofr4GVnm2LX4Nt+A+qNiI+yzyHeCNcsr/G1gMrABWAj97luH5vE0DNnqagypsVvJ8qQ/3xP4b7vfqNdyX1voqFO9dF9x9Mwdx90G8oKrf+hFTtVR8RYQxxhhjZwrGGGP+YEnBGGNMCUsKxhhjSlhSMMYYU6LKDTzVpEkTbd++faTDMMaYKmXJkiV7VDWlsu2qXFJo3749ixcvjnQYxhhTpYhI6bu+y2TNR8YYY0pYUjDGGFPCkoIxxpgSlhSMMcaUsKRgjDGmRMiSgoi8ISK7ReSXctaLiDwr7ikJV4hIn1DFYowxxjehPFOYDJxdwfpzcI9S2AW4HvewwFXaxoyN9JzYnxoP1qLnxP5szNgY6ZCMMcYvIUsKqjoP2FfBJucDb6nbAqChiLQIVTzhMHzKSNZ+MIKiRzJY+8HFDJ8yMtIhGWOMXyLZp9CKw6fX20Y5UwOKyPUislhEFqenp4cluECsy1qGa+HNUFAH18KbWJe1LNIhGWOMXyKZFEpPvwjlTMmnqq+oaqqqpqakVHqXdsR0S+oF/Z+DmtnE9Z/kfm6MMVVIJIe52Mbhc6625o85YKukT66cSpfdvYg7418c1bA3M0ZPi3RIxhjjl0gmhU+Bm0VkOtAfyPLMp1plpdRJgVqH2HTbFtoktam8gDHGRJmQJQURmQYMApqIyDZgDFATQFVfwj3Z9lAgDcgGrg5VLOGSkZtB83rNaVyncaRDMcaYgIQsKahqhZfeqHty6JtCtf9IyMjJYOfBnYz7fhxjB42NdDjGGOM3u6PZQT2b9uQfA/9BRk5GpEMxxpiAWFJwUPqhdApdhWTmZUY6FGOMCYglBQd9kfYFX274kr4t+kY6FGOMCYglBQdl5GZwVqezuLX/rZEOxRhjAmJJwUGZuZlk5Wbx5w/+HOlQjDEmIFVujuZodm6Xc9mTvYcbZt4Q6VCMMSYglhQc1L91f7ILsu3qI2NMlWXNRw4a9s4wvtv8HfVr1Se/KD/S4RhjjN/sTMFBGbkZJCUmsevOXZEOxRhjAmJnCg7KyMkgOTGZV5a8wu5DuyMdjjHG+M2SgoNOaH0CKXVTeH3p6zbrmjGmSrLmIwdNvmAyAMmJydbZbIypkuxMwSE5BTmMeG8EAMm1k8nMtaEujDFVj50pOGRfzj5+3PojAA+f9jDJickRjsgYY/xnZwoOyczNJLm2OxHUT6hPbmFuhCMyxhj/WVJwyP68/SVnB1OWT2HCggkRjsgYY/xnScEhA9oM4LurvwOso9kYU3VZUnDIqt2r+HbTt4C7ozkj15KCMabqsY5mh8zZOIe0fWmc1uE0jm12rPUpGGOqJEsKDsnIzSjpaO7auCtdG3eNcETGGOM/az5ySGZuZklH89asraS+khrhiIwxxn92puCQv/T9C3Vq1gGgXkI90valRTgiY4zxnyUFhyQlJtGodqOSxwfzD1LkKiI+Lj7CkRljjO+s+cghl71/GQu3LQQgTuI4ud3J5BTmRDgqY4zxj50pOMS7oxngm9HfRDAaY4wJjJ0pOKR4LoVi/573bzbs2xDBiIwxxn+WFBxyxbFX0LhO45LnX//2Nb9l/hbBiIwxxn/WfOSQxwc/fthzGz7bGFMV2ZmCA/Zk7+HEN048bJmNf2SMqYrsTMEB+3L2HTEn8/gh46kVXytCERljTGDsTMEBpTuZAXYf2s2aPWsiFJExxgTGkoIDirSI7indD1v25YYveXnxyxGKyBhjAmPNRw4Y2GYgA9sMPGxZcqINn22MqXrsTMEB8zbP4+O1Hx+2rGFiQ0sKxpgqx5KCA+ZvmV8yxEWx3i16c8cJd0QoImOMCYwlBQdk5Bw+xAVAs7rNOLndyRGKyBhjAmNJwQEZuRk0TGx42LL07HTaTWwXoYiMMSYwIU0KInK2iKwTkTQRuaeM9W1F5BsRWSoiK0RkaCjjCZVHz3iUS3teetiyhokNyczNRFUjFJUxxvgvZElBROKBScA5QA9gpIj0KLXZ/cB7qtobuAx4IVTxhNLmzM1HLEuITyAhPoFDBYciEJExxgQmlGcK/YA0Vd2oqvnAdOD8Utso0MDzOAnYHsJ4QuYvn/2FX/f9esTyUceOotBVGIGIjDEmMKG8T6EVsNXr+Tagf6ltxgJfisgtQF1gcFkVicj1wPUAbdu2dTzQYHnPz+ztpWEvRSAaY4wJXCjPFKSMZaUb2EcCk1W1NTAUeFtEjohJVV9R1VRVTU1JSQlBqMEpPcFOsdu/uJ0l25dEICJjjAlMKJPCNqCN1/PWHNk8dC3wHoCq/ggkAk1CGFNIPHzawyTVSjpi+YaMDWzbvy0CERljTGBCmRQWAV1EpIOIJODuSP601DZbgDMARKQ77qSQHsKYHKeq3NzvZuLj4o9Yl1zbhrowxlQtIUsKqloI3AzMBtbgvspolYg8JCLneTb7O3CdiCwHpgFXaRW7hnNT5ia6Pte1zHVN6zQltzA3zBEZY0zgQjognqrOAmaVWvaA1+PVwImly1UlGbkZ1KlZp8x1T571ZJijMcaY4NgdzUEqa4iLYst3LueLtC/CHJExxgTOkkKQ6tSswyltTylz3fJdy5m6cmqYIzLGmMDZfApBGtBmAAPaDChzXfFQF8YYU1XYmUKQ3l/9Pv9Z8Z8y1yUnJpORY1cfGWOqDjtTCNLSHUupXbN2met6Ne/FpKGTwhyRMcYEzs4UgpSRm1HmEBfg7m+oGV8zzBEZY0zgLCkEKTM3s9yrj3ILc+n7St8wR2SMMYGz5qMgTb1oKi51lbmuTs06FLmKyC3MJbFGYpgjM8YY/9mZQpA+XPNhuVcYiYhdgWSMqVIsKQTp3q/vJT27/OGa7jrxLhLiE8IYkTHGBM6aj4KUkXPk/Mze7hx4ZxijMcaY4NiZQhBUtdwJdopd/cnVzNk4J4xRGWNM4CwpBEFRPrz0Q2rVqFXuNnmFeew6uCuMURljTOAsKQShyFXE8S2Pr3Cb5MRk62g2xlQZlhSCsHbPWs5464wKt+nauGu5Q2sbY0y0sY7mIGTmZlbYyQzwtxP+FqZojDEmeHamEISM3PLnUii2cNtC3lj6RpgiMsaY4FhSCELL+i05v9v5FW6zbf82ZqyfEaaIjDEmONZ8FITUlqmktkytcJvk2jZ8tjGm6rAzhSA8t/A5Xl78coXb2DAXxpiqxJJCENL2pZFTmFPhNkc3PZqvRn0VpoiMMSY4lhSCkJFb8RAXAPESz8LfF4YpImOMCY4lhSAk1kikeb3mFW4jIlww/QKKXEVhisoYYwJnHc1BeGX4K5VuEydxNKjVgMzcTBrXaRyGqIwxJnB2phCEp394mvRD5Q+bXSy5djIZuXYFkjEm+llSCML4BePJK8qrdLtxg8fRuLadJRhjop9PzUciEq+q1iheii/DXABc2P3CcqfsNMaYaOLrmUKaiDwpIj1CGk0Vkl+UT35RPnVr1q1020vfv5QP13wYhqiMMSY4viaFY4H1wGsiskBErheRBiGMK+rFSzwrbliBiFS6bcNadgObMaZq8CkpqOoBVX1VVQcCdwFjgB0iMkVEOoc0wiiVXZBd6Y1rxWyoC2NMVeFTUhCReBE5T0Q+Ap4BngY6AjOAWSGML2qtSl/FjTNv9Gnb1JaptGvYLsQRGWNM8Hy9T+FX4BvgSVX9wWv5+yJyivNhRb+MnIwK52b29qeefwpxNMYY4wxf+xSuVNVrvROCiJwIoKq3hiSyKJeZm1npXArF5m+Zz4PfPhjiiIwxJni+nik8C/Qptey5MpbFjGOaHePT5agAh/IP8f3W70MckTHGBK/CpCAiA4CBQIqI3OG1qgEQH8rAot3RTY/m6KZH+7Rtw8SGdkezMaZKqKz5KAGohzt51Pf62Q9cHNrQots9c+7h2YXP+rRtcu1kcgp8u1LJGGMiqcIzBVWdC8wVkcmqujlMMVUJ6YfS6ZTcyadtuzbuyi9//SXEERljTPAqPFMQkYmeh8+LyKelfyqrXETOFpF1IpImIveUs82fRGS1iKwSkXcCeA0RkZGb4XNHs6oyccFEVDXEURljTHAq62h+2/P7KX8rFpF4YBJwJrANWCQin6rqaq9tugD/BE5U1QwRaervfiKlc6POtEvy7d4DEeH+r+/nmt7X0KBWTN8IboyJcpU1Hy3xfLlfp6pX+Fl3PyBNVTcCiMh04Hxgtdc21wGTVDXDs7/dfu4jYsadOc6v7ZNrJ5OZm2lJwRgT1Sq9T8EzOmqKiCT4WXcrYKvX822eZd66Al1F5HvPmEpnl1WRZ6ylxSKyOD298vkLwuHGz25kT/Yen7dPTrShLowx0c/Xm9c2Ad+LyL9E5I7in0rKlDVSXOlG9RpAF2AQMBL3gHtHXPyvqq+oaqqqpqakpPgYcmhN+2Ua8eL7VblvnP8GHZI7hDAiY4wJnq83r233/MThviTVF9uANl7PW3vqKL3NAlUtAH4TkXW4k8QiH/cREUWuIg7mHyQpMcnnMq0btLY5FYwxUc+npKCqgYzRsAjoIiIdgN+By4A/l9rmY9xnCJNFpAnu5qSNAewrrPbn7ad+rfrEie8T190z5x5ObXcqV/e+OoSRGWNMcHydeS0F95DZPYHE4uWqenp5ZVS1UERuBmbjvvv5DVVdJSIPAYtV9VPPurNEZDVQBPxDVfcG/GrCpGFiQ3b8fYdfZZITbZ5mY0z087X5aCrwLjAMuAEYDVTa46uqsyg1tLaqPuD1WIE7PD9Vxu5Du1myYwlDuwz1uUzDxIbW0WyMiXq+tn80VtXXgQJVnauq1wAnhDCuqLZy90qe+sG/WzdObX8q/Vr1C1FExhjjDF/PFAo8v3eIyLm4O4xbhyak6OfPsNnFBrUfFJpgjDHGQb6eKfxbRJKAvwN3Aq8Bt4csqijnzwQ7xeZumstVH18VmoCMMcYhvl599JnnYRZwWujCqRrO6HgGqS1T/SoTHxfP+r3rQxSRMcY4o7L5FJ7jyBvOSsTqrGuNazemZf2WfpVJTnQPc2GMMdGssjOFxWGJooq5Z849HNPsGP56/F99LtOkThOa12sewqiMMSZ4lQ2INyVcgVQlGbn+9yk0q9eMr0d/HaKIjDHGGT7NpyAiMwKZT6G6ysjN8Hl+Zm93fnknuYW5IYjIGGOcEbL5FKqz09ufTqdGvs265u0/K/7D3wf8nRb1W4QgKmOMCV6l8yl4fs8NTzhVw90n3R1QueTa7qEuLCkYY6KVT/cpiMgwEVkqIvtEZL+IHBCR/aEOLlqd8dYZAV1JZENdGGOina93NE8ELgJWaoxPNKyqzNs8jzo16/hd9sM/fUij2o1CEJUxxjjD1zuatwK/xHpCADhUcIiE+AQS4v2diA725exj16FdIYjKGGOc4euZwl3ALBGZC+QVL1TV8SGJKoodzD9It8bdAir74uIX6da4G7f0v8XhqIwxxhm+JoVHgIO451Lw/xC5Gmlerzk//+XngMo2TGxodzUbY6Kar0mhkaqeFdJIqoi0fWl8v+V7Rvca7XfZ5MRkfj/wewiiMsYYZ/japzBHRCwpAKt2r+KDNR8EVHZI5yFceNSFDkdkTNWzMWMjPSf2p8aDteg5sT8bM6J+Ft6Y4euZwk3AXSKSh3tuBcE9cVqDkEUWpTJyM/yeS6HY0U2PxqUuhyMypuoZPmUkaz8YgWvhN6ztP4nhjGTVbQsjHZbBxzMFVa2vqnGqWltVG3iex1xCAPdcCg1r+T/EBbjnVDh9SrnTWhsTM9ZlLcO18GYoqINr4U2sy1oW6ZCMR2VDZx+lqmtFpE9Z61U1sB7XKuyKY68gryiv8g3LkJSYxL6cfQ5HZEzV07n+Mazr9yz8dCtx/SfRLalXpEMyHpU1H90BXA887bXM+16FmDvszcrLIqlWUkBl7eojY9xmXf0ep+afy/bBYziqYS9mjJ4W6ZAAd1/H8CkjWZe1jG5J7rg6JneMdFhhVWHzkape73n4InC+qp4GfIN7BrY7QxxbVHpo7kN8nvZ5QGUb127MwDYDHY7ImKrn+y3f89SwsRSNzWPVbQuj5ovX3ddxMUWPZLD2g4sZPmVkpEMKO1+vPrpfVfeLyEnAmcBk3Iki5gQ6bDZA/Vr1mX7xdIcjMqbqmb1hNpsyN9H75d6RDuUw7r6Om2K6r8PXpFDk+X0u8JKqfkKM3sSWkeP/BDveLv/wchsUz8S81emrGdR+EJsyN7H70O5Ih1OibZ1u0O9ZqJkds30dviaF30XkZeBPuIe7qOVH2Wrlmt7X0LlR54DLL9y2kD3ZexyMyJiqRVXZkrWFHik96NOiDz/viJ7rVT4bPZ0W575B/L3JHDXi/ajp6wgnX+9T+BNwNvCUqmaKSAvgH6ELK3pd0/uaoMon1062zmYT00SE3f/YTZzEcU7nc8gpyIl0SCW27t/Kprt+4VD+oYDvR6rqfL1PIVtVP1TVXz3Pd6jql6ENLTq1eLoF2QXZAZdPTkwmKy/LwYiMqVpW7FrBF2lfAHDnwDu5sHt03OW/8+BOLvvgMtIPpdPp2U7E6qDQMdkEFKjcwlz2Zu+ldo3aAdcx+4rZDO442MGojKlaZq6fyde/fQ24h5Mf/bH/44iFwhdpXzC442BaNWhF3YS6/Jb5W6RDighLCn7IzM2kYWJDRCTgOr7f+j3LdsbeFQ3GFFu9ZzU9U3oC7nt3PlzzYVRcfDHz15kM7TwUgONbHs/i7YsjHFFkWFLwQ6GrkNM7BHe/3ox1M/j818DuczCmOli1exU9UnoAECdx9Grei6U7l0Y4KrhzwJ2cf9T5gHvkgiZ1mkQ4osjwtaPZAK0btA76PoPk2sk21IWJaW9f+DadGnUqed6neR/W710f9AFXMLYf2E6bpDYl0+Ve1P2iiMUSaXam4Icl25cw/sfgJptLTrSrj0zsysrNokiLSKyRWLJswtkTuCH1hghGBRMXTOTlxS+XPD+Uf4gek3rE5KjGlhT8sHbP2qDbGc/tei639r/VoYiMqVrmbp7LPXPuOWxZdkE2E36cEKGI3Gb+OpOhXYaWPK+bUJecwhzS9qVFMKrIsKTgh8zczKDuZgZIqZMSdB3GVFWr01eX9CcUS6yRyP3f3M+BvAMRiWlT5ibSD6VzfKvjD1ue2jI1JjubLSn4IZhxj4qt2LWC86ef71BExlQtq9JXlVx5VKxGXA2Obno0y3ctj0hMdWvW5Y3z3yBODv86PLPjmeQW5kYkpkiyjmY/3H3i3RS6CoOqI7l2Mhm5kb/8zphI+FOPP3FMs2OOWN6neR+WbF/CSW1PCntMuYW5nNvl3COWX9/3+jK2rv7sTMEPC39fyK5Du4Kqo2Fiw6i4JtuYcFNVhnQeQvuG7Y9Y969T/8XoXuG/iS2nIIeeL/Qsc5SBQlchl75/KUWuojJKVl8hTQoicraIrBORNBG5p4LtLhYRFZHUUMYTrPE/jmfJ9iVB1dEwsSFX9brKmYCMqUI2Zmyk+6TuZa5rUqdJ0P9bgfhm0zf0btG7zGbhGnE1WLJ9Cev3rg97XJEUsqQgIvHAJOAcoAcwUkR6lLFdfeBWIOpn7c7IzQh6kKwacTUYPyS4y1qNqYpWpa+ia+OuZa5TVYZPGx72wfFm/Tqr5C7mssRiZ3MozxT6AWmqulFV84HpQFk9rA8D44Co79EpHuYiWIPfGszWrK0ORGRM1bE6ffURnczFatWoxVFNjmLl7pVhjemUdqdwcY+Ly13ft0Vf1u5ZG8aIIi+UHc2tAO9vvm1Af+8NRKQ30EZVPxORcqf3FJHrcc8VTdu2bUMQqm8eOf0ROiV3qnzDSqRnp7M3Zy9tkto4EJUxVUPzes05ttmx5a4vnluhX6t+YYknpyCHEd1HEB8XX+42dwy4o8L11VEozxTKGjWuZCxaEYkDJgB/r6wiVX1FVVNVNTUlJcXBEP0zoPUAGtRqEHQ9yYnJ1tlsYs5Vva467Aax0v56/F/DevXRpEWTuGP2HRVuEydxPPbdY0FfdViVhDIpbAO8D4VbA9u9ntcHjga+FZFNwAnAp9Ha2VzkKiLlyRSU4MdYP6rJUTF5+7yJXS51cfKbJ5NflF/uNn1a9KF5veZhi2nWr7MqHcZeRJiyfEpMNSGFMiksArqISAcRSQAuAz4tXqmqWaraRFXbq2p7YAFwnqpGZa9OZm4mDWo1OOIGl0C8NOwlzuh4hgNRGVM1bMrcxObMzSTElz+1e3ZBNm0ntK0wcTglKzeLRdsX+TQIX6x1NocsKahqIXAzMBtYA7ynqqtE5CEROS9U+w0VpzqZwX2EMmfjHEfqMqYqWLV7FT2blt3JXKxOzTp0SO7A6vTVIY/nYP5Bxpw6hroJdSvdtm+LvjGVFEJ6R7OqzgJmlVr2QDnbDgplLMFKiE9g5NEjHalryfYl5BXl2QxsJmbsyd7D8S2Pr3S7vi368vOOn+nVvFdI42lerzl3Diz32pbDXHncldanYI7UJqkNj5zxiCN1Jde2jj9I+MEAAB6pSURBVGYTW67ufTUPnfZQpdtd2vNSWtVvFdJYXOqiy3Nd2HXQt9EJGtdpzIpdKygoKghpXNHCkoKPvtzwJQ98U+ZJjt8aJja08Y9MTHnsu8fYfWh3pdud2/VchnQeEtJYlu5YSkJ8As3qNfO5zG2zb2NV+qoQRhU9LCn4aFPmJnYc2OFIXed1O48JQyI7frwx4eJSF//+7t+HTaxTntzCXLo+1zWkzTWzfp1V5gB4FYmlzmZLCj7KyAl+iItigvBb5m+O1GVMtNuUuYnGtRv7dI9PYo1EFGXdnnUhiycpManCu5jLEkudzZYUfFSkRY5dQ711/1au+vgqR+oyJtr5cuWRt+I7m0Pl1v63MqDNAL/KnN/tfK449ooQRRRdbD4FH9178r2O1WXzNJtYMqTzEPq37l/5hh6DOwymwBWaTt2pK6byy+5feGzwY36Va9ewHfUS6lHoKqRGXPX+2rQzBR+9+8u7jt3VWDzRjmrwd0cbE+0WbvNvAOTr+l7HNb2vCUksM9bPoHOjzgGVPW3KaSzfGZnZ4cLJkoKPXl/6OpsyNzlSV2KNRJ468ylHhswwJtrdPvt2NmZs9Hn7gqICLn3/UseHgil0FfLlhi85p8s5AZXv2zI2+hUsKfgoIzeD5ERnOpoBbul/C1LmmIHGVB8udbFmzxp6pBwxlUq5asbXZOG2hWzYt8HRWHYc2MGwrsNoWb9lQOVTW8TGFUiWFHzk5NVHAL1f7h2W2/mNiaRNmZtoVLuR36ML92nRh6U7lzoaS5ukNrx14VsBlz+9w+kc36ryu7KrOksKPvrgTx/QLqmdY/XVqVnHbmAz1V5yYjKvn/e63+X6tOjj+GWpI94bEdTkVt1TunN93+sdjCg6WVLwgapS6CqkVo1ajtVpcyqYWKAog9oP8rvcP0/6J/869V+OxbElawvzNs8LuOmo2PBpw/3uOK9qLCn44GD+QU6dfKqjdQ5sM5D6teo7Wqcx0ea2L27j7eVv+11OUZ6Y/4RjV+h9/uvnnN357KBnUWtet3m171ewpOCDjFxn+xPAfd9DIEdQxlQlq9NX+3XjWrEacTV4ZuEzbMna4kgc2w9s5/xuZU0R75/Ulqks3mFJIeZl5Dh75RHAe6veY+qKqY7Waaq3jRkb6TmxPzUerEXPif39uswzEoqvPOrepHtA5Z28s/nB0x70e2iLsgxoM4B6Nes5EFH0sqTgg8Z1GvO3/n9ztM5t+7dV+9PQWBGuL+vhU0ay9oOLKXokg7UfXMzwKc7M7xEqOQU53NrvVpISkwIq37t5b0eSwrzN83jqh6eCrgfg2GbH8tzQ5xypK1pZUvBB6watubbPtY7WmZyYbFcfVRPh+rJem7kM18KboKAOroU3sS5rWUj245S6CXX9Hk7C2839bubW/rcGHcf7q993dC6E+/53X0jHZoo0Swo+mLpiKrd9cZujdSbXtvGPqot1WaH/slZVamkS0v95qJkN/Z+jW4PQzk4WrGcXPsvTPzwdcPmUuiks3xXcsBKqysxfZzK0y9Cg6vG2N2cv87fMd6y+aGNJwQe7Dvk2Q5M/hncdzkeXfuR4vSb8utQ/Fvo998eXdZLzX9afp31Oy8YNOOqi94m/L5lapz3JdQOiu/loyY4lQc1rLgiXvX9ZwPOYbMzYyFETUtm4dzN/nna9Y816qS1TWbJjiSN1RSNLCj4IRUfzgfwDvLfqPUfrNJHRtWVLkodMIv6+ZGoOeoJ/nn6L4/vYm72XiWdPZPXtP1E4Jo+fbvqaYV2HOb4fJwV65VExEQmqs3n4lJGkfXQZPLaftR8616yX2jI16jv5g2FJwQeN6zSmS+Mujta5P28/d37l28ThJnq9s/Id1u1Zx7Z71lI4Jo9pI1/lmYXPODoC7r6cfYw6btRhSeDYZsfSpE4TXlr8kmP7cVr9hPoBX3lULJiksC5EfTDHNTuOeVfNc6SuaGRJwQe3nXAbfz7mz47W2TCxod3RXA0c3/J4/nvJf6lTsw4AF3W/iP9e8l9EnBnssKCogBNeO4FFvy86Yl3NuJo8+cOTzFw/05F9Oe3r0V8HfOVRsVHHjuLszmf7Xa6gqIAETUL6u5v14vpPcqxZT0SY9su0ajt2mSUFHzz1w1P8svsXR+usn1Cf3MJcR6+KMOFT5Cri4bkP07J+S45pdkzJchGhbVJbbv/idrJys4Lez+Rlk2ndoDWpLVOPWFc3oS4vD3uZG2feyP68/UHvy0k/bv2R5xYGf+lmz6Y96dSok9/lxnw7hn4detD9og+Jvy+Zo0a8z4zR04KOp9j3W75ndtpsx+qLJpYUfPDR2o8cv1JIRJgxcoajdZrweeL7J/hm0zfUrln7iHVxEkdWXhaPfvdoUPvILczloXkP8egZj5Z75jG442CGdR3Gj1t/DGpfTvtuy3eOzEOuqnR8piN7svf4Ve64Zsfx3iXvser2hRSOyWPVbQvpmNwx6HiKVefOZksKPghFRzO4J+1weiIRE3qLfl/ExAUTmXLBFOKk7H+hR05/hNeXvh7UnADxEs/Lw17mhNYnVLjdC+e+wJDOQziQdyDgfTltVfoqeqYE3slcTETo1bwXS3f4Noz2zoM7mbxsMpcefSlN6zYNev/lSW1ZfedWsKTgg4zcjKAurSvPsHeGsWxndN+AFE2iZZiHFbtWMGnoJNoktSl3mxb1W3DvyfeyaPuRfQG+2J+3n8/TPvf5+vol25dw/KvHk1uYG9D+nLZuz7qgrjzy5mtns0tdjPpoVFg+Fz1SevDlqC9Dvp9IqN4zUDtk9V9XB91hVha7gc0/7juHR+Ba+A1r+09iOCNZdVt4hzFeuWulz3e33zHgDgCyC7JLOqJ9Nf7H8WzM2Mh53c7zafu+LftydNOjeWiuu7kp0r67+rtyz6L8dV6380g/lF7pdk/Mf4K8wjweOPUBR/Zbkfi4ePZk70FVadfQuXlWooGdKVSioKiAbzd969gH3FvDxIY21IUf3HcO3xyxYR4+WvMRF757IflF+T6XWbBtASe9cRJFriKfy6QfSue5n57jwUEP+hXf80Of57WfX2PlrpV+lXNa+qF0Pl77cdDDVBcb1H4Ql/S8pMJtXOpi9Z7VvDPiHWrEhedYd8qyKdXyXiNLCpXYk72HG2beEJK6h3UZRtuktn6ViZYmlEhoVqv9H3cO93uWrg2OC9u+tx/Yzo0zb+TtC98mIT7B53L9W/WnbkJdJi+b7HOZN5e9yWU9L6NDcge/Ymxerzmzr5hNtybd/CrntJ9+/4nXlr7mWH2qyjEvHlPu1Vz7cvax/cB23r7wbVo3aO3YfitTXTubLSlUIiM3NJ3MAKOOG8XANgP9KjN8ykjWfDCiyoyU6ZQ92XvIi9tLx/PfIf6+ZGqfMZ4bBlwRtv0/+t2j3Jh6IwPaDPCrnIgwYcgE7v/mfp8vG71z4J2MO3NcIGHSu0VvVu5ayes/+z8FplOc6mQuJiLUT6hfZv+bqnLtp9fywqIXHNufr6prZ7MlhUpk5mY6PsFOsbeWv8W47/3751+XtQyNYBNKpCTVSmL6xdPZ8I/lFI7J490/v86aPWvCsm+Xunj6rKe575T7Aiqf2jKVV4a94tMZxphvxvDzjp+pm1A3oH0BNKrdiLvn3M36vesDriMYq9NX0yOlh6N1ltfZPGnRJLZmbWXMqWMc3Z8vujbuypNnPuno3evRwJJCJTomd/S7bddXuYW5fv/jdqjbw+suzedDMvhatPl+y/d8v/V7BnccXLJsWNdhvDjsxZDve/nO5Qx8fSA142sG1VY9vNtwFmxbwKbMTeVus37vel5Y/AKdkv2/Wctbh+QO3H/K/Vw347qIXPJ814l3MbzrcEfrPLXdqRTp4f0yB/IOMP7H8Uy/eLqj86f7Kj4unjM6nkFWXvA3KUYTSwqVSE5M5rT2p4Wk7oaJDf26+mhz5mY+GvU23Ue479LsfOG7PH/+4yGJLVrkFeZxzafXHNH0IiKs3LWSyz+8PGT7zinI4fIPL+evx//VkQsN5m+Zzz+++ke56//1zb+444Q7HDkzvaXfLbRLaseug86P8FsRl7oochXRrF4zR+u9pOcl3Dnwj7HCCooKqJdQj9U3raZzo86O7ssfY78dy8uLX47Y/kPBkkIlXlnyCrfPvj0kdTev17zMO2LLUuQq4pL/XsLyXctZdZv7Ls07TrmeB759oNqdvnp7fP7j9EzpWealmd2adGPe5nmOT3hS3Jlf95EGbN2TwYltTnSk3r8P+Ds//f4T8zYfOZhaXmEeLnU5MqkMuI9i37rwLTJzMzlqfGrYLkzYkrWFc6aeE5K6r/zoSg7lHwLg2k+v5bWfXyOxRmJI9uWr6jhnsyWFSmTmZobkxjWAU9qdwtsXvu3Tti8veZnEGomHDcz3f33+j5yCHN5e4VsdVU2Rq4i5m+fy7DnPlrk+IT6B20+43e9+mcoUz6Smj2Zx8Ku/cd5bzgyGWLtmbZ4Y/ASTFk06Yp1LXfz3kv8G1ZdQltNeHc66jy4J24UJq3avcuymtdLW7FnD8l3LmbJsCou3L+byY0N3luir6tjZbEmhEqG8+igzN5PHvqt8usKD+Qd5cO6DvDTspcPGwImPi+eFc19g7LdjKXQVOhZXNFz2qqrkF+Xz9eivK7zM8Lo+11G7Zm2/7gOozOEzqd3saGf+pT0v5T8X/uewZfM2z2PQlEGO7cPbnoKtsPCWsF2YsCp9FT2aONvJDO7PZNrOnZz0+qlc+/7feHrI037fEBgKnRt15opjrnD08xdplhQq0TOlJ8c2OzYkdbvUxbgfKj/KrZdQj6V/WVrmFR39WvVj8fWLHb1hp/jOYffR5YiIXPY6ZfkUrvz4ykq3q1+rPm+e/yYFLudGm60rTUIy5DK4+0Lyi/I5fcrpHMg7gKryz//9k5uOv8mxfXjrltSLOM8UntL/ebo16EV2QXZI9gXuocQru9EsEMOnjCTry5vRR7PQufdy52djHd9HIOIkjodPfzgkN7dGjKqG7Ac4G1gHpAH3lLH+DmA1sAL4H9Cusjr79u2r1UVhUaHGPRinRa6icreZnTZbn/7h6UrrunXWrbp0x1JH4oofm6DUPKSgSs1DGj82wZF6fZV+KF2bPtlUl2xf4tP22fnZ2np8a911cFfQ+373l3e1w8QO2n18qsaPTdAeE/rphn0bgq63tFEfjtJ759yrn637THtO6qmFRYWO70NVdcO+DdpjQr+S1zJ95XRtN6Gdzlw/MyT7c7lcIak30p/J8mzYt0GbPtJRZUzNkH1WnAIsVl++t33ZKJAfIB7YAHQEEoDlQI9S25wG1PE8vhF4t7J6w50Ubpp5ky7fuTxk9Td4rIFm5GSUuS47P1s7PtNRP//180rreWXxK3ri6ydWmGB8UeQq0toPpKic9LhS85DGnfSEJj7QRHcc2BFUvf4Y/dFovf2L2/0q85cZf9H7/3d/0Pu+/3/364KtC4KupzLzN8/XuPsaaNyYmtp+3NFh/TL5asNX2umZTnrpfy/Vg3kHHau3yFWkKeNS9EDeAcfqLNZjQj+NO2mc5zM5TntM6Of4PgLRY0I/lZOeKPlfiZa4yhINSWEAMNvr+T+Bf1awfW/g+8rqDXdSOO7F4xw7Ai/Luj3rtKCooMx19/3vPr3kvUt8qqfIVaT9Xu2nk5dODjqmd395V3uM9xxdju+nt39xu3Z+trNuytgUdN2++GHLD7o/d79fZX7d+6s2Gdck4C8kl8ul6/esD6hsIA7/Mgn/l9yh/EP64qIX1eVyadreNEeO8H/L+E1bPd3KgeiOVPqMJ1qOyMs6g9m+f7vjZ35OvP5oSAoXA695PR8FPF/B9s8D95ez7npgMbC4bdu2fv8xgtF2Qlv9LeO3kNW/dMdSzczJLHPd3V/drb/v/93nuhb9vkgn/TQpoDhcLpfe+NmN+snaT8pcP/HHiXrmW2cGVLevcgpydNz8cQH/Q7265FVNP5QeUNk3l76pqa+khqz5o7Roag4ZOnWoDpo8SNftWRdUPTPXzwz5ZyTaHH4G4z5TuPaTa7XV06307q/u1lW7VwW9j+U7l2ujh9p6nb0HdhARDUnhkjKSwnPlbHsFsACoVVm94T5TaDO+Tblf2k4YNHmQztkw57BlLpcrqH/QQMo+9O1D2vul3hUeoWfnZ+vBvIO6evfqgGOryANfP6AXvXtRUHVs3LdR8wvz/SqzJXOLpoxL0WU7lgW1b39EU3NIYVGhTvxxojZ+orFO+HFCwEekn//6uT6z4JkQRhp9yjuCX7Fzhf7jy3/oOf85R1VVf9r2k+4+uLvS+rbv367vrHhHb5hxgx79wtGalZul8zfPVxlTM+iDiGhICj41HwGDgTVAU1/qrU4dzaqqF06/UN9f9f5hy95c+qae8NoJAR21Hsg7oM2ebKY/b//Z5zL/XfVf7TCxg0/9Bl9v/FqbPtlUf9z6o9+xVWRN+hptMq6JbsvaFlQ9Z751pk5ZNsWvMqM+HKWPzHskqP36KxqbQzZnbtYuT/XWuCrSRl6V3PPVPZr0WJKeN+08feGnF0re+85P9dIn5j+h9865V1VVH533qF4w/QId/8N4Xfz74pKzZicOIqIhKdQANgIdvDqae5baprenM7qLr/WGMykcyDugzy18LqT7uObja/TVJa+WPC++8mbx74sDrvPVJa/qgNcG+NTp7HK5NDMn068vpVnrZ2nKuJQjznCC8cyCZxz5W89Om609J/X0q8M9Iyej3H6dWFO6WStuTE3Nzs/2qez/ffJ/UZHcolVWbpa+ufRNbfxwu5IveE58TBs82LKkf6c81aJPwR0DQ4H1ni/++zzLHgLO8zyeA+wClnl+Pq2sznAmhfV71mvHZzqGdB9fbfjqsKP62z6/TW+ddWtQdRZ3Ok9fOb3C7RZsXaCnTzk9oDOSb3/7Vp9f+HygIR7GyStgXC6X9nqpl85YN6PSbX/d+6sOnTo06Cu2qhPvI1I56QmtN7a5bs3aqpszN2va3rRyyxW5irTuI3XLvZLO/CFS/UlRkRRC8RPOpPDTtp+078vhba7KzMn0+8qbsmzL2lZh2/q6Peu0+VPNffryrMgnaz/RqSumBlR2w74N2vXpPsoD8dr16T6OHWUu37m80g7nwqJCPfH1E3XCjxMc2Wd1Ud4R6QerP9Am45ro0KlDddb6WUck0lBeeVTdRKo/yZKCA2anzdYzppwR0n28tuQ1vWXWLZpXmKdXfnSlo53ay3cu10fnPXrE8rzCPO3ybJfDmq0CtXLXSm35dEt9efHLPm2fU5Cj+7L3qapqs0c6KUFeUVGeZTuW6S+7fil3/ZPfP6mnvnmqnSX4ITs/W9/4+Q095c1TNCs3Szfu21jyXn7727c67J1hEY6waohUf5KvSUHc21YdqampunhxeAagysjJYPuB7SEb4AvgnZXvMGP9DI5pegw/bP2BGSNnHDa+UTD25eyjx6QezLp8Fn1a9AHcg8zFx8Wzfu96ujbu6sh+0valMWjyIDSvLrvyN9GtQS9eGTGe3MJcuqd0JyE+gSs+vIJ1e9ex48AO7jrxLh467SHixybgejQTCupAzWzi70umcEyeIzFN+mkSX238io8v+7jM9R+v/Zjjmh3n95SX5g/PLHiGsXPHMqTTEBZtWsPmQ2vpltSLGaOn0TG5Y6TDM6WIyBJVTa10Q18yRzT9hOtMobhpI5TZfMO+Ddr28R7KA/Ead18DnbtpruP7eG3Ja9r/1f5a5CrSvMI8HfL2EJ2dNtvx/XhftcKJj2n8fUl6ypun6PzN8zW/MF9nrJuh6/esP6xDN5Sn0YfyD2nTJ5vqmvQ1hy3PL8zXt5a9ZWcIDtl5YKem/LtDyM74jHOw5qM/BHK61n388UHfLFIZ95fiE55OvcdDso8iV5Fe/O7F2vWpPipjami9sc2DvkmpLIF0noX6NPrBbx/Um2fefMSys/9zdthuUosF0XQjnimfJQUv3keknPS4NniwpV4w/QJVdV+jf9yLx2mXZ7toy6db6pwNc3Rz5mblgfiQf9DD9c8UjvFZoulmrGIH8g7oofxDJc+XbF+iKeNSgr4XwhwuGt97cyRfk0I1Gu+1fN7j47PwFg660rnquKtQVU5pdwqTL5jMZ3/+jIX/t5CT2p5EmwZt6NGwL3H9J4Vk+ORi7mGNQ7sPcL9+XXhzSOYHKDZj9DSOGvE+8fclc9SI95kxeprj+/BXvYR6bMzYyKtLXgVg2sppTBgygVYNWkU4suolGt97E7iY6GjuObE/az+4GNfCm4jrP4mjRrzPqtsWVlhmY8ZGhk8ZybqsZSHrPAvHPiCw119dfLf5Owa9OgypmUu3Br34dPQ7dGrUKdJhGRN2vnY0x0RSCNeXb7SK5dffc2J/Vn9wESy8hbj+z3PUiA9iJiEa482SgjFAjQdrUfRIRkguezWmKvE1KcREn4KJXeHqtzGmunBuYl9jotCM0dMYzkjWDb6/pOnMGFM+SwqmWuuY3NH6EIzxgzUfGWOMKWFJwRhjTAlLCsYYY0pYUjDGGFPCkoIxxpgSlhSMMcaUqHJ3NItIOrA5wOJNgD0OhlPVxPLrj+XXDrH9+u21u7VT1ZTKClS5pBAMEVnsy23e1VUsv/5Yfu0Q26/fXrt/r92aj4wxxpSwpGCMMaZErCWFVyIdQITF8uuP5dcOsf367bX7Iab6FIwxxlQs1s4UjDHGVMCSgjHGmBIxkxRE5GwRWSciaSJyT6TjCScR2SQiK0VkmYhU+2nrROQNEdktIr94LWskIl+JyK+e38mRjDFUynntY0Xkd8/7v0xEhkYyxlARkTYi8o2IrBGRVSLyN8/yWHnvy3v9fr3/MdGnICLxwHrgTGAbsAgYqaqrIxpYmIjIJiBVVWPiBh4ROQU4CLylqkd7lo0D9qnq456DgmRVvTuScYZCOa99LHBQVZ+KZGyhJiItgBaq+rOI1AeWABcAVxEb7315r/9P+PH+x8qZQj8gTVU3qmo+MB04P8IxmRBR1XnAvlKLzwemeB5Pwf3PUu2U89pjgqruUNWfPY8PAGuAVsTOe1/e6/dLrCSFVsBWr+fbCOCPVYUp8KWILBGR6yMdTIQ0U9Ud4P7nAZpGOJ5wu1lEVnial6pl84k3EWkP9AYWEoPvfanXD368/7GSFKSMZdW/3ewPJ6pqH+Ac4CZPE4OJHS8CnYBewA7g6ciGE1oiUg/4ALhNVfdHOp5wK+P1+/X+x0pS2Aa08XreGtgeoVjCTlW3e37vBj7C3ZwWa3Z52lyL2153RziesFHVXapapKou4FWq8fsvIjVxfyFOVdUPPYtj5r0v6/X7+/7HSlJYBHQRkQ4ikgBcBnwa4ZjCQkTqejqdEJG6wFnALxWXqpY+BUZ7Ho8GPolgLGFV/IXocSHV9P0XEQFeB9ao6nivVTHx3pf3+v19/2Pi6iMAz2VYE4F44A1VfSTCIYWFiHTEfXYAUAN4p7q/dhGZBgzCPWzwLmAM8DHwHtAW2AJcoqrVrkO2nNc+CHfTgQKbgL8Ut7FXJyJyEvAdsBJweRbfi7tdPRbe+/Je/0j8eP9jJikYY4ypXKw0HxljjPGBJQVjjDElLCkYY4wpYUnBGGNMCUsKxhhjSlhSMDHFM2LknZGOozIicpWItIx0HCb2WFIwJjpdBVhSMGFnScFUeyJyn2cujTlAN8+yXiKywDNI2EfFg4SJSGcRmSMiy0XkZxHpJCKDROQzr/qeF5GrPI83icijIvKjiCwWkT4iMltENojIDV5l/iEiizz7e9CzrL1n7PtXPePffykitUXkYiAVmOoZ/762iDwuIqs95av1ENgmsiwpmGpNRPriHtakN3ARcLxn1VvA3ap6LO47QMd4lk8FJqnqccBA3AOIVWarqg7AfTfpZOBi4ATgIU8MZwFdcI850wvo6zUoYRfP/noCmcAIVX0fWAxcrqq9gNq4hyfo6Yn33wH8KYzxSY1IB2BMiJ0MfKSq2QAi8ilQF2ioqnM920wB/usZI6qVqn4EoKq5njKV7aN4HK2VQD3PWPYHRCRXRBriHm/qLGCpZ7t6uJPBFuA3VV3mWb4EaF9G/fuBXOA1EZkJfFbGNsY4ws4UTCzwdSyX8r79Czn8fyWx1Po8z2+X1+Pi5zU89T6mqr08P51V9fVSZQGKKONATVULcZ9lfIB7gpgvfHgtxgTEkoKp7uYBF3ra5esDw4FDQIaInOzZZhQw1zP2/DYRuQBARGqJSB1gM9DD8zwJOMPPGGYD13jGuUdEWolIZRO9HACKR7etBySp6izgNtxNUMaEhDUfmWrNM1/tu8Ay3F/u33lWjQZe8nzpbwSu9iwfBbwsIg8BBbhH1NwoIu8BK4Bf+aMZyNcYvhSR7sCPnqaog8AVuM8MyjPZE18O7smRPhGRRNxnHbf7s39j/GGjpBpjjClhzUfGGGNKWFIwxhhTwpKCMcaYEpYUjDHGlLCkYIwxpoQlBWOMMSUsKRhjjCnx/xdCzza4i74GAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Let's plot it!\n",
    "\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "\n",
    "doc_nr = 2\n",
    "data = cosine_similarity(word_count_vector,word_count_vector[doc_nr])\n",
    "\n",
    "doc_similarity = pd.DataFrame({'docs':[item for item in range(0,len(data))],\n",
    "                             'score':[score[0] for score in data]})\n",
    "# naming the x axis \n",
    "plt.xlabel('documents') \n",
    "# naming the y axis \n",
    "plt.ylabel('similarity') \n",
    "# giving a title to my graph \n",
    "plt.title('How similar is doc no. '+str(doc_nr)+' to other docs') \n",
    "plt.plot(data, color='green', linestyle='dashed', linewidth = 1, \n",
    "         marker='o', markerfacecolor='blue', markersize=4) \n",
    "#plt.scatter(doc_similarity['docs'],doc_similarity['score'])\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(2, array([1.])),\n",
       " (18, array([0.49377355])),\n",
       " (21, array([0.47563728])),\n",
       " (14, array([0.30528945])),\n",
       " (12, array([0.29726208]))]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Top 5 similar to doc no. 2. Of course 2 is most simialrt to itself, no surprise!\n",
    "heapq.nlargest(5, enumerate(cosine_similarity(word_count_vector,word_count_vector[2])), key=lambda x: x[1])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# From text to concepts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Latent Dirichlet Allocation (LDA)\n",
    "\n",
    "https://towardsdatascience.com/light-on-math-machine-learning-intuitive-guide-to-latent-dirichlet-allocation-437c81220158"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LDA Model:\n",
      "Topic 0:\n",
      "[('base', 63.5511034235865), ('simulation', 33.92702856977487), ('position', 28.96142794550962), ('structure', 26.351593241571972), ('hydrogen', 25.050559510587156), ('bond', 21.597450524812277), ('anticodon', 21.52416961590458), ('type', 19.19369837085992), ('water', 18.849570767941522), ('modify', 18.825015516011227)]\n",
      "\n",
      "Topic 1:\n",
      "[('pcr', 466.9489397167446), ('use', 373.96378019336), ('sequence', 370.71230462628785), ('rna', 225.86277667399852), ('figure', 225.65714914899704), ('protein', 208.77765882259703), ('site', 201.7074936411588), ('sample', 197.75050966812987), ('probe', 188.64258556735285), ('mutation', 185.78733447472777)]\n",
      "\n",
      "Topic 2:\n",
      "[('trx', 134.33126796930108), ('cd', 100.88335085278402), ('cell', 92.35440672014174), ('signal', 85.82499868761822), ('prf', 71.7894061268765), ('figure', 60.11730820188047), ('genome', 58.62975464340464), ('protein', 47.68103149067885), ('surface', 45.44740748479866), ('disulfide', 44.40624015969484)]\n",
      "\n",
      "Topic 3:\n",
      "[('pcr', 4.336858320999754), ('sequence', 2.915658828750252), ('use', 2.7990726064858675), ('tag', 2.435706444738862), ('tnt', 2.1983728420118314), ('figure', 2.1921750860900855), ('gene', 2.1604446189769773), ('probe', 2.0242878563330278), ('target', 2.015448543716651), ('chip', 2.0120088620062497)]\n",
      "\n",
      "Topic 4:\n",
      "[('cell', 162.4216868494132), ('pmo', 94.9788783297214), ('sequence', 89.35125767668312), ('use', 84.80240861912678), ('sars', 74.00002529577175), ('conjugate', 73.56824376243364), ('figure', 69.3427187391344), ('proseq', 64.08659333542413), ('structure', 62.601866094527146), ('fig', 58.33889294120366)]\n",
      "\n",
      "Topic 0:\n",
      "[('base', 63.5511034235865), ('simulation', 33.92702856977487), ('position', 28.96142794550962), ('structure', 26.351593241571972), ('hydrogen', 25.050559510587156), ('bond', 21.597450524812277), ('anticodon', 21.52416961590458), ('type', 19.19369837085992), ('water', 18.849570767941522), ('modify', 18.825015516011227)]\n",
      "\n",
      "Topic 1:\n",
      "[('pcr', 466.9489397167446), ('use', 373.96378019336), ('sequence', 370.71230462628785), ('rna', 225.86277667399852), ('figure', 225.65714914899704), ('protein', 208.77765882259703), ('site', 201.7074936411588), ('sample', 197.75050966812987), ('probe', 188.64258556735285), ('mutation', 185.78733447472777)]\n",
      "\n",
      "Topic 2:\n",
      "[('trx', 134.33126796930108), ('cd', 100.88335085278402), ('cell', 92.35440672014174), ('signal', 85.82499868761822), ('prf', 71.7894061268765), ('figure', 60.11730820188047), ('genome', 58.62975464340464), ('protein', 47.68103149067885), ('surface', 45.44740748479866), ('disulfide', 44.40624015969484)]\n",
      "\n",
      "Topic 3:\n",
      "[('pcr', 4.336858320999754), ('sequence', 2.915658828750252), ('use', 2.7990726064858675), ('tag', 2.435706444738862), ('tnt', 2.1983728420118314), ('figure', 2.1921750860900855), ('gene', 2.1604446189769773), ('probe', 2.0242878563330278), ('target', 2.015448543716651), ('chip', 2.0120088620062497)]\n",
      "\n",
      "Topic 4:\n",
      "[('cell', 162.4216868494132), ('pmo', 94.9788783297214), ('sequence', 89.35125767668312), ('use', 84.80240861912678), ('sars', 74.00002529577175), ('conjugate', 73.56824376243364), ('figure', 69.3427187391344), ('proseq', 64.08659333542413), ('structure', 62.601866094527146), ('fig', 58.33889294120366)]\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{0: ['base',\n",
       "  'simulation',\n",
       "  'position',\n",
       "  'structure',\n",
       "  'hydrogen',\n",
       "  'bond',\n",
       "  'anticodon',\n",
       "  'type',\n",
       "  'water',\n",
       "  'modify'],\n",
       " 1: ['pcr',\n",
       "  'use',\n",
       "  'sequence',\n",
       "  'rna',\n",
       "  'figure',\n",
       "  'protein',\n",
       "  'site',\n",
       "  'sample',\n",
       "  'probe',\n",
       "  'mutation'],\n",
       " 2: ['trx',\n",
       "  'cd',\n",
       "  'cell',\n",
       "  'signal',\n",
       "  'prf',\n",
       "  'figure',\n",
       "  'genome',\n",
       "  'protein',\n",
       "  'surface',\n",
       "  'disulfide'],\n",
       " 3: ['pcr',\n",
       "  'sequence',\n",
       "  'use',\n",
       "  'tag',\n",
       "  'tnt',\n",
       "  'figure',\n",
       "  'gene',\n",
       "  'probe',\n",
       "  'target',\n",
       "  'chip'],\n",
       " 4: ['cell',\n",
       "  'pmo',\n",
       "  'sequence',\n",
       "  'use',\n",
       "  'sars',\n",
       "  'conjugate',\n",
       "  'figure',\n",
       "  'proseq',\n",
       "  'structure',\n",
       "  'fig']}"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from sklearn.decomposition import LatentDirichletAllocation\n",
    "\n",
    "NUM_TOPICS = 5\n",
    "NUM_TOPICS_ELEMENTS = 10\n",
    "\n",
    "\n",
    "\n",
    "def print_topics(model, count_vectorizer, top_n=10):\n",
    "    lda_dict = {}\n",
    "    for idx, topic in enumerate(model.components_):\n",
    "        print(\"Topic %d:\" % (idx))\n",
    "        topic_members_weights = [(count_vectorizer.get_feature_names()[i], topic[i])\n",
    "                        for i in topic.argsort()[:-top_n - 1:-1]]\n",
    "        print(topic_members_weights)   \n",
    "\n",
    "        topic_members = []\n",
    "        for element in topic_members_weights[:NUM_TOPICS_ELEMENTS]:\n",
    "            topic_members.append(element[0])\n",
    "        lda_dict[idx] = topic_members                   \n",
    "        print()\n",
    "    return lda_dict\n",
    "        \n",
    "# Build a Latent Dirichlet Allocation Model\n",
    "lda_model = LatentDirichletAllocation(n_components=NUM_TOPICS, \n",
    "                                      max_iter=15, \n",
    "                                      learning_method='online')\n",
    "lda_Z = lda_model.fit_transform(word_count_vector)\n",
    "# print(lda_Z.shape)  # (NO_DOCUMENTS, NO_TOPICS)\n",
    "\n",
    "print(\"LDA Model:\")\n",
    "# print_topics(lda_model, count_vectorizer)\n",
    "\n",
    "lda_dict = print_topics(lda_model, count_vectorizer)\n",
    "print_topics(lda_model, count_vectorizer)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Topic 1\n",
      "[7.06108265e-05 9.99717418e-01 7.09319817e-05 6.99246187e-05\n",
      " 7.11141215e-05]\n"
     ]
    }
   ],
   "source": [
    "# Let's chcek what is a topic of doc 2:\n",
    "print('Topic',lda_Z[2].argmax())\n",
    "print(lda_Z[2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Topic 1\n",
      "[9.90119397e-05 9.99602281e-01 1.00486927e-04 9.82261309e-05\n",
      " 9.99940818e-05]\n"
     ]
    }
   ],
   "source": [
    "# Doc 18 was second most similar to 2. Will it have the same topic?\n",
    "print('Topic',lda_Z[18].argmax())\n",
    "print(lda_Z[18])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Search"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.feature_extraction.text import TfidfTransformer\n",
    "\n",
    "count_vectorizer_new = CountVectorizer()\n",
    "word_count_vector_new = count_vectorizer_new.fit_transform(corpus_sentences)\n",
    "tfidf_transformer = TfidfTransformer(smooth_idf=True, use_idf=True)\n",
    "tfidf_transformer.fit(word_count_vector_new)\n",
    "tf_idf_corpus_vector = tfidf_transformer.transform(word_count_vector_new)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32m************************************************************************************************************************************************\u001b[0m\n",
      "Search documents or '0' to stop:  Covid\n",
      "NO RESULTS\n",
      "Press ENTERhydrogen\n",
      "\u001b[32m************************************************************************************************************************************************\u001b[0m\n",
      "Search documents or '0' to stop:  hydrogen\n",
      "\n",
      "\u001b[43mTOPIC 0 ['base', 'simulation', 'position', 'structure', 'hydrogen', 'bond', 'anticodon', 'type', 'water', 'modify']\u001b[0m\n",
      "   1. PMC1636460.xml.json\n",
      "      \u001b[36m310. Simulations contain mcm U seem lose U NH hydrogen bond U P simulation lack modification position close form hydrogen bond simulation time\u001b[0m\n",
      "   1. PMC1636460.xml.json\n",
      "      \u001b[36m374. bulky modification contain hydrogen bond donor acceptor may act sterically restrict movement discourage displacement U weak stack interaction stabilize position hydrogen bond contact across loop\u001b[0m\n",
      "   1. PMC1636460.xml.json\n",
      "      \u001b[36m373. modification remove base longer enthalpically favorable hydrogen bonding interaction unless U rotates solution leave unmodified favorably interact U electrostatic interaction hydrogen bond closer base truncate amino group\u001b[0m\n",
      "   1. PMC1636460.xml.json\n",
      "      \u001b[36m336. general modify unmodified hydrogen bond functional group anticodon loop\u001b[0m\n",
      "   1. PMC1636460.xml.json\n",
      "      \u001b[36m307. side chain mcm U form hydrogen bond often part ASL\u001b[0m\n",
      "   1. PMC1636460.xml.json\n",
      "      \u001b[36m353. Pseudouridine show stabilize tRNA structure water mediate hydrogen bonding network\u001b[0m\n",
      "   1. PMC1636460.xml.json\n",
      "      \u001b[36m372. modify partake hydrogen bond across anticodon loop effectively hold U\u001b[0m\n",
      "   1. PMC1636460.xml.json\n",
      "      \u001b[36m309. model present herein seem correlation presence U NH hydrogen bond U P one conformation\u001b[0m\n",
      "   1. PMC1636460.xml.json\n",
      "      \u001b[36m341. absence ASL rearranges provide new hydrogen bonding partner truncate N position\u001b[0m\n",
      "   1. PMC1636460.xml.json\n",
      "      \u001b[36m377. NMR study indicate SCH weakens ureido hydrogen bonding interaction compare\u001b[0m\n",
      "   1. PMC1636460.xml.json\n",
      "      \u001b[36m305. Simulations contain modify base ψ exhibit additional hydrogen bond exocyclic amine U simulation time respectively\u001b[0m\n",
      "   1. PMC1636460.xml.json\n",
      "      \u001b[36m338. instance wild type trajectory hydroxyl group hydrogen bond simulation time group tRNA\u001b[0m\n",
      "   1. PMC1636460.xml.json\n",
      "      \u001b[36m356. water mediate hydrogen bond ψ believe restrict ψ motion stabilize ψ base pair prevalently found position\u001b[0m\n",
      "   1. PMC1636460.xml.json\n",
      "      \u001b[36m304. simulation contain hydrogen bond C N exocyclic amine vary extent simulation average occupancy\u001b[0m\n",
      "   1. PMC1636460.xml.json\n",
      "      \u001b[36m376. behavior suggests delicate balance force cross strand hydrogen bonding portion prevent displacement U Figure\u001b[0m\n",
      "   1. PMC1636460.xml.json\n",
      "      \u001b[36m339. prefer interact sugar oxygen residue also hydrogen bond base oxygen U U carbonyl\u001b[0m\n",
      "   1. PMC1636460.xml.json\n",
      "      \u001b[36m274. MD trajectory propagate f time step employ SHAKE algorithm hydrogen atom nonbonded cutoff Å\u001b[0m\n",
      "   1. PMC1636460.xml.json\n",
      "      \u001b[36m311. exception trend wild type simulation contain hydrogen bond wild type wild type simulation respectively\u001b[0m\n",
      "   1. PMC1636460.xml.json\n",
      "      \u001b[36m375. wild type simulation bulk substituent hydrogen bond donor acceptor solvent expose U exhibit markedly increase conformational freedom\u001b[0m\n",
      "   1. PMC1636460.xml.json\n",
      "      \u001b[36m370. benzene water pair partially positive hydrogen water molecule interact negative π cloud benzene ring\u001b[0m\n",
      "   1. PMC1636460.xml.json\n",
      "      \u001b[36m246. Modified base generally thought increase stability tRNA structure provide additional hydrogen bonding contact well base base stack interaction metal ion binding site\u001b[0m\n",
      "   1. PMC1636460.xml.json\n",
      "      \u001b[36m308. detectable hydrogen bond observe occur mcm carbonyl oxygen hydroxyl wild type simulation simulation contain mcm U simulation time\u001b[0m\n",
      "   1. PMC1636460.xml.json\n",
      "      \u001b[36m382. However human tRNALys may rotate moiety around C N bond allow side chain hydrogen bond across loop thereby stabilize U correct position ribosome canonical fashion\u001b[0m\n",
      "   1. PMC1636460.xml.json\n",
      "      \u001b[36m290. Criteria hydrogen bonding Å distance heavy atom heavy atom H heavy atom angle\u001b[0m\n",
      "   1. PMC1636460.xml.json\n",
      "      \u001b[36m348. Although full water bridge present production run detailed table Supplementary Data component hydrogen bond ψ NH water P ψ NH water ψ P present large percentage production run typically\u001b[0m\n",
      "\n",
      "\u001b[43mTOPIC 1 ['pcr', 'use', 'sequence', 'rna', 'figure', 'protein', 'site', 'sample', 'probe', 'mutation']\u001b[0m\n",
      "   13. PMC1874606.xml.json\n",
      "      \u001b[36m2488. Deoxyinosine know relatively low melt temperature compare natural base due weaker hydrogen bonding\u001b[0m\n",
      "\n",
      "\u001b[43mTOPIC 4 ['cell', 'pmo', 'sequence', 'use', 'sars', 'conjugate', 'figure', 'proseq', 'structure', 'fig']\u001b[0m\n",
      "   0. PMC1636417.xml.json\n",
      "      \u001b[36m0. surveillance diagnostic application fine scale pathogen identification near neighbor discrimination important therefore assay monitor specific level desirable many type sample clinical environmental\u001b[0m\n",
      "   0. PMC1636417.xml.json\n",
      "      \u001b[36m1. successfully use method base DNA RNA detection assay must couple large database nucleic acid sequence information assay design ensure desire information provide interpretation raw data\u001b[0m\n",
      "   0. PMC1636417.xml.json\n",
      "      \u001b[36m2. Several well establish technique use PCR amplify individual target piece sequence genome provide detection organism\u001b[0m\n",
      "   0. PMC1636417.xml.json\n",
      "      \u001b[36m3. method roughly divide approach target individual short sequence length probe bp method examine longer probe\u001b[0m\n",
      "   0. PMC1636417.xml.json\n",
      "      \u001b[36m4. advantage use short probe uniqueness probe assure unique primer also select method give good specificity\u001b[0m\n",
      "   0. PMC1636417.xml.json\n",
      "      \u001b[36m5. approach capable provide fine scale identification several genetically close organism select sufficient number probe\u001b[0m\n",
      "   0. PMC1636417.xml.json\n",
      "      \u001b[36m6. However rapidly lead large number total probe require detect organism interest\u001b[0m\n",
      "   0. PMC1636417.xml.json\n",
      "      \u001b[36m7. addition select probe initial selection process determine unique often later found less specific organism sequence less specific condition differ original condition\u001b[0m\n",
      "   0. PMC1636417.xml.json\n",
      "      \u001b[36m8. particularly problem organism belonging family high mutation rate also pathogen relatively neighbor pathogen sequence\u001b[0m\n",
      "   0. PMC1636417.xml.json\n",
      "      \u001b[36m9. addition PCR approach focus short unique probe capable detect presence new significant mutation easily resolve base sequence detail\u001b[0m\n",
      "   0. PMC1636417.xml.json\n",
      "      \u001b[36m10. Approaches use longer individual probe avoid many issue cost less specific\u001b[0m\n",
      "   0. PMC1636417.xml.json\n",
      "      \u001b[36m11. issue mean approach suitable provide information desire provide impetus work\u001b[0m\n",
      "   0. PMC1636417.xml.json\n",
      "      \u001b[36m12. High density resequencing microarrays produce variable length segment bp direct sequence information\u001b[0m\n",
      "   0. PMC1636417.xml.json\n",
      "      \u001b[36m13. target sequence fall longer target regime PCR approach rather hybridize longer less specific probe microarray many shorter specific probe place microarray allow detailed determination entire PCR amplicon\u001b[0m\n",
      "   0. PMC1636417.xml.json\n",
      "      \u001b[36m14. also mean specificity primer use relaxed\u001b[0m\n",
      "   0. PMC1636417.xml.json\n",
      "      \u001b[36m15. successfully use detect single nucleotide polymorphism SNP genetic variant viral bacterial eukaryotic genome\u001b[0m\n",
      "   0. PMC1636417.xml.json\n",
      "      \u001b[36m16. use SNP detection clearly establish ability provide reliable quality sequence information\u001b[0m\n",
      "   0. PMC1636417.xml.json\n",
      "      \u001b[36m17. case microarrays design study limited number genetically similar target pathogen many case detection method rely recognize hybridization pattern identification\u001b[0m\n",
      "   0. PMC1636417.xml.json\n",
      "      \u001b[36m18. Taking advantage sequential base resolution capability resequencing microarrays require SNP detection resequencing recently successfully adapt recently use different approach organism identification multiple bacterial viral pathogen allow fine detailed discrimination closely related organism track mutation within target pathogen\u001b[0m\n",
      "   0. PMC1636417.xml.json\n",
      "      \u001b[36m19. new methodology differ earlier work use resolve base query similarity search DNA database identify likely specie variant match base call hybridization observe\u001b[0m\n",
      "   0. PMC1636417.xml.json\n",
      "      \u001b[36m20. system capable test pathogen simultaneously could detect presence multiple pathogen\u001b[0m\n",
      "   0. PMC1636417.xml.json\n",
      "      \u001b[36m21. software program resequencing pathogen identifier REPI use simplify data analysis perform similarity search genetic database use basic local alignment search tool BLAST\u001b[0m\n",
      "   0. PMC1636417.xml.json\n",
      "      \u001b[36m22. REPI program use BLAST default setting would return sequence might represent hybridization expect value quantity calculate BLAST program indicates likelihood sequence match found would occur random chance database\u001b[0m\n",
      "   0. PMC1636417.xml.json\n",
      "      \u001b[36m23. screen case insufficient signal however final determination pathogen detect degree discrimination possible require manual examination return result\u001b[0m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Press ENTER\n",
      "\u001b[32m************************************************************************************************************************************************\u001b[0m\n",
      "Search documents or '0' to stop:  hydrogen BLAST\n",
      "\n",
      "\u001b[43mTOPIC 4 ['cell', 'pmo', 'sequence', 'use', 'sars', 'conjugate', 'figure', 'proseq', 'structure', 'fig']\u001b[0m\n",
      "   17. PMC1751549.xml.json\n",
      "      \u001b[36m3063. Sequence neighbor identify BLAST link Conserved Domain Database CDD RPS BLAST algorithm\u001b[0m\n",
      "   0. PMC1636417.xml.json\n",
      "      \u001b[36m183. Filtering necessary send HybSeq ProSeq BLAST single query bias score strain insertion deletion relative ProSeq especially use BLAST parameter maximize use base call\u001b[0m\n",
      "   0. PMC1636417.xml.json\n",
      "      \u001b[36m22. REPI program use BLAST default setting would return sequence might represent hybridization expect value quantity calculate BLAST program indicates likelihood sequence match found would occur random chance database\u001b[0m\n",
      "   17. PMC1751549.xml.json\n",
      "      \u001b[36m3095. Related Structures link provide one similar protein know structure identify BLAST\u001b[0m\n",
      "   0. PMC1636417.xml.json\n",
      "      \u001b[36m75. also recognize long sequence longer WORD size BLAST may use\u001b[0m\n",
      "   0. PMC1636417.xml.json\n",
      "      \u001b[36m57. NCBI BLAST taxonomy database use query image obtain February\u001b[0m\n",
      "   0. PMC1636417.xml.json\n",
      "      \u001b[36m77. Upon completion algorithm return Task loop perform BLAST subtask\u001b[0m\n",
      "   17. PMC1751549.xml.json\n",
      "      \u001b[36m3094. Related Structure service also integrate NCBI protein BLAST service\u001b[0m\n",
      "   0. PMC1636417.xml.json\n",
      "      \u001b[36m85. result BLAST query expect value return tabular format blastall program\u001b[0m\n",
      "   0. PMC1636417.xml.json\n",
      "      \u001b[36m79. BLAST program use NCBI Blastall blastn version define set parameter\u001b[0m\n",
      "   0. PMC1636417.xml.json\n",
      "      \u001b[36m127. SubSeqs different sample report different bit score top ranked return BLAST\u001b[0m\n",
      "   0. PMC1636417.xml.json\n",
      "      \u001b[36m72. produce group SubSeq contain portion HybSeq chance produce limited list return BLAST query\u001b[0m\n",
      "   0. PMC1636417.xml.json\n",
      "      \u001b[36m37. first portion algorithm handle issue specific microarrays remainder deal sequence suitable use query BLAST algorithm\u001b[0m\n",
      "   0. PMC1636417.xml.json\n",
      "      \u001b[36m88. BLAST algorithm give rank score report accounting size database expect value bit score\u001b[0m\n",
      "   0. PMC1636417.xml.json\n",
      "      \u001b[36m71. subtask us procedure examine HybSeq find long possible subsequence base call SubSeq submit query BLAST\u001b[0m\n",
      "   0. PMC1636417.xml.json\n",
      "      \u001b[36m21. software program resequencing pathogen identifier REPI use simplify data analysis perform similarity search genetic database use basic local alignment search tool BLAST\u001b[0m\n",
      "   17. PMC1751549.xml.json\n",
      "      \u001b[36m3067. database scan run BLAST engine identifies sequence neighbor significant similarity result sequence identifier taxonomy index store Entrez provide Related Sequences link protein record collection\u001b[0m\n",
      "   0. PMC1636417.xml.json\n",
      "      \u001b[36m29. simple criterion expect value cutoff non optimize BLAST parameter use consider pathogen detect REPI algorithm provide list candidate organism make final simple conclusion relate result one prototype sequence another\u001b[0m\n",
      "   17. PMC1751549.xml.json\n",
      "      \u001b[36m3086. Clicking graphical alignment footprint human ACE enzyme complex lisinopril one see text representation correspond BLAST alignment Cn view alignment launch click Get Structure data Figure\u001b[0m\n",
      "\n",
      "\u001b[43mTOPIC 2 ['trx', 'cd', 'cell', 'signal', 'prf', 'figure', 'genome', 'protein', 'surface', 'disulfide']\u001b[0m\n",
      "   10. PMC1781101.xml.json\n",
      "      \u001b[36m2046. BLAST search page open click BLAST main menu bar virus select list BLAST Search Subject selection page single family family select\u001b[0m\n",
      "   10. PMC1781101.xml.json\n",
      "      \u001b[36m2048. BLAST program option select query submit\u001b[0m\n",
      "   10. PMC1781101.xml.json\n",
      "      \u001b[36m2043. GIB V execute BLAST search ORF sequence genome segment\u001b[0m\n",
      "   7. PMC1802563.xml.json\n",
      "      \u001b[36m1337. However BLAST analysis reveal none extension predict encode functional alternative protein domain data show\u001b[0m\n",
      "   10. PMC1781101.xml.json\n",
      "      \u001b[36m2049. user select BLAST Search result present graphical display Graphic View list display Feature List\u001b[0m\n",
      "\n",
      "\u001b[43mTOPIC 1 ['pcr', 'use', 'sequence', 'rna', 'figure', 'protein', 'site', 'sample', 'probe', 'mutation']\u001b[0m\n",
      "   21. PMC2018635.xml.json\n",
      "      \u001b[36m3602. Positive clone sequence analyze BLAST search\u001b[0m\n",
      "   13. PMC1874606.xml.json\n",
      "      \u001b[36m2488. Deoxyinosine know relatively low melt temperature compare natural base due weaker hydrogen bonding\u001b[0m\n",
      "\n",
      "\u001b[43mTOPIC 0 ['base', 'simulation', 'position', 'structure', 'hydrogen', 'bond', 'anticodon', 'type', 'water', 'modify']\u001b[0m\n",
      "   1. PMC1636460.xml.json\n",
      "      \u001b[36m310. Simulations contain mcm U seem lose U NH hydrogen bond U P simulation lack modification position close form hydrogen bond simulation time\u001b[0m\n",
      "   1. PMC1636460.xml.json\n",
      "      \u001b[36m374. bulky modification contain hydrogen bond donor acceptor may act sterically restrict movement discourage displacement U weak stack interaction stabilize position hydrogen bond contact across loop\u001b[0m\n",
      "   1. PMC1636460.xml.json\n",
      "      \u001b[36m373. modification remove base longer enthalpically favorable hydrogen bonding interaction unless U rotates solution leave unmodified favorably interact U electrostatic interaction hydrogen bond closer base truncate amino group\u001b[0m\n",
      "   1. PMC1636460.xml.json\n",
      "      \u001b[36m336. general modify unmodified hydrogen bond functional group anticodon loop\u001b[0m\n",
      "   1. PMC1636460.xml.json\n",
      "      \u001b[36m307. side chain mcm U form hydrogen bond often part ASL\u001b[0m\n",
      "   1. PMC1636460.xml.json\n",
      "      \u001b[36m353. Pseudouridine show stabilize tRNA structure water mediate hydrogen bonding network\u001b[0m\n",
      "   1. PMC1636460.xml.json\n",
      "      \u001b[36m372. modify partake hydrogen bond across anticodon loop effectively hold U\u001b[0m\n",
      "   1. PMC1636460.xml.json\n",
      "      \u001b[36m309. model present herein seem correlation presence U NH hydrogen bond U P one conformation\u001b[0m\n",
      "   1. PMC1636460.xml.json\n",
      "      \u001b[36m341. absence ASL rearranges provide new hydrogen bonding partner truncate N position\u001b[0m\n",
      "   1. PMC1636460.xml.json\n",
      "      \u001b[36m377. NMR study indicate SCH weakens ureido hydrogen bonding interaction compare\u001b[0m\n",
      "   1. PMC1636460.xml.json\n",
      "      \u001b[36m305. Simulations contain modify base ψ exhibit additional hydrogen bond exocyclic amine U simulation time respectively\u001b[0m\n",
      "   1. PMC1636460.xml.json\n",
      "      \u001b[36m338. instance wild type trajectory hydroxyl group hydrogen bond simulation time group tRNA\u001b[0m\n",
      "   1. PMC1636460.xml.json\n",
      "      \u001b[36m356. water mediate hydrogen bond ψ believe restrict ψ motion stabilize ψ base pair prevalently found position\u001b[0m\n",
      "   1. PMC1636460.xml.json\n",
      "      \u001b[36m304. simulation contain hydrogen bond C N exocyclic amine vary extent simulation average occupancy\u001b[0m\n",
      "   1. PMC1636460.xml.json\n",
      "      \u001b[36m376. behavior suggests delicate balance force cross strand hydrogen bonding portion prevent displacement U Figure\u001b[0m\n",
      "   1. PMC1636460.xml.json\n",
      "      \u001b[36m339. prefer interact sugar oxygen residue also hydrogen bond base oxygen U U carbonyl\u001b[0m\n",
      "   1. PMC1636460.xml.json\n",
      "      \u001b[36m274. MD trajectory propagate f time step employ SHAKE algorithm hydrogen atom nonbonded cutoff Å\u001b[0m\n",
      "   1. PMC1636460.xml.json\n",
      "      \u001b[36m311. exception trend wild type simulation contain hydrogen bond wild type wild type simulation respectively\u001b[0m\n",
      "   1. PMC1636460.xml.json\n",
      "      \u001b[36m375. wild type simulation bulk substituent hydrogen bond donor acceptor solvent expose U exhibit markedly increase conformational freedom\u001b[0m\n",
      "   1. PMC1636460.xml.json\n",
      "      \u001b[36m370. benzene water pair partially positive hydrogen water molecule interact negative π cloud benzene ring\u001b[0m\n",
      "   1. PMC1636460.xml.json\n",
      "      \u001b[36m246. Modified base generally thought increase stability tRNA structure provide additional hydrogen bonding contact well base base stack interaction metal ion binding site\u001b[0m\n",
      "   1. PMC1636460.xml.json\n",
      "      \u001b[36m308. detectable hydrogen bond observe occur mcm carbonyl oxygen hydroxyl wild type simulation simulation contain mcm U simulation time\u001b[0m\n",
      "   1. PMC1636460.xml.json\n",
      "      \u001b[36m382. However human tRNALys may rotate moiety around C N bond allow side chain hydrogen bond across loop thereby stabilize U correct position ribosome canonical fashion\u001b[0m\n",
      "   1. PMC1636460.xml.json\n",
      "      \u001b[36m290. Criteria hydrogen bonding Å distance heavy atom heavy atom H heavy atom angle\u001b[0m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Press ENTER\n",
      "\u001b[32m************************************************************************************************************************************************\u001b[0m\n",
      "Search documents or '0' to stop:  0\n"
     ]
    }
   ],
   "source": [
    "from termcolor import colored, cprint\n",
    "\n",
    "def print_results(dictionary):\n",
    "#     dict(sorted(dictionary.items()))\n",
    "    for key in dictionary.keys():\n",
    "        print()\n",
    "        cprint('TOPIC '+str(key)+\" \"+str(lda_dict[key]),on_color=\"on_yellow\")\n",
    "        \n",
    "        for result in dictionary[key]:\n",
    "            print('  ',result[0])\n",
    "            print('     ',colored(str(result[2])+\". \"+result[1], 'cyan'))\n",
    "    \n",
    "       \n",
    "while True:\n",
    "   cprint(144*'*',color=\"green\")\n",
    "   # get a search phrase to extract keywords from\n",
    "   search_phrase = input('Search documents or \\'0\\' to stop:  ')\n",
    "   \n",
    "   preprocessed_search_phrase = text_preprocessor.removeStopwords(\n",
    "       text_preprocessor.lemamatizePosSensitive(\n",
    "           text_preprocessor.wordTokenizeCorpus(\n",
    "               text_preprocessor.cleanSentences(\n",
    "                   text_preprocessor.wordSentTokenize(search_phrase)))))[0]\n",
    "   \n",
    "   if search_phrase == '0':\n",
    "       break\n",
    "   else:\n",
    "                \n",
    "        #generate tf-idf for the given search phrase\n",
    "        tf_idf_vector=tfidf_transformer.transform(count_vectorizer_new.transform([\" \".join(preprocessed_search_phrase)]))\n",
    "                       \n",
    "        query_sim = cosine_similarity(tf_idf_corpus_vector,tf_idf_vector)\n",
    "        \n",
    "        if query_sim.max() == 0.0:\n",
    "            print('NO RESULTS')\n",
    "        else:\n",
    "            nr = 1\n",
    "            results = {}\n",
    "            for sentence_nr in heapq.nlargest(50, \n",
    "                                              enumerate(query_sim), \n",
    "                                              key=lambda x: x[1]):\n",
    "                topic_nr = lda_Z[sentence_dict[sentence_nr[0]]].argmax()\n",
    "                doc_title = docs_list[sentence_dict[sentence_nr[0]]]['title']\n",
    "                if topic_nr in results.keys():\n",
    "                    results[topic_nr].append([str(sentence_dict[sentence_nr[0]])+'. '+docs_list[sentence_dict[sentence_nr[0]]]['title'],corpus_sentences[sentence_nr[0]],sentence_nr[0]])\n",
    "                else:\n",
    "                    results[topic_nr] = []\n",
    "                    results[topic_nr].append([str(sentence_dict[sentence_nr[0]])+'. '+docs_list[sentence_dict[sentence_nr[0]]]['title'],corpus_sentences[sentence_nr[0]],sentence_nr[0]])\n",
    "                    \n",
    "            print_results(results)    \n",
    "#                 print(str(nr)+'.',\n",
    "#                       docs_list[sentence_dict[sentence_nr[0]]]['title'], \n",
    "#                       '( Topic',lda_Z[sentence_dict[sentence_nr[0]]].argmax(),')')\n",
    "#                 print('  ',colored(corpus_sentences[sentence_nr[0]], 'yellow'))\n",
    "#                 nr += 1\n",
    "                \n",
    "        input('Press ENTER')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Zadanie domowe"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "Stwórz wyszukiwarkę dla swoich dokumentów. \n",
    "\n",
    "Zauważ, że w pierwszej komórce mamy zakomentowane biblioteki:\n",
    "\n",
    "# import cv2\n",
    "# import fitz\n",
    "# import docx2txt\n",
    "# import textract\n",
    "# import pytesseract\n",
    "\n",
    "Służą one do parsowania różnych typów dokumentów (zbadaj klasę Parser, aby dowiedzieć się co odpowiada za czytanie czego). Zainstaluj wymagana bibliteki. \n",
    "\n",
    "Stwórz folder z różnymi dokumentami i wyszukuj w tym folderze. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
